{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d12a5cce",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from sklearn import linear_model\n",
    "from sklearn.neural_network import MLPRegressor\n",
    "from sklearn.preprocessing import PolynomialFeatures\n",
    "from sklearn.metrics import r2_score, mean_squared_error\n",
    "from scipy.stats import spearmanr, pearsonr\n",
    "import seaborn as sns\n",
    "from sklearn.decomposition import PCA\n",
    "import umap\n",
    "import hdbscan\n",
    "import sklearn.cluster as cluster\n",
    "from sklearn.metrics import adjusted_rand_score, adjusted_mutual_info_score\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "import shap\n",
    "import random\n",
    "import itertools\n",
    "import warnings\n",
    "from tqdm import tqdm\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "assays = ['DNase', 'H3K36me3', 'H3K27me3', 'H3K27ac', 'H3K4me1', 'H3K4me3', 'H3K9me3', 'Methylation_pos', 'Methylation_neg', 'MNase']\n",
    "assay_index_choice = 10 # MNase\n",
    "\n",
    "RESOLUTION = 25\n",
    "window_size = 401\n",
    "operative_half_window_size = 80 # will lead to a total of the central 2*operative_half_window_size + 1 features being chosen\n",
    "pairwise_flag = True # If true then H3K27ac features will be concatentated with assays[assay_index_choice-1]\n",
    "if(pairwise_flag):\n",
    "    pairwise_multiplier = 0\n",
    "else:\n",
    "    pairwise_multiplier = 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d09d0fd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames = [\"pos_\" + str(x) for x in list(range(-(window_size//2), (window_size//2) + 1))]\n",
    "colnames = colnames + [\"log10p1(TPM)\", \"cell_type\", \"chrom\", \"position\", \"strand\", \"assay_index\"] \n",
    "df = pd.read_csv('../Data/MNase.Training_Data.csv', sep=\",\", names=colnames, low_memory=False)\n",
    "df_sorted = df.sort_values(['chrom', 'position', 'cell_type', 'assay_index'], ascending=[1, 1, 1, 1])\n",
    "df_sorted_unique = df_sorted.drop_duplicates()\n",
    "df_H3K27ac = df_sorted_unique.iloc[range(4-1, len(df_sorted_unique), len(assays)), :]\n",
    "df_other_assay = df_sorted_unique.iloc[range(assay_index_choice-1, len(df_sorted_unique), len(assays)), :]\n",
    "df_merged = pd.merge(df_H3K27ac, df_other_assay, how='inner', on=['cell_type', 'chrom', 'position', 'strand'], suffixes=['_H3K27ac', '_'+assays[assay_index_choice-1]])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fb5bf796",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "'continue' not properly in loop (<ipython-input-3-38f6727b59d6>, line 32)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-3-38f6727b59d6>\"\u001b[0;36m, line \u001b[0;32m32\u001b[0m\n\u001b[0;31m    continue\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m 'continue' not properly in loop\n"
     ]
    }
   ],
   "source": [
    "middle_position_first_feature = window_size // 2\n",
    "middle_position_second_feature = -1 + ( (window_size + 6) + (window_size + 6 + window_size + 1) )//2\n",
    "\n",
    "df_two_assays = df_merged.iloc[:, \n",
    "                               np.r_[middle_position_first_feature-(operative_half_window_size):\n",
    "                                     middle_position_first_feature+(operative_half_window_size)+1, \n",
    "                                     middle_position_second_feature-(operative_half_window_size):\n",
    "                                     middle_position_second_feature+(operative_half_window_size)+1,\n",
    "                                     window_size,\n",
    "                                     window_size + 1,\n",
    "                                     window_size + 2,\n",
    "                                     window_size + 3]]\n",
    "\n",
    "df_subsampled = df_two_assays\n",
    "\n",
    "# Train on all but CXCR4 and TGFBR1 chromosomes\n",
    "odd_chroms = [\"chr\" + str(c) for c in range(1, 23, 1)]\n",
    "odd_chroms.remove(\"chr2\") # exclude chr2 where CXCR4 lies\n",
    "odd_chroms.remove(\"chr9\") # exclude chr2 where TGFBR1 lies\n",
    "\n",
    "df_odd_chroms = df_subsampled.loc[df_subsampled['chrom'].isin(odd_chroms)]\n",
    "xTrain = df_odd_chroms.iloc[:, pairwise_multiplier*(2 * operative_half_window_size + 1):2 * (2 * operative_half_window_size + 1)] \n",
    "yTrain = df_odd_chroms.iloc[:, 2 * (2 * operative_half_window_size + 1)]\n",
    "\n",
    "even_chroms = [\"chr\" + str(c) for c in range(2, 23, 2)]\n",
    "df_even_chroms = df_subsampled.loc[df_subsampled['chrom'].isin(even_chroms)]\n",
    "xTest = df_even_chroms.iloc[:, pairwise_multiplier*(2 * operative_half_window_size + 1):2 * (2 * operative_half_window_size + 1)] \n",
    "yTest = df_even_chroms.iloc[:, 2 * (2 * operative_half_window_size + 1)]\n",
    "\n",
    "if(len(xTrain) < 10):\n",
    "    print(\"len(xTrain) = \"+str(len(xTrain)))\n",
    "    continue\n",
    "\n",
    "quadratic_transform = PolynomialFeatures(degree=1, interaction_only=True).fit(xTrain)\n",
    "xTrain = quadratic_transform.transform(xTrain)\n",
    "xTest = quadratic_transform.transform(xTest)\n",
    "\n",
    "print(xTrain.shape, xTest.shape, yTrain.shape, yTest.shape)\n",
    "\n",
    "# Fit a ridge regression model\n",
    "l1_alpha = 0.001\n",
    "model = linear_model.ElasticNet(alpha=l1_alpha, l1_ratio=0.0, max_iter=1000)\n",
    "\n",
    "xTrain = np.asarray(xTrain)\n",
    "yTrain = np.asarray(yTrain)\n",
    "xTest = np.asarray(xTest)\n",
    "yTest = np.asarray(yTest)\n",
    "\n",
    "model.fit(xTrain, yTrain)\n",
    "\n",
    "# Compute statistics\n",
    "yTrain_Pred = model.predict(xTrain)\n",
    "yTest_Pred = model.predict(xTest)\n",
    "mse_Train = mean_squared_error(yTrain, yTrain_Pred)\n",
    "pc_Train, _ = pearsonr(yTrain, yTrain_Pred)\n",
    "sc_Train, _ = spearmanr(yTrain, yTrain_Pred)\n",
    "mse_Test = mean_squared_error(yTest, yTest_Pred)\n",
    "pc_Test, _ = pearsonr(yTest, yTest_Pred)\n",
    "sc_Test, _ = spearmanr(yTest, yTest_Pred)\n",
    "\n",
    "print(\"Train: MSE = \", round(mse_Train, 3), \"Pearson =\", round(pc_Train, 3), \"Spearman =\", round(sc_Train, 3))\n",
    "print(\"Test: MSE = \", round(mse_Test, 3), \"Pearson =\", round(pc_Test, 3), \"Spearman =\", round(sc_Test, 3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "333ad3aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "inserted_pvalue_choice = 1.5\n",
    "peak_width_choice = {}\n",
    "peak_width_choice[\"CXCR4\"] = 6\n",
    "peak_width_choice[\"TGFBR1\"] = 6\n",
    "assay_index = assay_index_choice\n",
    "\n",
    "# Perform inference for epigenome editing data\n",
    "def ise(gene, trained_model, assay_index, inserted_lnp1_minuslog10p_value = 3, peak_width = 2, pairwise_features=False):\n",
    "\n",
    "    X = np.load(\"../Data/\" + gene + \".npy\")\n",
    "\n",
    "    # obtain the middle portion of this\n",
    "    X = X[:, (window_size//2)-operative_half_window_size:(window_size//2)+operative_half_window_size+1, :]\n",
    "\n",
    "    # Perform inference by introducing p-value changes with a peak width\n",
    "    yPred = []\n",
    "    center = operative_half_window_size\n",
    "    positions = range(center - center, center + center + 1)\n",
    "    for pos in positions:\n",
    "        X_modified = np.copy(X)\n",
    "        for p in range(pos - peak_width // 2, pos + peak_width // 2 + 1):\n",
    "            if( (p>=0) and (p < max(positions)) ):\n",
    "                if(X_modified[:, p, 2] > 10): # Remember this is ln( -log10(p-value) + 1)\n",
    "                    # If H3K27me3 peak exists, then p300 doesn't work\n",
    "                    print(\"H3K27me3 exists!\")\n",
    "                    pass\n",
    "                else:\n",
    "                    if (pairwise_features):\n",
    "                        # Modify the H3K27ac peak                                                \n",
    "                        X_modified[:, p, 3] = X_modified[:, p, 3] + (X_modified[:, p, 9] * inserted_lnp1_minuslog10p_value)\n",
    "                    else:\n",
    "                        # Modify the assay itself\n",
    "                        X_modified[:, p, assay_index] += inserted_lnp1_minuslog10p_value\n",
    "\n",
    "        # Prepare input\n",
    "        if (pairwise_features):\n",
    "            X_modified = np.concatenate([X_modified[:, :, 3], X_modified[:, :, assay_index]], axis=1)\n",
    "        else:                 \n",
    "            X_modified = X_modified[:, :, assay_index]\n",
    "\n",
    "        yPred_value = trained_model.predict(quadratic_transform.transform(X_modified))\n",
    "        yy = yPred_value[0]\n",
    "        yPred.append(yy)\n",
    "\n",
    "    # Prepare input for predicting native expression\n",
    "    if(pairwise_features):\n",
    "        X = np.concatenate([X[:, :, 3], X[:, :, assay_index]], axis=1) \n",
    "    else:\n",
    "        X = X[:, :, assay_index]\n",
    "\n",
    "    # Instead of scaling, divide by yPred\n",
    "    yPred_value = trained_model.predict(quadratic_transform.transform(X))[0] + 0.00000001 # to avoid divby0\n",
    "    print(\"Predicted TPM for \", gene, \" = \", (np.power(10, yPred_value) -1))\n",
    "    yPred = (np.power(10, yPred) -1) / (np.power(10, yPred_value) -1)\n",
    "\n",
    "    return yPred\n",
    "\n",
    "\n",
    "def p_value_mapping(inserted_lnp1_minuslog10p_value):\n",
    "    minuslog10p_value = np.expm1(inserted_lnp1_minuslog10p_value)\n",
    "    p_value = np.power(10, -1 * minuslog10p_value)\n",
    "    return round(minuslog10p_value, 4)\n",
    "\n",
    "\n",
    "def convert_to_2D(idx, nrows, ncols):\n",
    "    return idx//ncols, idx%ncols\n",
    "\n",
    "\n",
    "# Load p300 epigenome editing data\n",
    "df_p300_epigenome_editing = pd.read_csv(\"../Data/p300_epigenome_editing_dataset.tsv\", sep=\"\\t\")\n",
    "\n",
    "TSS = {}\n",
    "STRANDS = {}\n",
    "CHROMS = {}\n",
    "GENES = {}\n",
    "gRNA_STRANDS = {}\n",
    "for index in range(len(df_p300_epigenome_editing)):\n",
    "    tss = df_p300_epigenome_editing.iloc[index, 13]\n",
    "    gene_strand = df_p300_epigenome_editing.iloc[index, 3]\n",
    "    chrom = df_p300_epigenome_editing.iloc[index, 2]\n",
    "    gene = df_p300_epigenome_editing.iloc[index, 0]\n",
    "\n",
    "    TSS[gene] = int(tss)\n",
    "\n",
    "    if(gene_strand == \"plus\"):\n",
    "        STRANDS[gene] = \"+\"\n",
    "    elif(gene_strand == \"minus\"):\n",
    "        STRANDS[gene] = \"-\"\n",
    "    else:\n",
    "        print(\"something wrong with strand!\")\n",
    "\n",
    "    CHROMS[gene] = chrom\n",
    "    GENES[gene] = 1                                                    \n",
    "\n",
    "GENES_LIST = set(list(GENES.keys()))    \n",
    "\n",
    "df_GENES_values = {}\n",
    "df_GENES_means = {}\n",
    "for gene in GENES_LIST:\n",
    "    df_GENES_values[gene] = df_p300_epigenome_editing[df_p300_epigenome_editing[\"p300 target gene\"] == gene]\n",
    "    df_GENES_values[gene][\"Position_wrt_TSS\"] = pd.to_numeric(df_GENES_values[gene][\"gRNA position  wrt TSS (hg38)\"], errors='coerce')/RESOLUTION\n",
    "    df_GENES_values[gene][\"gRNA_strand\"] = df_GENES_values[gene][\"gRNA target strand\"].apply(lambda s: \"red\" if(s==\"minus\") else \"blue\")\n",
    "    df_GENES_values[gene][\"gRNA_strand_bool\"] = df_GENES_values[gene][\"gRNA target strand\"].apply(lambda s: -1 if(s==\"minus\") else +1)\n",
    "\n",
    "    df_GENES_means[gene] = df_GENES_values[gene].groupby('Position_wrt_TSS').mean()\n",
    "    df_GENES_means[gene].index.name = 'Position_wrt_TSS'\n",
    "    df_GENES_means[gene].reset_index(inplace=True)\n",
    "\n",
    "# Perform in-silico epigenesis\n",
    "assay_color = ['black', 'red', 'green', 'blue', 'cyan', 'pink', 'brown']\n",
    "xticklabels = range(-operative_half_window_size, operative_half_window_size + 1)\n",
    "\n",
    "GENES_LIST = [\"CXCR4\", \"TGFBR1\"]\n",
    "\n",
    "fig, axes = plt.subplots(nrows=len(GENES_LIST), ncols=2, figsize=(40, 30), sharey=False)\n",
    "fig.tight_layout(pad=1, w_pad=20, h_pad=25)\n",
    "\n",
    "for idx, gene in enumerate(sorted(GENES_LIST)):\n",
    "\n",
    "    idx_x, idx_y = convert_to_2D(idx, nrows=len(GENES_LIST), ncols=1)\n",
    "    ax_1 = axes[idx_x, 0]\n",
    "    ax_2 = axes[idx_x, 1]\n",
    "\n",
    "    gene_features = np.squeeze(np.load(\"../Data/\" + gene + \".npy\"), axis=0)\n",
    "    gene_features = gene_features[(window_size//2)-operative_half_window_size:(window_size//2)+operative_half_window_size+1, :]\n",
    "\n",
    "    df_values = df_GENES_values[gene]\n",
    "    df_means = df_GENES_means[gene]\n",
    "\n",
    "    inserted_lnp1_minuslog10p_value = inserted_pvalue_choice\n",
    "    peak_width = peak_width_choice[gene]\n",
    "    gene_ise = ise(gene, model, assay_index-1, inserted_lnp1_minuslog10p_value, peak_width, pairwise_flag)\n",
    "\n",
    "    # Create a scatter plot of the means with the predictions of those positions\n",
    "    gene_ise_at_means = []\n",
    "    alan_means = []\n",
    "    gRNA_strands = []\n",
    "\n",
    "    for p_idx in range(len(df_means)):\n",
    "        position_m = df_means.iloc[p_idx, 0]\n",
    "        alan_mean = df_means.iloc[p_idx, 1]\n",
    "        gRNA_strand = df_means.iloc[p_idx, 9]\n",
    "        if(position_m + operative_half_window_size < 0):\n",
    "            continue\n",
    "        elif(position_m > operative_half_window_size):\n",
    "            continue\n",
    "        else:\n",
    "            gene_ise_at_means.append(gene_ise[int(position_m) + operative_half_window_size])\n",
    "            alan_means.append(alan_mean)\n",
    "            gRNA_strands.append(gRNA_strand)\n",
    "\n",
    "    pc, pp = pearsonr(list(alan_means), gene_ise_at_means)\n",
    "    sc, sp = spearmanr(list(alan_means), gene_ise_at_means)\n",
    "\n",
    "    gRNA_strands_colors = []\n",
    "    for iii in gRNA_strands:\n",
    "        if(iii == -1):\n",
    "            gRNA_strands_colors.append(\"red\")\n",
    "        else:\n",
    "            gRNA_strands_colors.append(\"blue\")\n",
    "\n",
    "    ax_1.scatter(list(alan_means), gene_ise_at_means, color=\"#FF1493\", s=1000) # =\"#FF1493\")\n",
    "    ax_1.set_xlim(0, 1.1 * max(alan_means))\n",
    "    ax_1.set_ylim(0, 1.1 * max(gene_ise_at_means))\n",
    "    ax_1.tick_params(axis='both', which='major', labelsize=40)\n",
    "    ax_1.tick_params(axis='both', which='minor', labelsize=40)\n",
    "    ax_1.set_xlabel(\"Mean experimental fold change\", size=60)\n",
    "    ax_1.set_ylabel(\"Model prediction's fold change\", size=45)\n",
    "#     if(pp > 0.05):\n",
    "#         p_asterisk = \"NS\"\n",
    "#     if(sp > 0.05):\n",
    "#         s_asterisk = \"NS\"\n",
    "    ax_1.set_title(\"Correlation between experimental and model predictions fold change\\nPearson = \"+\n",
    "                   str(round(pc, 2))+\n",
    "                   \" (\"+str(round(pp, 3))+\n",
    "                   \") Spearman = \"+\n",
    "                   str(round(sc, 2))+\n",
    "                   \" (\"+str(round(sp, 3))+\n",
    "                   \")\", size=40)\n",
    "\n",
    "    # Determine whether we are doing H3K27ac ISE in the background of another assay's features\n",
    "    # or we have marginal features and are doing that track's ISE\n",
    "    if(pairwise_flag):\n",
    "        epigenetic_features = gene_features[:, 9] # MNase\n",
    "        epigenetic_features_2 = gene_features[:, 3] # H3K27ac\n",
    "        color_for_assay = assay_color[3]\n",
    "        label_for_assay = assays[3]\n",
    "    else:\n",
    "        epigenetic_features = gene_features[:, assay_index-1]\n",
    "        color_for_assay = assay_color[assay_index-1]\n",
    "        label_for_assay = assays[assay_index-1]\n",
    "\n",
    "    # Scale the model predictions     \n",
    "    scaling_ratio = np.median(df_means['Measured fold change'])/np.median(gene_ise - 0.0)\n",
    "    scaled_model_predictions = 0.2 * (scaling_ratio * (gene_ise - 0.0)) + 1.0\n",
    "\n",
    "    # Scale the epigenetic features\n",
    "    epigenetic_features_scaling_ratio = max(df_means['Measured fold change'])/max(epigenetic_features - 0.0)\n",
    "    scaled_epigenetic_features = (epigenetic_features_scaling_ratio * (epigenetic_features - 0.0)) + 0.5\n",
    "\n",
    "    scaled_epigenetic_features_2 = (epigenetic_features_scaling_ratio * (epigenetic_features_2 - 0.0)) + 0.5 \n",
    "\n",
    "    ax_2.plot(xticklabels, scaled_model_predictions, 'o-', color=\"#4daf4a\", linewidth=5, markersize=2, label=\"(Scaled) Model Predictions \" + label_for_assay)\n",
    "    ax_2.plot(xticklabels, scaled_epigenetic_features, 'o-', color=\"#8470FF\", linewidth=5, markersize=1, label=\"(Scaled) Epigenetic Features MNase\") # + label_for_assay)\n",
    "    ax_2.plot(xticklabels, scaled_epigenetic_features_2, 'o-', color=\"darkblue\", linewidth=5, markersize=1, label=\"(Scaled) Epigenetic Features H3K27ac\") # + label_for_assay)\n",
    "\n",
    "    ax_2.bar(df_means['Position_wrt_TSS'], 0.0 + (df_means['Measured fold change']), color=\"#f781bf\", bottom=0, width=2, label=\"Experimental mean from qPCR\")\n",
    "\n",
    "    gRNA_strand_groups = df_values.groupby(\"gRNA target strand\")\n",
    "\n",
    "    color_dict = {}\n",
    "    color_dict[\"+\"] = {\"plus\":\"blue\", \"minus\":\"red\"}\n",
    "    color_dict[\"-\"] = {\"plus\":\"red\", \"minus\":\"blue\"}\n",
    "    direction = {\"blue\":\"rightward\", \"red\":\"leftward\"}\n",
    "    color_index = 0\n",
    "    for name, group in gRNA_strand_groups:\n",
    "        ax_2.plot(group['Position_wrt_TSS'], 0.0 + (group['Measured fold change']), 'o', color=color_dict[STRANDS[gene]][name], label=\"qPCR facing \"+direction[color_dict[STRANDS[gene]][name]], markersize=15)\n",
    "        color_index += 1\n",
    "\n",
    "    ax_2.set_xlim(-operative_half_window_size-10, operative_half_window_size+10)\n",
    "    ax_2.set_ylim(-1, 1.0 + max(df_means['Measured fold change'])*1.5)\n",
    "    x_vals = ax_2.get_xticks()\n",
    "    ax_2.set_xticklabels(['{:3.0f}'.format(x * RESOLUTION) for x in x_vals])\n",
    "    ax_2.yaxis.set_major_locator(MaxNLocator(integer=True))\n",
    "    ax_2.tick_params(axis='both', which='major', labelsize=35)\n",
    "    ax_2.tick_params(axis='both', which='minor', labelsize=35)\n",
    "    ax_2.set_xlabel(\"Peak Position (in bp) w.r.t TSS\", size=50)\n",
    "    ax_2.set_ylabel(\"Gene expression fold change\", size=50)\n",
    "    ax_2.set_title(gene+\" with H3K27ac + \"+assays[assay_index-1]+\"\\nincreasing \" + str(peak_width * RESOLUTION) + \"bp peaks by -log10(p_value)=\"+str(p_value_mapping(inserted_lnp1_minuslog10p_value)), size=40) #, y=1.1)\n",
    "\n",
    "    ax_2.legend(loc='upper center', prop={'size': 30}, ncol=2)\n",
    "\n",
    "plt.show()\n",
    "plt.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
