{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a3f83a6a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "from chrmt_generator import TranscriptomeGenerator, TranscriptomePredictor\n",
    "from chrmt_train import maximum_likelihood_loss\n",
    "from tensorflow.keras.models import load_model\n",
    "from tqdm import tqdm\n",
    "from scipy.stats import spearmanr, pearsonr\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "\n",
    "RESOLUTION = 100\n",
    "window_size = 41\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8f6701f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /export/home/users/sbatra/.local/lib/python3.6/site-packages/tensorflow_core/python/ops/init_ops.py:97: calling GlorotUniform.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Call initializer instance with the dtype argument instead of passing it to the constructor\n",
      "WARNING:tensorflow:From /export/home/users/sbatra/.local/lib/python3.6/site-packages/tensorflow_core/python/ops/init_ops.py:97: calling Zeros.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Call initializer instance with the dtype argument instead of passing it to the constructor\n",
      "WARNING:tensorflow:From /export/home/users/sbatra/.local/lib/python3.6/site-packages/tensorflow_core/python/ops/init_ops.py:97: calling Ones.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Call initializer instance with the dtype argument instead of passing it to the constructor\n",
      "WARNING:tensorflow:From /export/home/users/sbatra/.local/lib/python3.6/site-packages/tensorflow_core/python/ops/resource_variable_ops.py:1630: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "If using Keras pass *_constraint arguments to layers.\n",
      "WARNING:tensorflow:Large dropout rate: 0.75 (>0.5). In TensorFlow 2.x, dropout() uses dropout rate instead of keep_prob. Please ensure that this is intended.\n",
      "WARNING:tensorflow:Large dropout rate: 0.75 (>0.5). In TensorFlow 2.x, dropout() uses dropout rate instead of keep_prob. Please ensure that this is intended.\n",
      "WARNING:tensorflow:Large dropout rate: 0.75 (>0.5). In TensorFlow 2.x, dropout() uses dropout rate instead of keep_prob. Please ensure that this is intended.\n",
      "WARNING:tensorflow:Large dropout rate: 0.75 (>0.5). In TensorFlow 2.x, dropout() uses dropout rate instead of keep_prob. Please ensure that this is intended.\n",
      "WARNING:tensorflow:Large dropout rate: 0.75 (>0.5). In TensorFlow 2.x, dropout() uses dropout rate instead of keep_prob. Please ensure that this is intended.\n"
     ]
    }
   ],
   "source": [
    "# Prepare model to perform inference with\n",
    "trained_model_path = \"../jeff/long_train_dropout_0.75_transcriptome_41_128_3_0.0_mse_tss.model-300.hdf5\"\n",
    "trained_model = load_model(trained_model_path)\n",
    "\n",
    "# trained_model_path = \"../Models/Run_transcriptome_101_16_10_0.0_mle_genome_wide.model-150.hdf5\"\n",
    "# trained_model = load_model(trained_model_path, custom_objects={'maximum_likelihood_loss': maximum_likelihood_loss})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4122733",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading TSS data\n",
      "TSS strand information is invalid\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T01A02.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T01A03.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T01A04.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T01A05.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T01A06.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T01A07.chr2.npy\n",
      "Loading Transcriptome data /scratch/sanjit/ENCODE_Imputation_Challenge/2_April_2020/Data/Gene_Expression/genome_wide_TPM_npy/T01_TPM.chr2.+.npy /scratch/sanjit/ENCODE_Imputation_Challenge/2_April_2020/Data/Gene_Expression/genome_wide_TPM_npy/T01_TPM.chr2.-.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T02A02.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T02A03.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T02A04.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T02A05.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T02A06.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T02A07.chr2.npy\n",
      "Loading Transcriptome data /scratch/sanjit/ENCODE_Imputation_Challenge/2_April_2020/Data/Gene_Expression/genome_wide_TPM_npy/T02_TPM.chr2.+.npy /scratch/sanjit/ENCODE_Imputation_Challenge/2_April_2020/Data/Gene_Expression/genome_wide_TPM_npy/T02_TPM.chr2.-.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T03A02.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T03A03.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T03A04.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T03A05.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T03A06.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T03A07.chr2.npy\n",
      "Loading Transcriptome data /scratch/sanjit/ENCODE_Imputation_Challenge/2_April_2020/Data/Gene_Expression/genome_wide_TPM_npy/T03_TPM.chr2.+.npy /scratch/sanjit/ENCODE_Imputation_Challenge/2_April_2020/Data/Gene_Expression/genome_wide_TPM_npy/T03_TPM.chr2.-.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T04A02.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T04A03.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T04A04.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T04A05.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T04A06.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T04A07.chr2.npy\n",
      "Loading Transcriptome data /scratch/sanjit/ENCODE_Imputation_Challenge/2_April_2020/Data/Gene_Expression/genome_wide_TPM_npy/T04_TPM.chr2.+.npy /scratch/sanjit/ENCODE_Imputation_Challenge/2_April_2020/Data/Gene_Expression/genome_wide_TPM_npy/T04_TPM.chr2.-.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T05A02.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T05A03.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T05A04.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T05A05.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T05A06.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T05A07.chr2.npy\n",
      "Loading Transcriptome data /scratch/sanjit/ENCODE_Imputation_Challenge/2_April_2020/Data/Gene_Expression/genome_wide_TPM_npy/T05_TPM.chr2.+.npy /scratch/sanjit/ENCODE_Imputation_Challenge/2_April_2020/Data/Gene_Expression/genome_wide_TPM_npy/T05_TPM.chr2.-.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T06A02.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T06A03.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T06A04.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T06A05.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T06A06.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T06A07.chr2.npy\n",
      "Loading Transcriptome data /scratch/sanjit/ENCODE_Imputation_Challenge/2_April_2020/Data/Gene_Expression/genome_wide_TPM_npy/T06_TPM.chr2.+.npy /scratch/sanjit/ENCODE_Imputation_Challenge/2_April_2020/Data/Gene_Expression/genome_wide_TPM_npy/T06_TPM.chr2.-.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T07A02.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T07A03.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T07A04.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T07A05.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T07A06.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T07A07.chr2.npy\n",
      "Loading Transcriptome data /scratch/sanjit/ENCODE_Imputation_Challenge/2_April_2020/Data/Gene_Expression/genome_wide_TPM_npy/T07_TPM.chr2.+.npy /scratch/sanjit/ENCODE_Imputation_Challenge/2_April_2020/Data/Gene_Expression/genome_wide_TPM_npy/T07_TPM.chr2.-.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T08A02.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T08A03.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T08A04.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T08A05.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T08A06.chr2.npy\n",
      "Loading Epigenome data ../Data/100bp_12_7_Data_20_July_2020/T08A07.chr2.npy\n"
     ]
    }
   ],
   "source": [
    "# Compute a true vs predicted spearman\n",
    "test_generator = TranscriptomeGenerator(window_size,\n",
    "                   256,\n",
    "                   shuffle=False,\n",
    "                   mode='testing'+'TSS_only',\n",
    "                   masking_probability=0.0)\n",
    "\n",
    "np.set_printoptions(precision=3, suppress=True)\n",
    "yTrue = []\n",
    "yPred = []\n",
    "for i in tqdm(range(100)):\n",
    "\n",
    "    X, Y = test_generator.__getitem__(i)\n",
    "    \n",
    "    yPred_value = trained_model.predict(X)\n",
    "#     print(X.shape, Y.shape, yPred_value.shape)\n",
    "    yTrue.extend(np.squeeze(Y))\n",
    "    yPred.extend(np.squeeze(yPred_value))\n",
    "\n",
    "pc, _ = pearsonr(yTrue, yPred)\n",
    "sc, _ = spearmanr(yTrue, yPred)\n",
    "# print(\"Pearson =\", round(pc, 3), \"Spearman =\", round(sc, 3))\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (20, 12)\n",
    "plt.plot(yTrue, yPred, 'o', markersize=4, color='green')\n",
    "plt.xlabel(\"True Normalized TPM\", size=40)\n",
    "plt.ylabel(\"Predicted Normalized TPM\", size=40)\n",
    "plt.xlim(-0.5, 4)\n",
    "plt.ylim(-0.5, 2)\n",
    "plt.title(\"Pearson = \"+str(round(pc, 3))+\" Spearman = \"+str(round(sc, 3)), size=50)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a609dec6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pre-compute the epigenetic features for a few exemplar genes\n",
    "STARTS = [136118149, 136116243, 99105113, 84905656, 33513875, 186694060]\n",
    "STRANDS = [\"-\", \"-\", \"+\", \"+\", \"+\", \"+\"]\n",
    "CHROMS = [\"chr2\", \"chr2\", \"chr9\", \"chr2\", \"chr2\", \"chr2\"]\n",
    "GENES = [\"CXCR4\", \"CXCR4_alternative\", \"TGFBR1\", \"High\", \"Low\", \"Medium\"]\n",
    "\n",
    "for gene_index in range(len(GENES)):\n",
    "    prediction_generator = TranscriptomePredictor(window_size,\n",
    "                           1,\n",
    "                           shuffle=False,\n",
    "                           mode='testing',\n",
    "                           masking_probability=0.0,\n",
    "                           cell_type=-1, # The last cell type in CELL_TYPES is T13\n",
    "                           chrom=CHROMS[gene_index], start=int(STARTS[gene_index]/RESOLUTION), strand=STRANDS[gene_index])\n",
    "                                              \n",
    "    for i in range(1):\n",
    "        X, Y = prediction_generator.__getitem__(i)\n",
    "        print(X.shape, Y.shape)\n",
    "        np.save(GENES[gene_index] + \".npy\", X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13d91cc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load Alan's H3K27ac p300 dataset\n",
    "df_CXCR4 = pd.read_csv(\"../Data/CXCR4.p300.tsv\", sep=\"\\t\")\n",
    "df_CXCR4_values = df_CXCR4[df_CXCR4['Position_wrt_TSS']!=\"Control\"]\n",
    "df_CXCR4_values[\"Position_wrt_TSS\"] = pd.to_numeric(df_CXCR4_values[\"Position_wrt_TSS\"], errors='coerce')/RESOLUTION\n",
    "\n",
    "df_TGFBR1 = pd.read_csv(\"../Data/TGFBR1.p300.tsv\", sep=\"\\t\")\n",
    "df_TGFBR1_values = df_TGFBR1[df_TGFBR1['Position_wrt_TSS']!=\"Control\"]\n",
    "df_TGFBR1_values[\"Position_wrt_TSS\"] = pd.to_numeric(df_TGFBR1_values[\"Position_wrt_TSS\"], errors='coerce')/RESOLUTION\n",
    "df_TGFBR1_values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b39411dc",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Define in-silico epigenesis\n",
    "def ise(gene, assay = 2, inserted_minuslog10p_value = 3, peak_width = 2):\n",
    "    \n",
    "    X = np.load(gene + \".npy\")\n",
    "\n",
    "    # Perform inference by introducing p-value changes with a peak width\n",
    "    yPred = []\n",
    "    center = window_size // 2\n",
    "    positions = range(center - center, center + center + 1)\n",
    "    for pos in positions:\n",
    "        X_modified = np.copy(X)\n",
    "        for p in range(pos - peak_width // 2, pos + peak_width // 2 + 1):\n",
    "            if( (p>=0) and (p < max(positions)) ):\n",
    "                X_modified[:, p, assay] = X_modified[:, p, assay] + inserted_minuslog10p_value\n",
    "                if(assay == 2): # Remove H3K27me3 peaks while adding H3K27ac peaks\n",
    "                    X_modified[:, p, 1] = 1.0 # max(1.0, X_modified[:, p, 1] - inserted_minuslog10p_value)\n",
    "                elif(assay == 1):\n",
    "                    X_modified[:, p, 2] = 1.0\n",
    "                elif(assay == 3):\n",
    "                    X_modified[:, p, 4] = 1.0\n",
    "                elif(assay == 4):\n",
    "                    X_modified[:, p, 3] = 1.0\n",
    "        yPred_value = trained_model.predict(X_modified)\n",
    "        yy = yPred_value[0][0]\n",
    "        yPred.append(yy)\n",
    "    \n",
    "    # Instead of scaling, divide by yPred\n",
    "    yPred_value = trained_model.predict(X)[0][0]\n",
    "    yPred = (np.power(10, yPred) -1) / (np.power(10, yPred_value) -1)\n",
    "\n",
    "    return yPred\n",
    "\n",
    "\n",
    "def p_value_mapping(inserted_minuslog10p_value):\n",
    "    minuslog10p_value = np.expm1(inserted_minuslog10p_value)\n",
    "    p_value = np.power(10, -1 * minuslog10p_value)\n",
    "    return round(minuslog10p_value, 2)\n",
    "\n",
    "\n",
    "assays = ['H3K36me3', 'H3K27me3', 'H3K27ac', 'H3K4me1', 'H3K4me3', 'H3K9me3']\n",
    "def assay_name(assay):\n",
    "    return assays[assay]\n",
    "\n",
    "\n",
    "#Perform in-silico epigenesis\n",
    "genes = [\"CXCR4\", \"TGFBR1\", \"High\", \"Medium\", \"Low\"]\n",
    "assay_color = ['red', 'green', 'blue', 'cyan', 'pink', 'brown']\n",
    "xticklabels = range(-(window_size // 2), (window_size // 2) + 1)\n",
    "for gene in genes:\n",
    "\n",
    "    gene_features = np.squeeze(np.load(gene+\".npy\"), axis=0)\n",
    "    plt.rcParams[\"figure.figsize\"] = (20, 10)\n",
    "    plt.figure(0)\n",
    "    ax = sns.heatmap(np.transpose(gene_features), annot=False, fmt=\"f\",\n",
    "                     vmin=0.0, vmax=2.0,\n",
    "                     yticklabels=assays, xticklabels=xticklabels)\n",
    "    plt.show()\n",
    "    plt.close()\n",
    "\n",
    "    for assay in range(6):\n",
    "\n",
    "        if(gene == 'CXCR4'):\n",
    "            df = df_CXCR4_values\n",
    "        elif(gene == 'TGFBR1'):\n",
    "            df = df_TGFBR1_values\n",
    "        else:\n",
    "            df = df_CXCR4_values.iloc[0:0,:].copy()\n",
    "\n",
    "        inserted_minuslog10p_value = 2.5\n",
    "        peak_width = 2\n",
    "        gene_ise = ise(gene, assay, inserted_minuslog10p_value, peak_width)\n",
    "\n",
    "        plt.figure(1)\n",
    "        plt.plot(xticklabels, 2.0 * gene_ise, 'o-', color=assay_color[assay], markersize=10, label=\"Model Predictions \" + assay_name(assay))\n",
    "#         plt.plot(xticklabels, 5.0 * np.transpose(gene_features[:, assay]), 'o-', color='darkorange', markersize=10, label=\"Epigenetic Features \" + assay_name(assay))\n",
    "    plt.plot(df['Position_wrt_TSS'], df['Fold_Change'], 'o', color='gray', label=\"Experimental Data from qPCR\")\n",
    "    plt.xlim(-22, 22)\n",
    "    plt.ylim(-3, 10)\n",
    "    ax=plt.gca() \n",
    "    x_vals = ax.axes.get_xticks()\n",
    "    ax.set_xticklabels(['{:3.0f}'.format(x * 100) for x in x_vals])\n",
    "    plt.tick_params(axis='both', which='major', labelsize=30)\n",
    "    plt.tick_params(axis='both', which='minor', labelsize=30)\n",
    "    plt.xlabel(\"Peak Position (in bp) w.r.t TSS\", size=50)\n",
    "    plt.ylabel(\"Gene expression fold change\", size=40)\n",
    "    plt.legend(loc=1, prop={'size': 18}, ncol=3)\n",
    "    plt.title(\"Gene \"+gene+\": inserting \" + str(peak_width * RESOLUTION) + \"bp peaks of -log10(p_value)=\"+str(p_value_mapping(inserted_minuslog10p_value)), size=35)\n",
    "    plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3fa033e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
