{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d12a5cce",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from sklearn import linear_model\n",
    "from sklearn.linear_model import LogisticRegression as logreg\n",
    "from sklearn.neural_network import MLPRegressor\n",
    "from sklearn.preprocessing import PolynomialFeatures\n",
    "from sklearn.metrics import r2_score, mean_squared_error\n",
    "from scipy.stats import spearmanr, pearsonr\n",
    "import seaborn as sns\n",
    "from sklearn.decomposition import PCA\n",
    "# import umap\n",
    "# import hdbscan\n",
    "import sklearn.cluster as cluster\n",
    "from sklearn.metrics import adjusted_rand_score, adjusted_mutual_info_score\n",
    "import matplotlib.pyplot as plt\n",
    "# %matplotlib notebook\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "# import shap\n",
    "import random\n",
    "import itertools\n",
    "import warnings\n",
    "from tqdm import tqdm\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "assays = ['H3K36me3', 'H3K27me3', 'H3K27ac', 'H3K4me1', 'H3K4me3', 'H3K9me3']\n",
    "assay_index_choice = 5 # 1-based\n",
    "\n",
    "RESOLUTION = 25\n",
    "window_size = 401\n",
    "operative_half_window_size = 80 # will lead to a total of the central 2*operative_half_window_size + 1 features being chosen\n",
    "pairwise_flag = True # If true then H3K27ac features will be concatentated with assays[assay_index_choice-1]\n",
    "if(pairwise_flag):\n",
    "    pairwise_multiplier = 0\n",
    "else:\n",
    "    pairwise_multiplier = 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d09d0fd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames = [\"pos_\" + str(x) for x in list(range(-(window_size//2), (window_size//2) + 1))]\n",
    "colnames = colnames + [\"log10p1(TPM)\", \"cell_type\", \"chrom\", \"position\", \"strand\", \"assay_index\"] \n",
    "df = pd.read_csv('../../../Data/Training_Data.T13.sorted.csv', sep=\",\", names=colnames, low_memory=False)\n",
    "df_sorted = df.sort_values(['chrom', 'position', 'cell_type', 'assay_index'], ascending=[1, 1, 1, 1])\n",
    "df_sorted_unique = df_sorted.drop_duplicates()\n",
    "df_H3K27ac = df_sorted_unique.iloc[range(3-1, len(df_sorted_unique), len(assays)), :]\n",
    "df_other_assay = df_sorted_unique.iloc[range(assay_index_choice-1, len(df_sorted_unique), len(assays)), :]\n",
    "df_merged = pd.merge(df_H3K27ac, df_other_assay, how='inner', on=['cell_type', 'chrom', 'position', 'strand'], suffixes=['_H3K27ac', '_'+assays[assay_index_choice-1]])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "e5ee9829",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos_-80_H3K27ac</th>\n",
       "      <th>pos_-79_H3K27ac</th>\n",
       "      <th>pos_-78_H3K27ac</th>\n",
       "      <th>pos_-77_H3K27ac</th>\n",
       "      <th>pos_-76_H3K27ac</th>\n",
       "      <th>pos_-75_H3K27ac</th>\n",
       "      <th>pos_-74_H3K27ac</th>\n",
       "      <th>pos_-73_H3K27ac</th>\n",
       "      <th>pos_-72_H3K27ac</th>\n",
       "      <th>pos_-71_H3K27ac</th>\n",
       "      <th>...</th>\n",
       "      <th>pos_75_H3K4me3</th>\n",
       "      <th>pos_76_H3K4me3</th>\n",
       "      <th>pos_77_H3K4me3</th>\n",
       "      <th>pos_78_H3K4me3</th>\n",
       "      <th>pos_79_H3K4me3</th>\n",
       "      <th>pos_80_H3K4me3</th>\n",
       "      <th>log10p1(TPM)_H3K27ac</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.316077</td>\n",
       "      <td>0.316077</td>\n",
       "      <td>0.316077</td>\n",
       "      <td>0.316077</td>\n",
       "      <td>0.316077</td>\n",
       "      <td>0.316077</td>\n",
       "      <td>0.316077</td>\n",
       "      <td>0.316077</td>\n",
       "      <td>0.316077</td>\n",
       "      <td>0.316077</td>\n",
       "      <td>...</td>\n",
       "      <td>0.334390</td>\n",
       "      <td>0.334390</td>\n",
       "      <td>0.334390</td>\n",
       "      <td>0.334390</td>\n",
       "      <td>0.334390</td>\n",
       "      <td>0.334390</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>T13</td>\n",
       "      <td>chr1</td>\n",
       "      <td>27466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.314752</td>\n",
       "      <td>0.314752</td>\n",
       "      <td>0.326776</td>\n",
       "      <td>0.329428</td>\n",
       "      <td>0.347780</td>\n",
       "      <td>0.347780</td>\n",
       "      <td>0.347780</td>\n",
       "      <td>0.347780</td>\n",
       "      <td>0.347780</td>\n",
       "      <td>0.342905</td>\n",
       "      <td>...</td>\n",
       "      <td>0.318344</td>\n",
       "      <td>0.318344</td>\n",
       "      <td>0.319475</td>\n",
       "      <td>0.327719</td>\n",
       "      <td>0.327719</td>\n",
       "      <td>0.327719</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>T13</td>\n",
       "      <td>chr1</td>\n",
       "      <td>46955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.830275</td>\n",
       "      <td>0.882636</td>\n",
       "      <td>0.907202</td>\n",
       "      <td>1.049832</td>\n",
       "      <td>1.049832</td>\n",
       "      <td>1.049832</td>\n",
       "      <td>0.992307</td>\n",
       "      <td>0.857431</td>\n",
       "      <td>0.818399</td>\n",
       "      <td>0.554225</td>\n",
       "      <td>...</td>\n",
       "      <td>0.318344</td>\n",
       "      <td>0.318344</td>\n",
       "      <td>0.318344</td>\n",
       "      <td>0.318344</td>\n",
       "      <td>0.321409</td>\n",
       "      <td>0.330797</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>T13</td>\n",
       "      <td>chr1</td>\n",
       "      <td>48262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.580898</td>\n",
       "      <td>0.830275</td>\n",
       "      <td>0.882636</td>\n",
       "      <td>0.907202</td>\n",
       "      <td>1.049832</td>\n",
       "      <td>1.049832</td>\n",
       "      <td>1.049832</td>\n",
       "      <td>0.992307</td>\n",
       "      <td>0.857431</td>\n",
       "      <td>0.818399</td>\n",
       "      <td>...</td>\n",
       "      <td>0.328009</td>\n",
       "      <td>0.318344</td>\n",
       "      <td>0.318344</td>\n",
       "      <td>0.318344</td>\n",
       "      <td>0.318344</td>\n",
       "      <td>0.321409</td>\n",
       "      <td>0.030907</td>\n",
       "      <td>T13</td>\n",
       "      <td>chr1</td>\n",
       "      <td>48263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.336684</td>\n",
       "      <td>0.328435</td>\n",
       "      <td>0.307145</td>\n",
       "      <td>0.313420</td>\n",
       "      <td>0.326993</td>\n",
       "      <td>0.347079</td>\n",
       "      <td>0.399997</td>\n",
       "      <td>0.462830</td>\n",
       "      <td>0.405791</td>\n",
       "      <td>0.405791</td>\n",
       "      <td>...</td>\n",
       "      <td>0.318992</td>\n",
       "      <td>0.333754</td>\n",
       "      <td>0.324479</td>\n",
       "      <td>0.362663</td>\n",
       "      <td>0.372292</td>\n",
       "      <td>0.372292</td>\n",
       "      <td>0.015702</td>\n",
       "      <td>T13</td>\n",
       "      <td>chr1</td>\n",
       "      <td>48566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10616</th>\n",
       "      <td>0.314752</td>\n",
       "      <td>0.314752</td>\n",
       "      <td>0.296073</td>\n",
       "      <td>0.298432</td>\n",
       "      <td>0.302057</td>\n",
       "      <td>0.314752</td>\n",
       "      <td>0.314752</td>\n",
       "      <td>0.311762</td>\n",
       "      <td>0.461223</td>\n",
       "      <td>0.490912</td>\n",
       "      <td>...</td>\n",
       "      <td>0.354387</td>\n",
       "      <td>0.329730</td>\n",
       "      <td>0.323239</td>\n",
       "      <td>0.323239</td>\n",
       "      <td>0.369133</td>\n",
       "      <td>0.401791</td>\n",
       "      <td>0.019116</td>\n",
       "      <td>T13</td>\n",
       "      <td>chr9</td>\n",
       "      <td>5448205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10617</th>\n",
       "      <td>0.293496</td>\n",
       "      <td>0.293496</td>\n",
       "      <td>0.301077</td>\n",
       "      <td>0.301733</td>\n",
       "      <td>0.293827</td>\n",
       "      <td>0.299434</td>\n",
       "      <td>0.300517</td>\n",
       "      <td>0.293496</td>\n",
       "      <td>0.295203</td>\n",
       "      <td>0.295781</td>\n",
       "      <td>...</td>\n",
       "      <td>0.343971</td>\n",
       "      <td>0.410588</td>\n",
       "      <td>0.531649</td>\n",
       "      <td>0.468521</td>\n",
       "      <td>0.474043</td>\n",
       "      <td>0.487957</td>\n",
       "      <td>1.399599</td>\n",
       "      <td>T13</td>\n",
       "      <td>chr9</td>\n",
       "      <td>5449832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10618</th>\n",
       "      <td>0.440758</td>\n",
       "      <td>0.405791</td>\n",
       "      <td>0.405791</td>\n",
       "      <td>0.405791</td>\n",
       "      <td>0.404085</td>\n",
       "      <td>0.424893</td>\n",
       "      <td>0.419179</td>\n",
       "      <td>0.349506</td>\n",
       "      <td>0.349657</td>\n",
       "      <td>0.342884</td>\n",
       "      <td>...</td>\n",
       "      <td>0.359020</td>\n",
       "      <td>0.331980</td>\n",
       "      <td>0.326088</td>\n",
       "      <td>0.314585</td>\n",
       "      <td>0.313231</td>\n",
       "      <td>0.315433</td>\n",
       "      <td>1.389310</td>\n",
       "      <td>T13</td>\n",
       "      <td>chr9</td>\n",
       "      <td>5475460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10619</th>\n",
       "      <td>0.944373</td>\n",
       "      <td>0.817767</td>\n",
       "      <td>0.775156</td>\n",
       "      <td>0.680218</td>\n",
       "      <td>0.558381</td>\n",
       "      <td>0.426065</td>\n",
       "      <td>0.322553</td>\n",
       "      <td>0.324419</td>\n",
       "      <td>0.304418</td>\n",
       "      <td>0.295110</td>\n",
       "      <td>...</td>\n",
       "      <td>0.644951</td>\n",
       "      <td>0.644951</td>\n",
       "      <td>0.758850</td>\n",
       "      <td>0.758631</td>\n",
       "      <td>0.673328</td>\n",
       "      <td>0.548530</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>T13</td>\n",
       "      <td>chr9</td>\n",
       "      <td>5504759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10620</th>\n",
       "      <td>0.817767</td>\n",
       "      <td>0.775156</td>\n",
       "      <td>0.680218</td>\n",
       "      <td>0.558381</td>\n",
       "      <td>0.426065</td>\n",
       "      <td>0.322553</td>\n",
       "      <td>0.324419</td>\n",
       "      <td>0.304418</td>\n",
       "      <td>0.295110</td>\n",
       "      <td>0.294545</td>\n",
       "      <td>...</td>\n",
       "      <td>0.644951</td>\n",
       "      <td>0.758850</td>\n",
       "      <td>0.758631</td>\n",
       "      <td>0.673328</td>\n",
       "      <td>0.548530</td>\n",
       "      <td>0.635814</td>\n",
       "      <td>1.084823</td>\n",
       "      <td>T13</td>\n",
       "      <td>chr9</td>\n",
       "      <td>5504760</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10621 rows Ã— 326 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       pos_-80_H3K27ac  pos_-79_H3K27ac  pos_-78_H3K27ac  pos_-77_H3K27ac  \\\n",
       "0             0.316077         0.316077         0.316077         0.316077   \n",
       "1             0.314752         0.314752         0.326776         0.329428   \n",
       "2             0.830275         0.882636         0.907202         1.049832   \n",
       "3             0.580898         0.830275         0.882636         0.907202   \n",
       "4             0.336684         0.328435         0.307145         0.313420   \n",
       "...                ...              ...              ...              ...   \n",
       "10616         0.314752         0.314752         0.296073         0.298432   \n",
       "10617         0.293496         0.293496         0.301077         0.301733   \n",
       "10618         0.440758         0.405791         0.405791         0.405791   \n",
       "10619         0.944373         0.817767         0.775156         0.680218   \n",
       "10620         0.817767         0.775156         0.680218         0.558381   \n",
       "\n",
       "       pos_-76_H3K27ac  pos_-75_H3K27ac  pos_-74_H3K27ac  pos_-73_H3K27ac  \\\n",
       "0             0.316077         0.316077         0.316077         0.316077   \n",
       "1             0.347780         0.347780         0.347780         0.347780   \n",
       "2             1.049832         1.049832         0.992307         0.857431   \n",
       "3             1.049832         1.049832         1.049832         0.992307   \n",
       "4             0.326993         0.347079         0.399997         0.462830   \n",
       "...                ...              ...              ...              ...   \n",
       "10616         0.302057         0.314752         0.314752         0.311762   \n",
       "10617         0.293827         0.299434         0.300517         0.293496   \n",
       "10618         0.404085         0.424893         0.419179         0.349506   \n",
       "10619         0.558381         0.426065         0.322553         0.324419   \n",
       "10620         0.426065         0.322553         0.324419         0.304418   \n",
       "\n",
       "       pos_-72_H3K27ac  pos_-71_H3K27ac  ...  pos_75_H3K4me3  pos_76_H3K4me3  \\\n",
       "0             0.316077         0.316077  ...        0.334390        0.334390   \n",
       "1             0.347780         0.342905  ...        0.318344        0.318344   \n",
       "2             0.818399         0.554225  ...        0.318344        0.318344   \n",
       "3             0.857431         0.818399  ...        0.328009        0.318344   \n",
       "4             0.405791         0.405791  ...        0.318992        0.333754   \n",
       "...                ...              ...  ...             ...             ...   \n",
       "10616         0.461223         0.490912  ...        0.354387        0.329730   \n",
       "10617         0.295203         0.295781  ...        0.343971        0.410588   \n",
       "10618         0.349657         0.342884  ...        0.359020        0.331980   \n",
       "10619         0.304418         0.295110  ...        0.644951        0.644951   \n",
       "10620         0.295110         0.294545  ...        0.644951        0.758850   \n",
       "\n",
       "       pos_77_H3K4me3  pos_78_H3K4me3  pos_79_H3K4me3  pos_80_H3K4me3  \\\n",
       "0            0.334390        0.334390        0.334390        0.334390   \n",
       "1            0.319475        0.327719        0.327719        0.327719   \n",
       "2            0.318344        0.318344        0.321409        0.330797   \n",
       "3            0.318344        0.318344        0.318344        0.321409   \n",
       "4            0.324479        0.362663        0.372292        0.372292   \n",
       "...               ...             ...             ...             ...   \n",
       "10616        0.323239        0.323239        0.369133        0.401791   \n",
       "10617        0.531649        0.468521        0.474043        0.487957   \n",
       "10618        0.326088        0.314585        0.313231        0.315433   \n",
       "10619        0.758850        0.758631        0.673328        0.548530   \n",
       "10620        0.758631        0.673328        0.548530        0.635814   \n",
       "\n",
       "       log10p1(TPM)_H3K27ac  cell_type  chrom  position  \n",
       "0                  0.000000        T13   chr1     27466  \n",
       "1                  0.000000        T13   chr1     46955  \n",
       "2                  0.000000        T13   chr1     48262  \n",
       "3                  0.030907        T13   chr1     48263  \n",
       "4                  0.015702        T13   chr1     48566  \n",
       "...                     ...        ...    ...       ...  \n",
       "10616              0.019116        T13   chr9   5448205  \n",
       "10617              1.399599        T13   chr9   5449832  \n",
       "10618              1.389310        T13   chr9   5475460  \n",
       "10619              0.000000        T13   chr9   5504759  \n",
       "10620              1.084823        T13   chr9   5504760  \n",
       "\n",
       "[10621 rows x 326 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_subsampled"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "fb5bf796",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9474, 322) (4877, 322) (9474,) (4877,)\n",
      "Train: MSE =  0.176 Pearson = 0.677 Spearman = 0.718\n",
      "Test: MSE =  0.187 Pearson = 0.65 Spearman = 0.695\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[<matplotlib.lines.Line2D at 0x2aab15f3ed90>]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYkAAAD4CAYAAAAZ1BptAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8/fFQqAAAACXBIWXMAAAsTAAALEwEAmpwYAAB4F0lEQVR4nO29ebwlVXUv/l1Vdc65Q88TNHQziggqoHTQOCugSAY0GoP5RYnPhGeUl5fpvWA0gzExxERNfDEaNEYyqDGOGHEAjBpFhAaZx+6Woemm5+GOZ6javz+q1q61d+06wz3n9j2X3t/P537uOXV2Ve3atfde81qklIKHh4eHh4cLwUJ3wMPDw8NjeOGJhIeHh4dHKTyR8PDw8PAohScSHh4eHh6l8ETCw8PDw6MU0UJ3YJBYs2aNOumkkxa6Gx4eHh6LCrfddttepdRa129PKSJx0kknYfPmzQvdDQ8PD49FBSJ6tOw3r27y8PDw8CiFJxIeHh4eHqXwRMLDw8PDoxSeSHh4eHh4lMITCQ8PDw+PUngi4eHh4eFRCk8kPDw8PDxK4YmEh4fHUQulFL5w23bMNOKF7srQwhMJDw+PoxaP7pvG7/7Hnbj+/l0L3ZWhhScSHh4eRy0acQIAmG16SaIMnkh4eHgctYiTtDJnK/YVOsvgiYSHh8dRCyYSzUyi8CjCEwkPD4+jFp5IdMZAiAQRXUREDxLRFiK60vE7EdGHs9/vIqLnZsdPJ6I7xN9hIvqt7Lc/IaInxG8XD6KvHh4eHoyWJhJe3VSGvlOFE1EI4CMALgSwHcCtRHStUuo+0ezVAE7L/p4H4KMAnqeUehDAOeI6TwD4kjjvQ0qpv+63jx4eHh4uJMpLEp0wCEniPABblFLblFINAJ8FcInV5hIA/6xS3AxgBRGtt9qcD2CrUqo0r7mHh8fRgX2Tdew6PDvv98kN155IlGEQROJ4AI+L79uzY722uRTAZ6xjV2TqqU8S0coB9NXDw2MR4Nw/uwHPe9+N834fJhINr24qxSCIBDmO2SPetg0RVQH8PID/EL9/FMCpSNVROwF8wHlzosuJaDMRbd6zZ08P3fbw8Dja4SWJzhgEkdgOYKP4vgHAjh7bvBrA7UopHfaolNqllIqVUgmAjyNVaxWglLpaKbVJKbVp7VpniVYPDw8PJ2Jvk+iIQRCJWwGcRkQnZxLBpQCutdpcC+DNmZfT8wEcUkrtFL+/EZaqybJZvBbAPQPoq8eA0Wgl2DtZX+hueHjMCXGmZmomXt1Uhr6JhFKqBeAKAN8EcD+Azyml7iWitxHR27Jm1wHYBmALUqng7Xw+EY0h9Yz6onXp9xPR3UR0F4CXA/jtfvvqMXj8682P4pUf+t7Ar/ur/3QLXvmh7w78uh4eElqSaHlJogx9u8ACgFLqOqSEQB77mPisALyj5NxpAKsdx980iL55zC/2TdWxf6oBpRSIXKanueE7D3r7ksf8I2GbhJckSuEjrj36Aq8tv8Y8FiNa2rvJSxJl8ETCoy9wMBL/9/BYTEi8uqkjPJHw6AtMG2IvSgwlDk03oTwBLwVnf/XqpnJ4IuHRF1in6/eh4cOW3RM4+0+/hX+/9fHOjY9SeBfYzvBEwqMvMAMWeyoxdHjwyUkAwHcfWrxOAPMtoSY+C2xHeCLh0Re8TWJ40UrSjS8MBud1dqQx35u3zwLbGZ5IePQF1ncrz4gNHZhwR4uYSMy3rYDHyKflKIcnEh59waubhhdslA2DxbvM59vryCf464zFO3s8hgJe3TS84A1wMUsSzeTIEAkvSZTDEwmPvqCJhHchHDqwqiZYxESiNc8cvl2+dMvuCbzx6psx04jn9b6LCZ5IePQFZvQ8jRg+PBUkiXknEso0XN/5+CH8cNs+7Dg0M6/3XUzwRMKjL7Ak4W0SwwcmEovZu2m+02XoLLDZfXge++DQHJ5IePQFnbvJL6qhw1NCkphvm4QVTKcT/nlDtoYnEh59QXnD9dCCufAwXMRE4gjZJPg+sZ/PBXgi4dEXcu+mzm0Pzzbx/PfdiM2P7J/TvTb92Q34h+9undO5RyMamftoOMAU7kca8x1MF1tZYH3q8CI8kfDoCzpOootFtfvwLJ48PIuteybndK+9k3X8xdcfmNO5RyPqGZFYzPr1+Y6EZsmBiUKs/3uXWIYnEh4948lDs/iDL92NZpxoSaKbTKOzzXThdVr4PmvpYFBvpW6ci5krnu/4BTZcx4lCkijw1PQ2iRwDIRJEdBERPUhEW4joSsfvREQfzn6/i4ieK357JCtTegcRbRbHVxHR9UT0cPZ/5SD66tE/rvziXfj0jx7D97fszVOFd7Gx602rw8JfzJvaMIGJ8qKWJOa573LeNpNEq5tipTDTiPHovql5vf9iQN9EgohCAB8B8GoAZwJ4IxGdaTV7NYDTsr/LAXzU+v3lSqlzlFKbxLErAdyolDoNwI3Z96c8kkThwzc+jL2T9YXuSilYTxwFJILpOp9X71KSWMyb2jAhlyQWr+pkviUJ6ZXXjJXhAvvWa27FS//qO/N6/8WAQUgS5wHYopTappRqAPgsgEusNpcA+GeV4mYAK4hofYfrXgLgmuzzNQBeM4C+Dj0ePzCND17/EK6/b9dCd6UUvMlHQaA39G68QWazTatTqgWXsXIxq6C+cc9OvOFjP8RNW/biE/+97Yjd96lhkzgyWWCBNE9ULAzXN23dB2Bxz71BYBBE4ngAsqrJ9uxYt20UgG8R0W1EdLloc4xSaicAZP/XuW5ORJcT0WYi2rxnz+LJm//ovik8vn+6cJw9UqaHOC0Ac3eVkESN6+5tEp30va5N7Ujvc3sm6vjaXTsHcq23/evtuOWR/fjyHU/g779z5LyzupXchhnz3fekRN0kJYxFTGMHgkEQCZd/nT2s7dq8UCn1XKQqqXcQ0Ut6ublS6mql1Cal1Ka1a9f2cuqC4l1fugd/9JV7CseZ+5tptI50l7oGc19RGIg4ic7n9WOT6MdvPUlUz9z0mz95C97x6dsxWS++h6/euQM3bd3bcz9ascJs88gRfx7vxSxJzHswXYm6Sc7BxTx+g8AgiMR2ABvF9w0AdnTbRinF/3cD+BJS9RUA7GKVVPZ/9wD6OjSYqLcwVS9uGOyvPTXEkkSubsptEt0sJO3d1KGtS9Loh0ic/8Hv4rw/v6Gncx7eNQEg936R+PCND+NTP3ik5340E4WZZnzE1BfMcCxmR4D5liTk2LRiYbg2JIn++3DfjsP45x8+0vd1FgKDIBK3AjiNiE4moiqASwFca7W5FsCbMy+n5wM4pJTaSUTjRLQUAIhoHMArAdwjzrks+3wZgK8MoK9Dg1acOHXzDS1JDC+RYElAqVyC6Gbjqze7lSTy33VRoz7W6U/2TmHfVKOnc3TFMsc7ihM1J115K06gVL55zzdym0Tv9/vVf7oFnxtwbezZZozLPnkLHsoIcDeYb5uEabhOnJLEIIjEFZ++HX/0lXvx5KHZvq91pNE3kVBKtQBcAeCbAO4H8Dml1L1E9DYielvW7DoA2wBsAfBxAG/Pjh8D4PtEdCeAWwB8TSn1jey3qwBcSEQPA7gw+/6UQStWTo45t0kMv7opUao3SaLVu3eTdrFdIG7Y9Y5aiZoTd87PfaRUTjlR7r2vmx85gHt3HGrb5p4nDuH/fv7OrvN2/fixg/juQ3ucatYyzH8W2PxzM1ZgmpQMWN107PIRAMC3H1h8CpFoEBdRSl2HlBDIYx8TnxWAdzjO2wbg7JJr7gNw/iD6N4xoJgmCpGiqYSIxzOom5vQTpfQm3pVNQhtSO3k3mVxcABoIN9doJahGvfFFrr7GidLvqRfwuM00Y6zo+Ww3rrnpESRK4S0vPLnwW6MPdVOcqI6xL//jU7di90Qdv3Ph6XoTbAfORtuLYDP/aTny68vgUEOSGEAXTlk7jpu27sON9+/CLz/vhP4veAThI64XCHGinGoXXhTDrW7KjdW9VKabbXXH2Zr6YPN/Pzgw3ZvKCSgnEnPZvObj3f7xtffiPV+9z/lbPzaJOMm56jIwwe12LMJst+klrfx821MKhmtHWo5BpMHnIbrlkf092aSUUtg2xzQ2g4InEguEVuxWWbDhepjVTczpK6Fu6oZIaEmiA2smbRK9pP3ohLkEKLreUStRczKo5uqmI2WTYO+mzvf7tx89ik9+/yf6e6xURzVSNdv1u635EGSJBntR3xypGtdAZpNwJPgbhLqJx3JitoUD082uz/v+lr04/4PfdbrLHyl4IjGPUErhUMmEaMaJ072vvgjiJHjTiRPVU4K/biWJlqVuSv/Ppacm9k72Lkm41EpxkszZcA2k6qYjgXqXcSkA8J937sS1d+ZOid2omyphr5JEpm7qcF3DmJx9Pjzb1ERvkLCJhKsc7yAYFEl0Hukh1ce+yQaUAg7NdE9YBg1PJOYRH/mvLTj7T7+F3RNFj4ZW0slwvTBE4vR3fx2/8okftW0j1U29eB+xwbaT73vLqW7qf6Hu61KSkJtCuSSR4Jaf7MfOHspc8rWOmOG6h4jrVpJz0a6AMhcqUbrp17uUjLqVJOS7ZsJ61p98C7/88fbzci6IFcCZ1EsliYGom/Ixemxf91IBE8b5ts20gycS84iv3plG7O6dKHKwrThxqiwW2gW23krw/S3tA8VYXaSU1OF2oW7Knq3R6sUm0d2G1Q5j1RBAypV1A8npu+xGSaZuesM//BCv/ND3uu4Hv+8j8W6TRGk1UDdJ8hpC/anzF3UpSXRL9Hgz7kwk8s9ys77t0QNd3acXxEmi1WZxAqe3Xi/qpm17JnHSlV/Dlt2mm2+sgONXjIKoN0miH+eDQcETiXkEU/+KozJYK1FOXTGfMzXENgnTcI3sc/rhx48dwM3b9jnPq3crSYiNWbFLYh9rZKSSEolubRKHZ/KxdxHylnA6mJjt/j0NWt3UbvOSdoJubBKtODHUiJ2uD+Q2iW6fR1lzpQzy97l4kZXhlp/sx4eufwgPPHlYH4sTpcu7SqZnrt5NX74jVdkxg5jfJ8FoNcRxy0fxaIkkMd1o4aQrv4ZP/+gxfYwZq/m2zbSDJxLziHYG2s5xEkdekuiWW5dxEnb50g/d8DCuKikMpL1tOtkkXJJEHyI/E95ubBL1VowbH8iTK5a6wM7FJpE916CIRLvrSBVQNzaJ1EZmEodOY87eTd0a4ruNeTHUTQNMy/HuL9+Nv7Wi5ZMkTS+T3jf3QjIkiR7mHjNCzJjoa2TE6MTVY/jJXrckwQbtD9/4sD7WizQ4X/BEYh7Bi9O1obSSkojrmFUyycDSJE83Wnj/Nx7oqBboVXpJg+n4c/p/thGXRhTz/TvpV53qpj6IBL+Hblxg/+WHj+JdX8qDvexNSqlULTMXIs7PXR8QkZh25JViSCNvVzYJ6f7ZZYAkq5u6NSh364RgqJsGGEw3o+efSYRySd8dHNqLuonn/kjF3FrjRCEgwuoltVIjdCWTaOR4anWTt0k8NcGbgq2ySDKvINfkk+L19IA2k82PHMDff2crbn+svU63F9UJwGk5zEVVj8s9f2y//Wac4L8fLmbulefnaT966poB3ui7UV3Yut/Cu8u+zoVI8IY3KEmiXcBls8dNrpkkuRpRSxTtz+HNtVubRLfEx45dGBRyNamUEtKU9+lxONVNvXg38QZfixySREgIqfz5eXykZFbvMkvBfMITiXkEv1ibC2AJohmrwgSUXLg0cN60ZS8u+bvvz8nLgSfltCOhoESvRMKUJDIi0YxLuZ5ZK3fTB771EN70j7fgtkf3O/sLoKDO6hVK5TEN3aiIRqyIbHu8+/GZ1xHXjcFJiAxbVWi6kXa+X7NVdELorG5KN8Lu1U3dEQm5Juzxv/Pxg5iYnZs7KM8DQ1LNNu/0vvlGLRmKXtRNPBY1ax61MkkiCKicSDCj5ZIkFrBwlCcS8wjeDO3NSYrQ9oSRi0Jyq/fsOIQ7tx/CZI8bubxHJ3XSZL23xSddYBOxuMq4Hl5Ajez3+3amBsTD1jO5XGDnujnbfvCdYKvKbHVHP0Si2YMksfPQTEcOVs4PmxBwP6OAnJlsbbSSpOjd1FHd1JskkUuFnWwSZr8kLvnID/CWf7q1q/t958HdRoAgv3/5XK1EabVZIgIIZar+Xt45j0XNoW6KAkJIVPr8OXHOjzW6tOPNJzyRmEfwpmBvmi1rkkpIDmZK6JztBdwL+FxXanIJe7O+5qZHCq58EjLBH6/leisp5djtehLayGeJ5q6I67nuzWYq6M4XYSLxxbe/AECRsPTK0RkxF9m1Om2q9+04jJ/+i2/jX4WXiwtyftiqNJ4n1Sjoyn0yTUkxN++mbiWJpMs5bKubbClpc5eusL/6T7fiT/8zT1nSchCJRHg3JcK7SRLgXl45v1t76OJEIQwIAVHp8zszMGh1k5cknpJg7s5Wv8jvBSIhfjP99U19cS/gjbZTqg+pbpptxvjja+/Ff9y2vbS9cqmb2hjcdWW67CTOCluNTBdhV8T1XKNe5eLqZqE1WgnCgHDc8tHsnP4kCaM8puZS2xOJHQfTAL1v39++hK0hSVj95PGqhEFXfXZ5N3X7rLNdG67R1XVtdVO/wWx8Pa1usryn2Lsp7WORSPTk3ZTNaXudaiIRUKmtx7W2tXeTlySGD7c9egC/+Zkf9xXExXOryI0WuUuGYbhuxIV2csLe+sh+/NcDu/G+6+7Hh65/yNmHu7cf0vd3VVmTkKqs3YfTmILDbdIByOCjnEjEpRPajh5lScJeNK5U4XOWJERfurFJ1FsxalGg9dS25GBvcOSquSgg36eWJDpsqmO1VLLqJPm1lSSyr91KEtK7iR+50+bI77xbdZPLJjHlmJO2d1O/uZOmG2mhJ2ba7NKkrDYzJYnO6qZNf3YDPvitB41js3pOW0RCpUQiDMptPa73pKVvb5MYPvz6P2/GtXfumFPmUBv2ptlsJ0lknCxgujg2Hdzd3317C/76Ww/i6u9tw98K32rG4/un8XN/933ceH+aw76TR440CHIqERlYZkPqcLlbqU0iwaGZJv774T06FYZSqlDjul5ilGsai9gkQnytbiF19d2qm6pRgErAeYnaSxLcrgxy8+ZTO0kS3OWJDkRdvs8ikcjUTWHQ0X1SqTTuw1ZpdmKQ+LLdqptsW8cXb9+OZ/7xN3UVQLsdkBL2flOyHJxpIk7ytPZ28r48mC6vLyHH1jXflFLYO1nHh7+9xTiuU6Go4rwJM5tEJ8O1RMN7Nw0vWN86iKC2dh4y9sbViBOMZ2kkJOfLC13up4dnm21VKIezTZ8JnYtrk2BJIwoIuyfSzd326ZYbjvRuShehQj0jEr92za140z/egg9kEk7DofYpK4oTG9HCRSLB9+yGg5WLqyvDdTNBLQp0XqJ2UiCQJ60rvZ7D7baT4Zq5x07vSzoi2FISj1et0lmSsNVLdrxEGaRHWzfIVZPp/+89lLo/3/2EWdyoUC2uT0niwFSjtNJc6pqau8Bqw7V4Jtf9O8UCFSSJjBgRlddGaXefRR8nQUQXEdGDRLSFiK50/E5E9OHs97uI6LnZ8Y1E9F9EdD8R3UtE/1uc8ydE9AQR3ZH9XTyIvnYL9k7opKLpBvZG027jarQSjNciR7viwp2cbRmT395UbI6906bDNolWorDrcCpJ2ERCbkYyToK5USBdbExk+JqS22xqtYtbknAm+BNNmnGCn+ydwjP+8Bv44u3lNhMgX1y1KOjOJhEnqEWh9p23F6e9kKMORMIVm9GJuPH76jT3pEtzO0mi0yYrbRFG+vcuiUv3NgmT4HNUsi2JyD200UqchuNu8l8tydbRoZmmlaakTJLI1U1yrbiI5eESN9yypIocTBcGVDqu7QzXizp3ExGFAD4C4NUAzgTwRiI602r2agCnZX+XA/hodrwF4HeVUmcAeD6Ad1jnfkgpdU72Z1S+m2+wn3OvsQOMdgFBclO0J1MjlkQiKZwj20/Mtgwu/ImDZkZSPr/bandy4rPx1F4MzZbcwHMR3q7WxouMN9mDmTSzYqxSyIZqSxKuQCab++P8+p9vY1gH8rEfq4Zdiexsk2A9dTvPNAAIHXm5JBpxccy7lSQ6uTtLSaIgrbIkkdkk2qno7A205WBIXODfZxoxmnHS0THCtkkwkbDHQ9633nIbrrvJw7V0JF1HB6ebpW7nscpdYGWcxEwH76YyNSwTPFtaaCVZMF1Q7t3UTpKQ73fHwZnS/GjzgUFIEucB2KKU2qaUagD4LIBLrDaXAPhnleJmACuIaL1SaqdS6nYAUEpNIK2RffwA+tQ3eAKXBe78cOs+PFKSgwUwJ35BZWGlBZBotBLNARlEwhEtOjHbNCaWXZikYRGJTotYbkrbD6REwpYk6mLTS4Qk8cUfP4H/+/m79G9sdOVn4OudtHrcYZMo1/vb3lNAOhbLRysAgP1T7W1GPL5j1ah7dVMlABEhCqij4bqTJOFUN3Ug1nUdT9K+v/I6BXUTSxJRrkopgzkfZWqKtrfXm/5sM8Ev/cMPceYffbNtex5Kvv5olSUJczz494DSuevaPCWRuHv7IZz9nmJKfiYSB6YbVhS/STCk4ZrHTWY7cG3qvC/Yjgv1EnVTkiiEQYCAqNSlVp7Da9YVJ/EHX7obl159M+54/KD7QgPGIIjE8QAeF9+3o7jRd2xDRCcBeA4AmTT+ikw99UkiWum6ORFdTkSbiWjznj3FFA9zRTtJ4omDM3jjx2/G2//t9tLzp9twea02UoYkEpIzt6NF40RhKuPgeJ+yiUS+GWecaQ8R1yyVHJ5pGuKxaYjNN5S7th/C1+95Uv9m58nZfiDt2ylrxrUx2ZUCQfab75H+F78nCfjrtr1TeOcX7ypVpfG1Rqth94brjLOMQupouO5kk3Crm9rvvpKw7Dg4g9/59zucKqqpbgzXJSVGlVLCNdSUJOy6EmWQ6qbbHzvYti1QNFyPRG4iwf0aqYSpuskpSeTMwdY9kzg008Tj+01JeulIykgcmmmizFmEDcoAoEQfzTgJF5FI55sdWT3bchOJVqIQUkr4urFJ8P6RJ/grah/e+5/usrWDxiCIhGuV2KPQtg0RLQHwBQC/pZTiPL4fBXAqgHMA7ATwAdfNlVJXK6U2KaU2rV27tseul6OqiUQT2/ZM4qateY2Fv/t26klUicqHb6aND3urRD8KsLopLJxnq5uY64+TnKtmbj2/r8mRtksIB5iEja+VKMtAKjYjGSdRBu734/tnEAaEDavGUrHeMN7bRLTI9dmFgPj8RivBZ255HF8VVdUkeAzGqiEacdLRM6rRSnTenUpQtGPYkkXUg3cTo5NNQp7zO5+7A1/88RPaQ01ipo3hOlc3pc9iz7P3fPU+nPzOVINrb6Dd1pPo1btJWQSf7X62ZMXXHa2k78wlScgCUjyeNqPAm//B6YaxlgwXWMNwnUsSNjNkg9WwdrbXpkPiT58plSTaqZvk3GIGQHs3WWpeANh5sPuCV/1gEERiO4CN4vsGAPaKLW1DRBWkBOLflFJf5AZKqV1KqVgplQD4OFK11hEDL66Jeguv+MB3japYP9iS6gNXZJuzC2agU/eG62arxCZhTT6epHLD3GIVTNc5i7pMPy77tX+qoTlqqXJqGGJ7Z3dUntzbD0xj/fIRnR3T9Wz6u2GTyO+V97PIXa4Yc78LvhYv5k4GwHor1ptXFFLHtBxRR5vE3L2bgDxexSWx1FuJ5mRtYsTDw+/Qfu5P3fQIANMGwd+7rUw3V+8m+3zb8J1YkoSLSEipl8fTVqdy/w9MN03GzGI4KkExd5NEu/tLScJO92FfIwzS6nyqZN3IOc0MnSt3E9tD5pKufi4YBJG4FcBpRHQyEVUBXArgWqvNtQDenHk5PR/AIaXUTiIiAP8I4H6l1AflCUS0Xnx9LYB7cATBS9KlbmIjbDsdfzsiIV+4K+J6tBIiIHe0ME9E7Ykk/NvvfPygMzkaT7RO3jJ2PzesTKOODSJRUDe1vaQWk7cfmMGGlaM6rkBy0+1tEqpwzN7YgPIssVKSADrHSsiNtxK6JIn+1U0zzbgtcZXqpiczLzNX4ap6M9F69zI3a5Z2yzycOK6FIcuYDjqYrsBdx3x+uftumbpJ9o3Pn6zH2LJ7Eh/5ry1Gm4PTtneTeZ3ccK2cdhjX/dkmIbO9SuJfiLhWuSSR9sH8/Vc+8SN86cc5b82SRB6Amrfn9Vjmhjto9E0klFItAFcA+CZSw/PnlFL3EtHbiOhtWbPrAGwDsAWpVPD27PgLAbwJwCscrq7vJ6K7ieguAC8H8Nv99rUX8KSyDddxonSOIzvg5qt37sBVX38Ah2aahgjdjlO2F3e9laASBqiEgRknwdxdNmF5w28laX2D5aMVHJhuGlWvdDxCl4brVqKMPPjHdyQSnTOFtrRNYgYbVo5pzltesxBMZ0krgJ2qQRXuW8ad8/3HHLEnLjRaic5umhIJ8z724p+LC6xS7Re4/I3nmFuSiJ32KyDfIHNJIv999+HcwGtHyEubhL2R7Tw0gws++F3t+ZbbJPJrf/627fhHkVRPwqZTZWlKuKusbnJ5pcm+SXXTV+/cgb/65oOYqrd0m0MzDT0PqmFgvEMZJ6Hglp54yhyYamiDOXPzUpKU6lxXFoFUkiiORTNOSwZLlWlBkhAX5LUzyKp97RAN4iKZe+p11rGPic8KwDsc530fbnsFlFJvGkTf5gpeqLYkIdNUyMn9+P4Z/K/P/BgAcM7GFZBrupcssI2Mk62GgaGHzCUJZP3K1U1KKfzUaWtww/278ePHD+CkNePGfWTU5uP7p/HNe5/EW190MshyzWjGCcarEWabqaTE+Yuku59BJJJi8jUbzTjN5bRrYhbHrxjVC5KrcNnjYY+Jy3AtNzJGKZHIdhxWN3XycJKSRBQWvZtsSSKgvFDMv/zwUfzCczdg1XhV/15GlGabcUGfrfvgeJbAkf+j3kq0cdZ+Ltu7SY7xPTvy4LXZZlJQ/ZVFXG/dPYUtuyexdc8kjlsx6pQkfu8/7gQAvPVFJxf6a0tPugSsJeFq76eK27BtPw//PllvaeZpuhHrOXJgOjdcp8GFpk1QJ/gT9hgJPvac914PAHjkqp/R60+O3VSbfE+tOEEUBAhEMkGGqwgRMwc8fzSTmCitam7ECf7mhoeQJAq/88rTC9cYFHzEdQmasZtIHMxeaDUMDIOujCeot2Jncr78u1sfzwFp1ShAxQr+0r7rlropTlKVzxnrl2G0EuKu7fkGYE8wAPiPzY/jz752P/ZMFP3MW7HSeYMA4LgVTCTEs5VEXJehGSeZwTjdqCvCmKjv22YjdsVJuKJwy9xKm5Yk0VndFLdVN5WpbX7tms34s6/dj28ID6/0erahO33+tqVHHRyiK3dPo5Wrm8q8m2oOddODT+a2q3orLmyaZeqmPGGlORddm/ih6eLG5/L4AYBJS1ovBtvl12e1m6zbLSWJSS3lt8Raaep5MFIJRYBm+iHSLrBuSaKdd5OZa61NfQ8FnZYj7X/++0FH6p+pRgtJktdCaQqCqlRqg1MK+JsbHi6kBhk0PJEoAb98O8Edv9DjVowY6ia56G1jW8E4XfIb55ephmkwl/0bkC8gWw1WDQOsHKsYRM0Vyr8nE5ddgXXNOMFYJRcuj1sxAsAkgM0e1U1NkaAtCkjrfw9KScLePBzqJtu33d5wyvTiubrJrbu3kablCHV/OwXTxYnC/qkG/vvh1Put3oox04hx9fe2ohUnhQ2fnRLaeQTVWwmWjZhCfqNVHOe6cJe276P1+hmR+MLt2/XGdXCmYVzD9KJzx0l8/e6degPWkfUJ36v4DA9m+Zj+9eZHtT+/bHfSlV/DLT/ZD6DIiOVEIvN+Eu92JAqz+JX8YrnhOsZEVhNlst4yGCotSUR5BDpfoyLUTS5JIlGqQPQ0N+9wU4fjOq0kyRL8UeH3gw6COtuMrbQ8aXvej9YuqRXOmS94IlECfvk7LDczliSOWzGqs0sCxcAmOYkL7oklEdfcrhplNgk5AS0XWLv2QxQGqEZB6aRl7Jkoz+PUjBNDkliTTUS5Acn+yhQOLoxU0uRy/LhBQJprMySJNhsx30/eppXkUbi//uJUrVG26drqJh7jf/rBT7Bl92ShfT2T5IB087AJrcsgKQnUVL2Fj313K9533QP44u1PFDh8zsvVLqCu3oqxUqis5HPY7ZaMFFO4APkGz8/yNzc8jK/eleq85aZct9RNKQHOnjUb4x9s2Yvf+Lfb8f5vPpDdKzF+lzhuecpYPPhk6sl+1dcfwGdveczZnomH7VDBQ6wjssVY1SphwY00N1y39LNNN2Ij5oEJyUglLDBcrrQcEnGi8LBVV+WwQ5IwmBtbkkig60kAgBKv84BT6jLXHb9/Vk2tXeqJxIKDN5Mdh8wozlySGE1TUXAMQsMtSaQbpSW2SxdYSSRaOZGo2oZrK1WCzX1FARWIhEsfzoY3F5FoJUpzpkDuVmoSK9Ne0E7dNF6N0BR63pByrm3KMOyX22xc3k0yffTrzt2AJbWoVH3jUjfFicJ7vnofvnLHE0ZbpZS2CQHIpLn2kkRiSTWT9Vi7dO6ZrBckl7FsfNupmzigUno0uSSgeivB0i4N17KNUTekFRcyANiG662ZazVzvO3SdqxakhK3B3dNIEkUphotbXMo8+iy048keu0U03aMVIJCtT1D3ZTda6reMjZqjswfqQSF54tEWg5XFxOl8LDFUExYUhVgxzU5JAkibas0JYmiuilOEmstp+2ZSKzxksTCw+U5kCRKLxTmmJjLmWmaxl2Zn6ZtWg7xG3MOtSgs6MM19yb0rBJRmBKJdvEHALQtwlXKtBUrvZkCwLLRikPtZaqC2rlyjtciw34QBqS5Nqm/vWfHYbz7y3frZ5OceR4nYS5AfU0ijFRCYyM5ON3AnRmXyuM7KgzXLrdCQIy/jpNw2SSKm7FhvKy3tMpuplGs9y0libKxY+M5q8hcfeV2o9UIYUCFftqGayBPUyHnTr2ZWO6hxYhrjtVYnUk3DWsuSvCc27ZnCtPNGErlG2oZQzHZMDd03kBHHZLESCZJuNRNk4ZNIjY2Yk0kolAf1+omR2W69F5szwEe3pUTiUYrEYbr3HmjLCcbB50a6ibDJlGUJFqCAQXyeewliSGCi0jEKicSx2aeP8wRy+L2BpGIHESiJE5iVovEaapqF2diG64ZUUAF6cPFfeaSRDEVQjNJvZsYS2uRJlZ7J+t4yfv/Cw88mYvdnSSJsSwVhs7FI2wSUvL69gO78a83P6ajvA1/c4d3k9SbhwFhtBpgVlzv5//uB7jkIz8wxoDzBDXjROdGsjdwSaSBVJIocoRFNYJc8FP1llapTTeKBZjYJvEr//gjvPJD34MLbBeRBNuVVqOR1b6w3ztQTMsB5NzoxGxLSyB1S5JoxsWIa84IPJrNjXaShLblzTa1tMobaplqUimTaeFm/M7kfEgLQpmZbfl9TjdiQ5KIY4WVmTS8L0vjUavkLrC54dptk1hSq+jn3HkoVztPN1rG+9AOIg4JGIDBJLm8m1w1a+yEmbZNwksSQ4BGK8HJmSspI04UDs00sXQk0lwZp0aYtlIk8MQYrYaYqsd6oQG2JFHUrY5UipKEridhxUkwoiy2ou6wY0iUxUyw0XxUbEzjmkgobD8wg8f2T+NBQSSUam+4XlKLjLEIKbdJuHTynDF1qt7SGyTvBXataB63MCCMRKYk8ViWw0opVVA3NWPl9PoCTHUf4LZJuFJAS0IwWW/pzfHQTLNgS5DSga3CYHDUt0EkWmUELXVysJkafi+SSLBr7cRsU3Ois82kwAEnFkOyK5M++Xzpjr20ZhrYpes4z1Hm7tu5S8v5rA3XkcNwXQkRkFuSmHJIEivHUuln/1Q9u2ZYIIKuBH8AsCSzzyWJMlPyC/daQJYYdUsSsWBonN5NDhdYm0g0vU1i+NCIE5x30irjWJwoHJxuYOVYVS9g5shZCggDMiSJWhTgh9v24ZUf+p5TLG05XPk4TsKd4C/9bqubQodNQsZZ2G72drI/XgTjYtGPVUMd1MdqFqkKYsJShvFahJYgEkFAOteRK0UIj+V0I9b9cEkSktsNiDBaDZ06/tSBwKFu0rW20wp6r/jAd3DPE4e0GkrHSQSB5r7lMxvflbJy7uREIrVJmO2X1MzYCFcUPCcZlASlrBZ6LQpQjcLSokPSJjErNvDcKSE2OVZhk+BrcPAdvzPeDJVS2nBu92tiNt+wJ2Zb+MR/b8PWPeVZk6XjgV477ALbcNgkHOtmYraFyYz5Sb2bcrvaPm2TyA3XksNPn8eUJHj8bTXUVL2FZqy0fcFVPc4lSUTCcC1/d9kkWoky0rNoSWK2iTAgLSEdCXgi4QDHK6xbZlLrVqJwcKaJFWMVzXHzwpluxIgCwlg1RL2Vezex8e3QTFMbNJuWJLF7YhYzjVhzYSOVsGhfYO+mEsN1JSTUCt5N+Web47OT/XFbqW4iIlQzLpUn6YxjMZdhvJb6pPNYhESoRrkqJr1H3l56pnB/y+pJ5EbHzCaRXU8mfmsI905D3cR1g2OFJw/NYtueKTy0a8LgzoFM3VRQFdrqJpPLnKzH2lC7Z6LccM3Y6vKwaiWoVUx1k00EmNDVKqF+RxLcXAbh1cVmumZpVd/L9iazs7VyASk7s29sOToApus4E8snD8/iz752v84X5YLcELW6yWW4joo2Ce0oMFHX5043WkiEJCHVTfxcuVtsmN3XJAb8bLbLNUdzMxFx5Vdy5XEy1E3Su2mqvSQxUsn3Ah1s2ya56KDhiYQDrSSPV1gqOCVWNy0bqejNlA3W040Yo5Uw3ajjPN/MqIiq5Y3RnEAJzvvzG/Gmf/yRaZOwUkLoBH/CJiEnShhkLrAOwgKkE15uyJOWuomvP25xuhzUxxtH3ZIk2oEXEW8AoZAkeNzGxPiwdDTdaGkO1a5BwM8lVVijlVBzyQ9ZBsZinIQShVxyo3MrVvnGy3ESYbH0p1OSyO6xbCTCtFA37ZmYLTgPjFfN8XW64TbjzHAt1U3mdbTUk7k+lxUdkvrueibhTtZb2s9+thkXssDKuuV/8KW7tdF3xlY3uSQJUSNkTxeFgRhSkrCD6QybRCVAFJrV3dgeKOf+VD11Glih1U0ZkYhCQSRMlVyiTJUYrwWZEh9gW1OSp3pxSBLydSSCSLBgZ3g3dVA3LalFeh62kjS1uZQQ5xueSDggddM3/u5L8fpzNwBIN6d6M8GI0BdLddNoNUQtClFv5puTzIXE5SYld8qLY/OjB4S6KSx4FdkJ/g7PNrFqLPel14Zr6TYnNpYoDIwNedpSN/H17VQRbBvh+xrFlBw2D5nLaIkVOCbjJHjcpA1EShLjVVvdZBJMw3BdCbVK4q7tB8Uzpaogovw9GJJEkscINJPEUOGkz17k0F1Eg9/n8rGK4Ya5b6qBmWZspGiRKiSgmLkXEN5NNaluSu/x4RsfxrP/+Ju6X7VKMT4GyDemc0/My7DUm7lhV8bAGJtbbEpGn/7RY/qznUcoSZROC6LbiDn7pOU+3g4u5mO0WlRNSkliphHjD750t7NKXSpJAMtGIwQE7GObRCXQc4ffPRMJW900riWJdM4xw5dLErmW4Nz3Xo8b7tuVXi8MDHWYIUl0EXFNBMO9fqwa6THnNCJeklhgSCKxbukINmULjV9cJcwX8IxQN41VQ83N8ySRL3M6456bSa7PlAbtWSFetkvwl2TcoEyP7YqTkJJEGJBhb7DjJDj+wc42mgb15eK2kb3V4ZYp78GcGG8AIZHegNlwLonE4dm0wNF0Iw8S47Uk6VErMQ3XbJN4fP80/p9IUcDqpkoQaK8q07spNzqnkoRpk6gEZp6ftB/l6qYVo9Usf1B6HaXS9yszhdqLe5uDSHBNi3GHuumD1z+EiXrLUI25iAS/r5PXjOORq34GI5UAs8J1c01m+EwZGkuSsAxNf/xzZ1p9ySUNW40ZJ3mSyJ1dEAm2B0iHCx5ilwtsrRJmNgmFr961QxOxMUtCY+NyFBCWjlT0ezYliWyNGvUk8mssEXaxWCksG42ya6e15Zng37fzMPZNNfCNe5/M7hFALo3EQSSkI4ZkvIh43uWShCy9m0oS+Xw+EvBEwoGGxWHwRG5l6olqlHPlvNnNZAnbUm4+9axIN+588k4JSYI3DiYSy0crYpNKr1NWT2KqkeZvWSkliSxOoswFtkAkCuqmtK09+aqZRJMTCW5HTnXTEoNImCkjwgCohulzT862QGSmWp6YbWn98pI2Ngm5kQVBHifxrft2YbLewu+98ukAUm+pVpwgCkmnKDfVTUKSiBNHnESxnkRua8p86IXhesVYBVON2MhFtHeyoa8HmDaCWhRg72QDDzx52LClpDYJK07CUjfxvKtFIZaNVApJ4uT4cLt6M9bS2oosBibNApvPk+8+tNvI/wWkc0IW2JJSba2Sp79m8Ly0sxW4wMTHjI0xDdd2MF2YbaJyrq2zvH2m6y2dnluqjKtRoON7cnVTsTIdICWJlDFbxpXuMjd4Jkx2CV3pZguYNjlXWg7puRZlsRRxYtrTcrVogijIma0jAU8kHGAKzhtmFOYiYiPzPGEOOI+TEJJEZgwMAtKBOtwGSCdNFKaBZcxtrR6vCkkic4GVWWCzjWjfVEPnvJHZRqMgQDUMS9NyhEQGt2XHSXDbyCYSUWARiVhzRDaXffbGFfg/r8qzUbLKiDeAgHIxeaoRoxIGhnpqYral+5VzcelvRloOKx8Uq5vYm+aUtUsA5EbZKCBUIib0OTGQAUupNwlzlnmq8LKUKietTt2jkyQfu2WjFcSJwr6phvY+mWm0DP2x3E9PXbsE+6ca+JVP3KIloCTrk22TsMea00JUowArx6vYb6kspM0GSDfXeivRRGLpSAUjUZhlgc0H9zO3PI7P37bduFY1DIwKfNIdOyQy1JgAtB3gycOdJQlWV7kkCS5vyuvmvJNW4aVPX4swMLO3AsDLn7FOf141XtWSRBhAb+5VMd9S1+V8nQdUNFxLD7tWojSx4bFnbcIBm0gIaYXvBZSrm2QWgyCrrR4nZi2Ugk3CIhKdsjH3A08kHLB102GQV/fiAKZalHJQubqphVGhbkqySSzzzTP33orTQidRSFpvu3K8mksSOpgu59h4k3z/Nx7EW6/ZDABudVM7SaJaLknoBZMtIo4ot20S9VaiXflsSeIlp63Ba55zvOaWeJPLJYmcA5qqt1CxxmdiNq/DUXSBNTkzXjSpC2yAmWaM6UYLtSjQagouqFMRm5wZcZ2rrVpxojlWZgAqQpL49gO78O+3Pqbv+6W3vxBvf9mpiMXGwpUKD043de6lqUZsSBKS6z513RLsPDSDvZN1HQmfz732wXSyMtrq8WqBm5W++Xy92Was1U1LRyLUKkFBknChEqVebgxWN8XZhjVmOTvwOHSjbtIqSZkfTOUcNJAzGe/7hWfhZaev05KElDCYMQBSqYLtBiGRVhNFocnJSyJBjvnMUk6SSa5LRM1sIHe6sG0iqbqpSCTk/SXNl+MfZR5QsbCRjVYiYy9grYGEyz44KHgiYUEppaMamQOUATBsk6CMg5rS6qYEo5VIG4+Z4ksj4CN7p3DPE4fQStKNthIE2n975Vg1927K0nK4XOskDEkiDLQrpKvAfapukpKE27spCgPcdOUr8I3ffgkA5qZNY2ZKJIo2Cd6IKyGhIiayliQEB9RKFCqRyaFOzLb0eC4dMYmEUXIyI8L8XCNRymkdmmlivBbpe7B3UxTm3iANoW5i9WE6VkoHRvLmJNNyfPpHj+EfvrdN5wyqhKR149xmuShny04F0/WWoVKTNTxOWTOu1Xe62lgzV3VKdVOjZQZ68RytRSFWjlXTwD3xvnm4+Ha1KJUkDksiEYUFF1gXoiAwJEwpSQQBFYzxK8fTcXClo7eRZ8UtqpvsLLDMhXOcRF14RI1EAd7/urOwaryKE1aNaQN9GARYPZ6qolKJId+km+JdBlR0Shit5skEW7FCNUylcU0kaqwyNp+zGrnVTYHI3SRzkkkpmVPXtMS84swFfC2Xd1MnT8N+MBAiQUQXEdGDRLSFiK50/E5E9OHs97uI6LmdziWiVUR0PRE9nP1faV93PvCF25/Aa//+JgAOm0SWdIu54SUjkVZxzDRahrqJuSxpdPuLrz+An/1/30+NqZkkwahl6oCA0kkrUy2U1UBYIWwSUgTNc9Ar4/cxw3BtqZsyQhSFhONWjGoRPVV7mSUko2yx2ROTn6cSpMGAPJG1JEEmB1TJUqIzJupNrWu3iYS8VSuRSQNJb+r7pho6ABBIufJmkhZ7camb5AbfjBPtRcMc4mgl1It1st5CvZkYHDrr+3mcDSKREfDpZmzoj6VfgIya5c1bBvTZ6iZpd9CSRCXAqvEqlDKL1xTVTakkwfN1SSZJ2C6wLtjvKc8CW1RjAua8tFOe23ClOs/TnJuxSLwOw4w4y9rYo9UQb/ipjbj9Dy/EqvGqJqJhAKzOkg5WQtMFtWVJEvY6G6mECClV/SRKIaCUIGoikT33rolcYiJK9w3J1OTZZvOiQ7FS+JebH8WXf2wmmeT8TolS2laRqpuyOZs5YhQkiQ51UvpB30SCiEIAHwHwagBnAngjEZ1pNXs1gNOyv8sBfLSLc68EcKNS6jQAN2bf5x3sxgbkRELqMbkoEJBuZLxYZ5ppnEQ149hYZ+qq27B7op5Nhnz4k0TpamVEZKTlKCMSMupScu6uNAGpuimd1KmB1S1J2BxKNaI0cln0IQoI5OC8eJwqmcdNZBOJjEti7rYamkbPidmW3hC0C2z2CNIbhGsFEOWGayANpBuvRoJYJplqj0x1E/v7i0ybUn3BRIe53Ol6jKl6jNlmqucOKJUIeAPmjV2q/5hIKAWDSATieVcLSdCuW2y7wDbjxIg7YLVRNQz0vaTKyY4mZkmCVUW1MNSSRCciUY3ynFtpX3IuOKCiZ9GykYrmmNdnOc7KMG7F0qTXTf8zUzHrlCSUwYBJh4Cxar4ug4C0JAHka862SRCKEvtIJUQQ5BHXLI0zAeK+7xaSRCWrY21nLQZSghVqSUbhD798D343q+THCIPUbiI978aqEZpxWoGSJQnbwcRVO2ZQGIQkcR6ALUqpbUqpBoDPArjEanMJgH9WKW4GsIKI1nc49xIA12SfrwHwmgH0tSMkN6jVTWHupsdBdkBqdOMiJ9ON2LBJpAuICpHNAPDQkxOpx43gzhKlMNuUaapTL4w4UaX6xpW2JMEcdKtIXKR305oltUKcBE8yu2aztkko81pB4JIkcnVTSiTMTTQgyqK4pVrKUjex4dqWJLJ7LR2JsP3AjNY3A7mr5L6pBsZroTEOrSTJ8lqlbW11ExO6ZpxgppHGNPA7GNfOCWmA3Gwzzgzh6e+84fN4Lx/N34esByG5voAIm05ciWOXjRjqwpxIsF3KdIFtxspQ37DkwZIEYBKJRKVElCxJQr/nLEK/nqnk2vndR5bLpcxIHASkE//l7UlLCGuX1toGfrHKxhVMFwRpsOCMQ5JoJco4R2ZSHa3mQZAhkZYkDs82tSSXMnxS3VRM5lgNg0ySEERCSBLMTMjUKlFI+hz7ecIgZ4rstcM0LhJrqyHUTXxOnCSG+pQx7Oqm4wE8Lr5vz45106bduccopXYCQPZ/HRwgosuJaDMRbd6zZ8+cH4Jhu8sBouRkNlkrLkkiIxK1MFc3lUkSTx6eRSUw1U1xYtY9ZvVI0+LiJUzDdaDdbfMIUCFJUG64Xre0hkacGPmfGnrzMKcEq5vkJGTdrs158ThFmTis1U3NXJIAzAR6piTR1JHYuQts+hvf/sTVY3h8/3Tm3pgRiWwR7Z2oGzaJeivd+GtRkElnaZoNOT5SWuOoed5YmZOfbqQBcnWhRpTPw0TnzPXL9LOsEMyGtEkERPj8b7wAN//B+XrzAlIbQyI8rGpRgOefshpve+mpeMaxS1NJQhKJmdwFlpkFGV0tiShfT9of2ImgnlVAG2lDJGzVqIy4TueVKUmEWWwCkL7Hj1+2qZA7jFHNIsbNtBy5Dr8aBVrCC/X8yiQJYcc4Y/1So7+yL2uycW60EmOTNtVNRW68GqXEg+MkOCaH1/x4tahKixwMVD7mucRj5+xiFSdL29KwnqeUySWJNKOyeCdDTiRcr9/ucVmbbs5tC6XU1UqpTUqpTWvXru3lVCekbtS2SbAqhDe/JbVIl0ZsJQpjlRC1irBJhKSNoTbCgLDrUL7oE5VuEEwkckNruTpARrtKjwfeBKWnUxTmXigvOz0dp28/sFv/3hJclQQbruWkDzPDdZlNIvX+CjXRmNVpOdJ2OoGe5Vp52OkCa3o3nbh6PCUScb5Zs/Q3lbkh18Q4PHFwRtfqjoLAiIeQAUutJMmkQRHnIaLqp+ppAFU9cwEGctUBE8HloxV8+R0vxDkbV+CsDSuMMeRNUtLgXA2SEsHJhhkkN16LcOWrn6HrckgiwQS+FgWa2OyT6qbMqMwYqYRaauA+jWTpTFqxKkTaS9jqJhnYyRunRBiQzqB86rpxvPTpa3FcidopdTwIDCM0T1smEjzNeIMNM3XMbDPG0pEI977nVXjOCbnJskgk8nGWqbpNF9iiTaIahggC0oWlWFq31ZISlUz64Pl6aLqJr9+zM+tLIIJoTWM3X4ttEmwLkx6BzSyIlNeVlCbiYbZJIOX+N4rvGwDs6LJNu3N3ZSopZP934whActf8Eux8Q7lNooKJ2aYmHqPVUBucWxkn9/RjlhrnMJaPVvQmvm5pLTXECSOn1qu38T6Rk0SG6nPKbTnpAyFJvOhpa7FuaQ1fv/tJ/TtLBbaus2bFSQApMSAqVm3LJYl0MUWWJMGLnDnrqqVyawjvm0LEdbboTlo9hn1TDUzMtvQmLdVu0ibRiBM8vn8GG1eOZc9GWTBdHtQojfyzzdjQr7N6brLe0hLhVCMnElrdlI13FBLO2bgCX37HC3HSmjF9HVZnAKZ30/LRiiFJHZpuFvJHyX5PCO5zQsRJsEQp/fUThySR2lTSNCVSkmglSVsikaqbzPeU3iN9HtsmEQaET/7qT+G2d1+A33vl6foYkHosPfDei/D8U1ZlzxagVgkNSUKqm8z8ZJbhOpO8x62ob9nXMCCsFkTC9FTMXVNTm4QqXIe9m1gyq0SBcNMujhm7ufJ6+YWP/gD/8N1t2TjmLrB2DIn2qMvaxJlNgr3ogDw+yJbIgeF3gb0VwGlEdDIRVQFcCuBaq821AN6ceTk9H8ChTIXU7txrAVyWfb4MwFcG0NeOkGKgLUlwIjE+vmwkwuHZlp40bJNgL5gwIHzwl87B59/204VsmX/0c2fip09ZjU0nrsTxK0dTm0SW/RPIN+tmrJxGqb+99BxtKwHYBdY0FNv+18cuH0k5q6VVXHjmMfjuQ7l6rlkqSZCWlOS1UpdBW92Uq5EMdZMwXMvxs9VNQB7NWszdlP5+4qqUQ31k35R+fna5BFIdN9/3yUOzmGnG2LhqVN9XShJybFtxgulGyyQSWR+kH/xUvaUXLQ8Vb+xSJWMQ8Iy7TM/J2wRBmvKZx+DQTFPYJEwvMBnfAaREIiBk0bchltYiU5JIzJgM9p5riiC0WiUNvmy0ksJGL1GxJD5D3RQAF5xxDM7asFz/zp52q5fUNFHUhDWrJKgDVQNKA/2EJCHVTaZnWK7ObCVJmi/NQdwkoxMIm4Tsh1Q3VVndZM3nSsRefBCSBOWSRKWobqqEqQdTrNI5JdOjB5R7xO2yYkj4OYLMoYWl3IpgtlpxZmMLTEYy/W3+JIn2/mldQCnVIqIrAHwTQAjgk0qpe4nobdnvHwNwHYCLAWwBMA3gLe3OzS59FYDPEdFbATwG4Bf77Ws3kCm4IzGRgTwVQm64jtBoJYZLnDZcx6mBc0ktwqaTVmkPjb983bNx/hnHYM2SGj79688DAPzix36YTtok1w3LXEM2x37WhuW45JzjdZ5/7qNUs6TnCkkiIFx4xjH49u++FOuWjuC4FaOYacZpgRtRPU9uBtyPpohL4DahQzzXLrDau4nVMaZ3Sm64zg2iS2sRJuotHJxupjUDsnPttBwnrk459Ef3TXeUJLg28wmrxnS/U6kobWsm+EttEpKjzv3g83GerLdKbRJy7AxXX+HRZQ0v3vqiU7Dr8Cw+ddMjODzbFFHfDiIhNtLDs03Uotx+snK8akoSKs8PBuTBdGlah/TaI5l0MdtsTyTK1E3MXb/s9HVYOhLhdR/9YTYORS0yH7LnQBQG2suKoeuPWC7TPHapi2hq7JYJNOV4MaKAjPxSkkjoOZ9t7PZ8XjZSQRjAVDdF8h2TZhTlvTmO46at+4zrsVEbcEkSke5vFJBWh0lX8mkdRV6UJMpiqQaBvokEACilrkNKCOSxj4nPCsA7uj03O74PwPmD6F8vkESCjdi5JJHZJIS6CQB2Z37So5UQ1TDUnglSJ8yLYMPKMa0j5QXOxrF6K9EGz9wbJyk1EEtOMRJucS7DNRvUTszSSbC+fXK2hdqSPFgnctgkmrHpYcXqpoJNIlvFzzhmaRo13kmSiHJJYslISiQOTDcwVo1EcZb02mzA3pht+E8entX5ekYqIUaz/E1cTQ/I03DzOZUo3Qhk/AmrHFqZd5NLkpAujtMudVMrV+HIcZOfJSct8RsvOxX37jiUEomZpu7PiCFJpP2WG+nEbMuQNtLUHLmqNLFsEixJcEoYPjbbSiWUXtVNzDQEei4W1UL2NYDciyeXJkmrwvK+Z9ch04tHGq5bSaLdzm3IvgYBGSo+M+I6l56luundP3MGjl0+gqetW5J6KgnDtUmAAnztN1+M6UaM93z1Xty0dZ/IRgA8tCuv4ghwMF0mSVhEQhquA2GTqIQBzjlhBQDgW/c9mQahMjNmBDgOsSTxVMNkvYXXnHMcPviGc8QCyIhE0zRcMxFhI9So4GJnsiJEDN5QJdfLCIJUv1tvxqhlG1+1jSQRCS5MHxMBNs1Y4cEnJ4wSpYG1cDnFwFQ9xuoluU6zkOCPPYWaNsFxxElkk/cvX38WAODRfamoLetJyGtKmwSr4w7PNjES5QY+2wVW1i+Qm9HKsQpmDsVZMF16nEX9DStH9bPVYzM2oiUIxkwzNjzGWJLYPWGqmwqG61asEwjqcTPUTfnmYBMJIDe8H5ppFuw26fmpdMp5wxpZcJ9Mard6vKqZFaDo3TQShTpNCc/LkSzp32wzKahDJSpRUJAktFunQ43mIhL2WuLrhQFpozqDr02WTUIaruPMhlTroG4KrfHma9z+6AGtRqwE7K2Xr9GfPeu4tA/s3ZS5tNtSCjMg0hmDpQ/plsvteZrsnig3XLMk0owVKlFq19x04kp85pbHjdiiqvFOhtsm8ZTCxGwTS0cqxqYa2USiRJJgdRO3tTdmwEylwWDjmOECy0SiVbRJSAOxPia8m/ZN1fGqv/meQVxsFQCX0WQbjPR6keANV3J6UeCOuLbtGUzEZi3DtVQ3MYfJm/9kvYWacEPNDdfp/5pj0wDyKN/xWpTGYkRpBbI1S6o6bQSnTJF1EXTFr0ySML2bMklCbL5Tjdwmwe+3Llwrdd+C3ODICeQA9wYqiYRdHY/7zTYJl4s2kG5sssKZS5IA0sBAfi+cPVfOOxcqoZljS8bNBHos8vauZ+Rpxe+sIuZAzXKBTZRUN4XiGtm8DzlOwt1v27sJAL73f16Om658hX4n/+fzd+GffvCI5tylC6zsfxjk3k1SWk/7YUqLPFa8ng9YSRfZcwkoZo61vZtkCiAAuOhZx+Ine6ewd7Ku14ycI/MpSXgiIaBUWqfBrrZlq5v4xTH3xeoIjrjmti7d7ApHbVrecKWOlWMx2FNKIpckzMnMm6/NwQBFjoo9QqYaLdz+2AHccP+uwjXls0qfdOaM7X6FBXtGzmlzHwGZijvnipjgTs6mSfq0JJGo1IOsadqD7L4y8WV1US1rJ90f2VCrE/wlUt2U2iRkRtMwM6pKddNUvegC22glhXEDzDgbvZkWm2FJLUIYEA7NNPNiQpZ3E6ubJJGQ3P+q8Uoh4tqWJICUCPO8HK2m5WUn6622qac5zQqj2Up0JHweBV3Ovctj/PxybFJ1k5Qk8nOqDqlArpfRDjYJflcnrB7DcStGC4wbz1GZliOwiESscsO1YUgX7Yx3TSlhOTDdKKjLXGMD5IZrtknESZqWg89nIjIjXLBNm4RXNx0RcJWupRaR4AUwXbBJmJwmB9MBaYGh5YIgPO/kVfjRT/Y7OR/O1cJGZEDYJBxpEzhTq1yY7FEE5Mnf7HtIaPfO2Rbe8qlbxbWLhmvALPwSZpxXbHuD2AswMCWJsCBJ5OqmpcLddJlwDU2UwrP/5FsAoNNhsMolNCSJdKyZ+69GAVA3I+iZY607JIk0TqJV8H0fr0aGakCW/pSGa1cRmPRYnNpe2qibiFID6KGZJpbU0v5Ke0MU5l5ZMjZm2YhMA1LDTDPWQZ0u7yYgk4TYJhHlTEVbSSKygukSkTtLG5Pz9u3UTaTnAEsFQaZuKkoSRHm/OQULwDaJNHeTW5IoV33ZmzS/t9QmkRTaBJRngU1tEtLu5CASGeMTK4UDU01sWDmKbXundL+phEgwcxNQKknMNPPaNbKfM82c+TRcYOcxLYcnEgI6176ln2VXS+ameXHxImVO01A3NcxN7FNvOQ8HZ0wRU1/fIUnUhCRhq3Wkdw1RatQNKD/HLkAjz2EsEZuyhM0R84Y+K3TGecS1W8LR7XTkM2eBza6pbRKBlj5krv6TI5GtU9xCq6syDzLJ8bGth+0IfA+bSHCiPr629ASbbSYFIjFWCw3X0kYrESqWXFJybYx6gYvNwUUkgLQWxaGZFtYsMe1e/DlVk6UEgDdJToENpJIEAOyfbuD46mimbsqvX9OSRG4/kdxpW0kitHM3JYYHkv1cTnWTJUlIF1iOBmck4trMdJk2jzTL6kyjOxdYox8FSSLQ7WRktGyfeh6yV1GJJBHKOZ2ec2C6geMFkQgCKn3/WpLI1FVJwnESgXF9pYRtT3R02NNyPGXAG6Zdt7eQlsM2XAvvppomEi1jwx2thqXJzthTSHJGrEffdWgWH/3OVqO9bTxjDkVLErPdEwk7ZbitIpPqM3mtgIoT075HpDfR8jgJLUlkY8mlO3ktyeyzuT67aJPhXEmGJAGHJNE04w1YOpzN0lPYBXTKUi8AprrJlqIA080z57jdm8Ty0QoOZzaJahgUbGJcFKkWBXqOLbUkCQDYP5kStIK6iSUJYXgfESqttpKElbuplSWbk89j6/Ft5Oq27B1qzpu0ey5Dq5uCfE4b49GDTcKez2XqpoByvX5gSBLpvVJGzCSW8nNVPA8zfQemG1if1WXhvpSpm0a0d1Og33dqkzBtOHwPAHjZ6evw8ix7wlBngX0qgaOtbU+PPC2HHXFtezflksR0My7lGmyEQSoxyIyh3IePfXcrNj96wGgviU8kckDxpsR5fVzPwBgvkSRscZgnp1zEOiGaNTFdKT3kubxAZBJD7QJbE4kVhSQhs79y1yT3x+CMuPxcPBbSBpT648cG18pR9CxFFtRNDq+ffGNMv9dbSUGK4ucATO+msimxfDQtQVoXSR4ZlSiPk+Ca1oCZZ0xKEkAxLQePx1S9pfsqn7WdJGHnCZqst3DOn16f/kZFImHH2qTHTCKhVY5BoKvmMXLDtQhotSQVlrxrDpsElyO1+yX7wdDqJspzkclzAiIRQ2TGSbhcniORBfbAVNNI9JgG0xW6CyBXN4WU21yawnDtUqFd9oKT8M6LzwDgvZuOGGSufYk8mM60SURhgPFqagwNKE9WBqRioctw7UIYkEiCl57Pm5OLP5CLMJUkTGN3N+omHSfhyFIr4fJuCoPAWFSufvE9A8olCd609AYhgrTkmNcEkTACAi3PGPlMm05chbM2LMdxK1LOjc8q2iQSk0hk75Slr4K6yRFkZm94XK3PhlRBuDZTiWVakogLKVw4VoVtVnnEf/5sOslfphqz03JIrzt+pzIWw+VKaj6ze6uwXVvTz8V2eWyJ2Z8opDQtR7OobpIZg+2YoEacVm5zqZtkX23JwWbcciIhigNZUhHP8yAwDenSfifjPoKAMFlvoREnhTr0Ze8/924KtKTEwXQAjBrjhpNAdj2vbjpCmNDqJrckMdss6ovZ9XKsGhmTWp7XCTa3AuR5YfY6qnvZrq8FSSLb8C455zi88bwT0r5YiyM1GAYFdZONithc9LGgpDKdw8MnCoOiJMFuvkFec1jagaR3kyROfCwvK5vf79kbluPaK16k1XR83nKxSDn5IhN1ICf8LEnYRIHVTbJ4jr0xlnk3cSZf6R9fpm7QkkTLIUkIJqUWBTpiXM5TThbI9hMuksPQUbvCO0uqatqpm4CilMhwx0kUtxXbdsHXC4M0wd9sKxbR9eJdV1hqFPcUXzq6wHa0SeQSnq77INVNAWm7lW24dqVh4RQsPK9kzZeQurBJBKRtLs1W0SbBbXT/dZ0UTySOCHLDtWmTCMUiBWwikbZlTqBMHG2HMOOM5Dm1KA0KY8J1wRnrcDxnMw3NBZn746f/WZL47QuernWioWORL6lVMGnVlbChDddN0yYRBg7DteN5KwGVSxKGC6zp/89qrylHQKBLkrDBEsLyUVvdlOjIbPu5gGI+Hr7/hWceW3hOabh2cdr2xgEU1XkMg0hYGx8T6sksyrqR2VSkTWLpSOpGy5JEnJgcsUx6yBuL3GDbqZuAYvwMw2VraStJOGJlapUQSuUbHXsSpe1C43zA3MRdkkQ7dZPdN10bhMgdJ0F5Te8oIIujL45vJTClBbvmiyQSUq3NzA0H0+mIa8u7CYCVs40lCa9uOiKYFPV/JQoR15GDSFTmTiQCIl2qUBIA3shWjFXwict+SnNVRcN1rletRoEmEiOVsODTL7GkFnaWJKyAuLSP7sLxTjfQKNApNQqG61J1E7sD5oQ5/U76vLJnYvC7stVNM40YjVaiVTXy+oBbvQQAv3n+03S/tSTD6qZmiSQh9Mmd1E3LRytoJQoHpxtFSSK7zmSjhWqY6++ld1MQEI5dNoLr79uFvZP1bKMtXkP2YbQHSYLtGLYqzE7gl34ulyS0XUnEFfC92eYXK5W7yjqkRrlJdsrdVAhytNVNkSASSdFwHQb52ixEXFseaOkxkxBImwTbKxhyro1W8+dkm4RpuDbVbfZnL0kcIehiIjXbJmFy0yaRsIK4SjiNdgioKEkAOaehc0jpwCWTk5DnjIuiKKOVvKaDqy/jtQhTnQKpLPdfvlZARWOZawN0BVmZLrAsSZiGayAdF7NEJfepsyTBm78s/sOus0BqA7CfCyjaJP7q9Wfj3y9/Pk5cPa7HaUOWelzHScQlNgnBBbYLpgNyYrb7cN1BJNKTlEKWVjt9BtsL789e+yw8vHsCn/rBIwXvJjlnI22T6F6S4LgGm3PXrq0lnL4+ll3eFXF9XCbtPnFwRj8nX07OBd3/DuqmyCAS1m+F9CkZ8YIsDiTUTWJtcjW//NpC7SO8kCTDIOdfEJhjI/cZlmCNYDppkzCeqUiohr186VMGHHlqc0s8F5pZoRs5OXgS5MWCRBWyrtVNOScQOsRR9vzRUoHFVUguQ24cI1WRWK6ESEzUW04PHoYrmI65pbKiQxJVcUwbLcXE33TSSlxwxjpsXJm7B9fExjDlkiS6IBIMW5JgsI1hxpIkbJfXY5eP4HmnrAaQx1Rw6nHuT6PMu0mom7irZXOCJZs9k3Uj2howNwj5DLbE+/LT16UG8NlmIS2Hy3XTliR+75VPx2vOOc7ZP34+V+0IwNy4nXESFpGUnDfnP3ps3zQA0+julCTEJtkxwZ9FsGwhRyYezNVNor2wFxZdYCURztPpBCWEIAoCkLi2KUmIVOHacC1sEiXMJ6+5+Yy49kRCYKLeKnBnQCpS23p/BuscXeqmXrybXJ/HLUlCb5IWJyHP4bZEJqfu6suSTJJo56qr4z6ckoSlbnK5PjqMiJLDfsaxy/CJy37KWEy8SRLBqOzHKgiX4boMtgssg9+z/QztEt0xB8+px22PGxtST90uwR+QE7P9U42CW+doCce/zDFXOc+Tq3wpIzdcm8Tnilecht942dOc/XMRFqAY/wC4mYVQ6P4Bc/M/IUv//tj+lEjEwujOwXRlksSYI47FNPKaY1lwiZVpOUrUTdqgHVCpKkum5ZBERjIdBUlC2CHyAmccZ5Gg2UoMict1Xx6LfVONeYu69kRCYGK2VeDOGK4oRyDfhJirnqtNguGySbDnj97wLUlCLgTuP9dqto2sEkwk2vlYu+wMbJNQFvPiMo7b+aUAmTFTLDJHYFJApMuZpt/NPnUzvpLoy813Wcl7dlUbs5Grm/Jj7eMkcmLdzrtJ99OSZG3jO8P1DJXMZhEn7SSJorppxMHkmNdNz4mtl669mywVjQ1+1S5107KRClaOVfBoRiSUym0X/M7KGKmyEqK6L9bj2OOfR8KXqJuE40UhLUdQHFPppGD3r2CTqOUeTdLbi+MsGnGi7W/lNom0Dx/9zlb85md+XBiLQcATCYHJ2WYpJ5nnSzEnJdskdEWxOXo35Z/z8zlTq5YkHFKBdIFN25rqL75emU1ist7ShjkXKo5NgyWJQts2Hj5EKBgjq8Zizglarm6Cke6cxymPki0f33Htd+7mppeNFrlwwB08Z4M5X3l/V8S1y3DdLpiOUSh1O+b+zSX1puVms1xDsn9hcWOpRXntbbtsbtmzTFnecPzYYQmnr9tpZiX9fsKqMYxUAl1//ISsdjlgezcViYQpSbhLiDKK3k3md1YxEXLvJkOSoDwvUkh5MB2RyXjlAaJkrNNqFOD9rz8LG1eNIiA4VVGy1C97N9k2CbvSoWssVjoyTA8CfeVuIqJVAP4dwEkAHgHwBqXUAUe7iwD8LYAQwCeUUldlx/8KwM8BaADYCuAtSqmDRHQSgPsBPJhd4mal1Nv66Ws73LX9IH7tms3YPVHHC05d7Wxjc8EMtknoimIlQS/tYE7K/LO2SWh1U3Zd24gluDtWQUi/a/sejNFKiNlmWtKyDC7/eE4VXjjeRpJwqT5ctStajTj3jQ/a2yTaqfO++dsvwfYDM8axTly47Fs78Bh3SkXhMly3825y9bP4W96/Ms+eRiutYFYTrqAuNSgRYSRK04XnKerd/eNzbG84mceKyMwtJGEnODz92KV44L2v1r+fsGoMdz5+EAB07QbZb5u7Z7iIhKGO7UQkdOCemQ5EtpdxEjJS3LiniLjm03lM37BpI96waWP2/PlaY0YmCnMGiSWJRpwgUXCqmyJrLHjcVzoyTA8C/UoSVwK4USl1GoAbs+8GiCgE8BEArwZwJoA3EtGZ2c/XA3iWUuosAA8BeKc4datS6pzsb94IBJAues70WSpJCK5Qgusr5yUsRanKNpyuRFlytNwmkW1KDu+mSkAG0WCpwxbTXQuXg8uacYJXPGMdvvKOFxbbhG5OzUkk2nDTLp99eyxtCcP2buJburhLGxtWjuH5p5gEvxMXXg3z+IxuYEgS7QzXQvIqmxNLR/KCVW3VTZVAF1Fy9bWqJQlTanHl/gFQSCjpet9ALlH+0k9tNI3UDganneG6bHw3rhzFEwdnsoyrRamzTFJxqZvkXCsYrq3vdqJC1+dmLNVN7rlnpGBx2HwYpgtsulalR1SUEYmmVd9FSvT2vZlHdBU0GwT6JRKXALgm+3wNgNc42pwHYItSaptSqgHgs9l5UEp9SynFrMnNADb02Z854ViRhMu1eQBFH3/G8tH0xTC3IaOuXZy1+9r5Z7kAcu8mU90kF/zqJVWsFmLmMmGTkH1wbeC1zCVUKeCcjStw9sYVxTaOiS4JoYRrA2COy3DHzDYiW5XF48bBZAFZwXTWxtGt9xjDVDcJgyIVf3fh5neej+///sv1d0MF0q6ehNBTl8SkIQgIp6wZd/bDljK+esWLcMPvvNR5nUqYcr5SZQOUqyt4nnS0SWTv8dnHL8eW912sj7ukqbbqppJXtnSkgjhLYqhEjEfVwWS4NtoydK1uItlG9JvIsFVUSta2VAvlatMiAZNrhO1fsviWVLumv5HxHyiXoF0FzQaBflOFH6OU2gkASqmdRLTO0eZ4AI+L79sBPM/R7n8gVV0xTiaiHwM4DODdSqn/dnWAiC4HcDkAnHDCCb0/AdIFsnKsggPTzXLDtbVBMbi9zCxazXIEdW24lgvAYbheZsVJyOu+//VnW/2xbRK5OsCGy3e+0Maxq9mRo+0QhcU+H7t8BETAMUtHjLY2Jx1YxnHbcN2t9xjDVDeJMqXV1DZTtkHKfkt0UjfJBZ7bJMr7fOraJXjgyYlixLXlArtyvFqqf+ZYENu7iTefVlZhjTGiiYRburOfxZ4n5uZaPtfa1dPg5wLSdeRWN+Vtow4usBKdUoW7PJrKJPswyG0ShUSBQpJgJ6NOTEcuSeSGa07Lwcg95DrbOudLkuhIJIjoBgDHOn56V5f3cD2RoQQnoncBaAH4t+zQTgAnKKX2EdG5AL5MRM9USh0uXEipqwFcDQCbNm2as7PwsctHcWC63HBd5t20ZkkNr3rmMXjri07Rx2pRgAmUe7IUrl0iSo9bNgnbcAuYXCZgejfJa7sliXyBuYgBAJ0BVEZ0ynKcnZBnbM2PPW3dEtz+7gsLG51WN2njoFtV4MoC2w0M7yYxbqPVsCsiYcM0XLvUTcJ3Pvu53Zw4ZW0qSbR7qk59rIQBJmZbBe8mPrfVMFOIMEHiueDy0pLHbbWaK12Ga6518u5iYjXbTLLcTSY3XlZDuxMjZhM1u30sbBKMwCKu+lxDS+CWgqMwgFJZdcEOBIxtEhXDcB0YBFGqazWRLyHkC2a4VkpdUPYbEe0iovWZFLEewG5Hs+0ANorvGwDsENe4DMDPAjhfZRm+lFJ1APXs821EtBXA0wFs7vxIcwMboMskCX4x9qYcBoR/eNMm41g3OnP7GvqzYbjOvJuyYLrca6J8o8glCbMPrg1VcjpluXnSdiGasellZEgzrzsLsy13DqiKQ5IA3BNaq5uiImEBoHdPnZajR2VpWSCaK1q+G5SlimBI98VOcRIAcOLqlEjsPDRb2salwjB/D7A/TgreTWk/AgCxWefEskmUgaXZQhp9x4bqdIHVNonyfgOpJKFEwSRnPYkeJMhCgj/rezPO1cR2X9P25jPIpI0SMiYmaWOTkBjLxjISKqqQTEnCDoJsJTFcaU+A4TVcXwvgsuzzZQC+4mhzK4DTiOhkIqoCuDQ7j72efh/AzyulpvkEIlqbGbxBRKcAOA3Atj772hbMrdtpwhk8cboR6VweGe1QNkE50lpLEtlP7WwdvPmN2DYJxzm1LtRNQJF7rYRmGcazN67Am3/6JOe5vcQ0aOOpzt1knsMppGuC6+oF0n3RTInQXh9fBsPF1KXOEx4vuSqm/Hocdb79wHRpm24S8TWyOAl7zF1qupFKWChy5MJpxyzFZ379+XjR09YYx12btzuYrj2RNCUJEUznMFy7CHIZCqnCS9RN8qghVVhSS5lNgonneC3KDdcdCPoS7QKbpvt4w6YNeOHTVjttEvJzGZGcL0miXyJxFYALiehhABdm30FExxHRdQCQGaavAPBNpG6tn1NK3Zud/3cAlgK4nojuIKKPZcdfAuAuIroTwOcBvE0ptb/PvrYFB6y51AZA/mJWzAORMNRNYlKcd/IqvOn5J+LsDSuM67WTJJaN2jaJctVMtWtJwlYxmOqmdo8Z9aAaqnSQJDiQq11f24GJz1glNN7zaNVUuXQL+RpcahoZNNipfCkAPOv45aiEhMtfcmppG5cjgUTq3ZSpbKwBzPtj6vQ7XZPx06euLlzT5ZrqUilpSarkVlKSmKy3CoS7U26oMnRKFZ5Xo3O3sdVcuU3CfJDjVoziX9/6PLzymcfoczpJEqNC3UREeP/rz8amk1Y5i0UBKLWHMOyyy4NCX1dVSu0DcL7j+A4AF4vv1wG4ztHOmQNAKfUFAF/op2+9grn1iZKsqGxAXdGFSNfpZdqQ+4vkjpePVvDe1zxLf+eF1o0kYcdJuAiL3BTLDJaAW5KQm107Yyxz2N0QTNtwbV+XA8PZSNhremS+7mg1MsZwEOom17t+1vHL8ZwTVmDNkpqWAtuNw3gtwsN/frHzNy6Y1ImQsSSRelTZvxU50ZFK6LzmT5+yGj/ctq/tvQBrc23rAsvt3c/PhKreSrBnoo51y1InAa267RCsVwZb6rCJhi4oVBKrZBu0qw5pjPGi01Ipi5+107saidIszfZ6NiWJIiNXtv57cd/uBf1KEk8ZvOqZqW3+OSescP7OFdy60ftpY9scvJvaLQCX4dpGrm4ypRkX8y05yN4kCbIWT+mpPRmZq7a6yeoSL2i+Zq85zfg5xqqhsdBOymwB/aibXAv3OSesxJfe/kKMVMKOEded0MlFlZF6N6lSw7Xd1w2rRnXcBWPb+y7GP7/1vK765fIEcjEknepp5OqmGHsm6li7pGb0uRdjteu+Zee+9jkbsn7lx2Qf7dra3ahPtaqso9RHmWdTuTNAxSHt96pm7RfzI58sQrzwaWvwwHsvKs2rz+m3l3ejburVcF3CudjIo3bbqJssF9iozTlmWvPya1aF+idRadsybxAbLhfY0vuwr3nkJixMFDizbNIjlWBvk1FL3fTi09bg3370WM+ShEnc25/bThXTDUYqAQ7NtPd8AtIxbLRiVB0BjzowS/T19155eiGbbxAQ0KWQ5tq8XUPRKU6Cx362mWDPZB3rltWM4/ZmDXQnUbSLk7jvT1+l7QalhmvxLAFRxiC1Z6rYE7CTTSIKAlTDoC2RWFIrSvu9un73C08kBNoVXulFkuhV3dQpgyaDNxiXkZQxUgnx+xc9A+efsc7oi4sDlZuirOZVbJcHXE034iyCuDvC1ovh2hUnIcFEgdsldobBDsjVTaHm0JbWIk18erVJ9KIC4Z97ddtl/OXrzsJVX38AazIOuwwccR07vJtcXHklDOCa9rwpl3n76XZdShJRByLJa+/Jw7NoxkpLEpxm3fQyKp/TNux5J7/KQDwqaeN6x7Kiogtcd6aTTaISpmUJbM2AvLYM7i1bS3/8c2eWFssaBDyR6BErRrs3XHcbEVwWTWpDL8IOhtvfeFlu+HzWccvx/tef5cxJJTfFtpKEMChPN+IsLUf+e7t9L0/G17bL6X2YSDjqGgNFw3Wvxbg4UnysGmJJLcL7XvtsvPwZa/GjbfuN+3cL+d5c6SEk2gWadYOXnb4OLzvdFatqohpyMF3xXjLhYDd4/+vPwnknrWp7n14liTJ1ExNw9uxauzQnhtXI3JTtRJDtUJb1tdC/UhVTfg+ZdaHdGLI7eKdqf9WInJJEZBAJs6yv/TsAvOWFJ7e9T7/wRKJHzIvhWjRrd043huvCOQHp5GI2uo6TyDbtdNI3Cwn+2m18UQ+ShM5EWmLHYLWIJhI9Gq6JUldDNur/8vPSCH1XBt9uIJ+JYxza3Rvojlj2g0oYIE4UWklSsEPldTi6e86yeQOkc7ARuw2+7WwS5eqm9J1wUsZ1kkiEgVNi6Yaod0uUjchx8UVGdMv63O3mc73ZXTBdFAR4+rFLceraJcbxXiWJ+YYnEj2iG19kV6GUdujWKMeLflA6SdMFtp26KW23cdUY9k7WcfzKUdO411aS6H4sWIy33WaPWz6CHYdmtbqJn38utd9rUVjg+s9cvxwA8IpndObUJeRmclKWPrxT2/le4PxOZ5txSTBdb0xGGXJbV36snd2hc5xEJknsd0kSpqNBL0SiW5SpT416EKJEaTumKpckOqibogCfekvRQcBI8ll12SS84XqoMd6F7s/lRdIOPaubBjRJTBfYdobrtN1p65bg3y9/PoioVG9roxcX2BVjVaMmMHP4x60YxY5Ds3menexadgGcbrCkFhWi5p+9YTnuec+r2lalc0G+Bi7BWdq2T5tEt+CNZLaZONRNZPzvB2W1p8OAnCqdznES6Rx7nCWJZSPiN1OSIOucQcAMoMs/uySJSkRt53OrS8N1p7Tsab9M+xEwGCLfCzyR6BHd+CJXexTry/LS2MhzFw1mknQbcS3LhZKDI243Jr2om37txSfj589er7/vm2wASIkEHj2gY1V4vHr1bgKAj/3KuQaXyuiVQMh+AJ31z+1SVgwSRtGrguHaDLDsB8yo2MF0Ze+5U6pw7vf+qQZGK6HBjJ29cTnOWL9Mf9e1W+YYVPnXv3g2nn38cuMYdSFJyPxt7Ri1//WKp2FytoVLzytX1wHlgbtl76ddjMZ8whOJLkGEQrnOMrgCgNrB5MbKJ1+72hBzgfTjbrfgXF4x3aubut8cl41UjOysHNi4fkXKVbLkwP3o1bsJSKWGQWEu/vrzrW4qq4Wc/pYxGQPoQ8XhDRdS+ZzvlOAvCEgbw1eNV4359ff/37nOc9Y4iH03eP25xYoEZZKxlCR4c37uCSvbSo4rxqr4y9ef1bEfrqqPgDumCfA2iaHHj/7gfMw2ulOC95qWo1t1k50FtV+U1Riw4cqfI7s5KBfYMhyflbi0M3b26t00aMylQNF8r2+pp+8mLcdcwVywzA4cBUEpl5vbJMqvyfVNuJBXGZ553DK86+Iz8AvPPb7HXpdDdkuOm5QQ+fhf/aKZnn+uKNMIlJUc1unIvU1iOLHOqn3QDjrtcrfeTYH83IVNYkDqpigzFMeJ6ioth0xx0HOcRB9qlvXLzYhgXiS9ejfNBy565rH4ubOP69iOy0zOV+oEhmQg7Kk01zocLvAc5CyqQKrLL/Mm6lRPAki9gSbqrY5u5kSEX3/JKW3bvP7cDfj8bdvbtpEoI+Iy/mDQap5OeeJsBwudUfkI2ySOLEk6StCzJNFlUNagDddAeb1ps00xK6sRJ9GmO7yZ9NPlY5aZagW+1hDQCHzsTefiZ85a37FdQPNvjwBMSaJgkxig4fO3Lng6ABgpPcKASudvpzgJIJ+LyweQ8vqvf/FsPHLVz3Td3mVrA8yNetDvr5SgZsftADlvk3gKoecssD2qmwY5SThArpvcTdIGUGbos1FWT6K3PpqLhe83F++mhUIYUF/SVLeQKsSCd9MA1RUXP3t9YRMOiNpsfNym/JpsI1sxOj91EdqBX409l6VN4kjZAvg+41Zp1oWySXhJYh7QT6rwdhuJjpMYoLipi6W0uSb/Fgv9c1nWzOK5vcWMuGA/L7uwblzZ3u10mBCQuy74oNFOkphvF8qonSTRjbopYwa6CVgdNMpclEcWgEhwH4rqJi9JPGVQ65Hiy2jcdhGiOnfTgAzXQL4w212TOc9WIolE/r9d0JDLVbJb/MUvPBuJUgXd7Rnrl+Hjb96EFz6tmGpkWMHJ4eYblTaShDZcz1M/7IqF9m9pn8rPZxtTN6lvBg2CW+Idc7jAzjdamZ3Hrt+tq+INcP13A08k5gG915PoTiXTS8xBt+jGJsH3k9lCAyESt60nweqmObDRbzwvTZux4+BM4bcLzzym5+stJF5w6mpMN9y1SgYJQ5IocYGdL++YdoSQ10I7SWI2S2cxCJtEr+AhsbtnqJuOhCgIYDpLEDhmxe4slE2ir9lCRKuI6Hoiejj7v7Kk3UVE9CARbSGiK8XxPyGiJ7KqdHcQ0cXit3dm7R8kolf1088jjV4T/PH+3Gnzf+1zjsdfvu7Zg5UkKmna7268qqQkwetlvEMQmizkPlcc6QjT+cAFZx6Dq17X2Xe+X0ibxHwarl2IwnIi0Y26aSbbHBfEJlEiSbiC6eYd2TI71nLYWFKLUGkzxvOFfiWJKwHcqJS6Ktv8r0Ras1ojq1X9EaTlTbcDuJWIrlVK3Zc1+ZBS6q+tc85EWgv7mQCOA3ADET1dKRX32d8jAp7k3ZYTzA3S7Tf/javG8EurTuivcxZcWShtRFqSEO6OWZ/LaoLrc/uQJGQfPbqDTPlu6/YrXdif+sGvPO9E7J6oO3/rJk5itpEubzttypEAT097ntbaSGbzhZc+fS2ufPUz8CvPP9E4ful5J+DcE1cOlEnsBv0SiUsAvCz7fA2A78AiEgDOA7BFKbUNAIjos9l596EclwD4rFKqDuAnRLQlu84P++zvEcFPn7oaX3z7C3DaMUu7aj/oSOpeUIvCjhG4LkmCF1MnSaLXAkwuHGkd7GJGNcw535PXjFu/9ZYuple84GlrSn/rxgWWE+N1U0d+0MgrB5r96zazwED7EhDe9tJinfPloxVsKkndPq/96fP8Y5RSOwEg++9Ko3k8gMfF9+3ZMcYVRHQXEX1SqKs6naNBRJcT0WYi2rxnz565PsdAQUR47glOzZsT8+Ha2i1qlaA0PQCDpYHYYbiWlbPc5w5A3bQA47JYIdNl2OnLX/GMdfitC07DiR2SEc4Hugmm4+jthfBu0pJEm6Uw34GQw4qOkgQR3QDgWMdP7+ryHq6R5d3mowDem31/L4APAPgfHc4xDyp1NYCrAWDTpk2Lx3FeIM+QuRCSRPtkZUDOeZo2Cbcvt428Itnc+3ikxevFDKmaW7PE5MhXL6npILgjjW7UTYyFUDd1YzMZFD7+5k2YrDfn/T6DQkcioZS6oOw3ItpFROuVUjuJaD2A3Y5m2wHIdIgbAOzIrr1LXOvjAP6z0zlPReSR1EeeSFSjUNeMLoO2STjiJDqVTezViO/CQqjhFiukVDhMnC/Pl3bv8kVPW4Pvb9nbMaPufIB7dSSIxGLzzOvXJnEtgMsAXJX9/4qjza0ATiOikwE8gdQg/csAwAQma/daAPeI636aiD6I1HB9GoBb+uzr0KJb76b5wMuevtaoAuaCyybBXbXd9Gx0qm3sMVgMq5GfVZbtCNfH37wJh2YWhsMuS8vh0T+RuArA54jorQAeA/CLAEBExwH4hFLqYqVUi4iuAPBNACGATyql7s3Ofz8RnYNUlfQIgP8JAEqpe4noc0iN2y0A71gsnk1zwULaJF537ga8rkMbzsL6rOPznP4sGYx14PoWKpXA0QomEmdvXLGwHbHQTRbc0WqxauCRQp6WY0FuP9Toi0gopfYBON9xfAeAi8X36wBc52j3pjbX/nMAf95P/xYLFtK7qRs86/jluO43X4zTj829tTj7Zyd10yDiJDy6RxAQ/vN/vQgndiineqTRqXzpQkNnEPDztAAfcT0E6EZfu9A487hlxveZzKd9tJPhegBxEroP65d1buSBZx0/uMJKg0LYhSSxkDiShuvFBk8khgA5kRhOfbILU1mKifEOLrCDUjdtfvcFHT2pPIYXedqL4dyEy4LpPDyRGAospHfTXDHNkkRHm8RgOLQ1S+ZWqtJjODDs6iZq44b+6V9/HrbunjzSXRoaeCIxBFhI76a5YpaTkHWMk2BJYt675DHEyGt8L3BHSpC7wBZ/e8Gpa/CCU8ujyZ/qGNJXdnRhMdgkbGhJotp+CmlJYhE9m8fgwXN7WNVNi3ENHil4IjEEGHbvJheYSHSSJIgIa5bUsGbcq4uOZgy7uomX3rASsYWEVzcNARYyTmKu4OyYSztkgQWAb/zWi7tq5/HURTdxEgsJHUw3pP1bSPiVOwQIFqEk8YE3nI1r79jRlVuqNzp7DLskkSf4G87+LSQ8kRgChItQkli3dAS/9uJTFrobHosEuU1igTtSAi465NVNRXibxBAgWITeTR4evWDYDcPcLR8nUYQnEkOAcMgXkIdHv1g6EmHZSIT1y0cXuitOeHVTOby6aQjQbflSD4/FipFKiNv+8MKhVanmlekWuCNDCE8khgCL0XDt4dErhrl4lE8VXo7hfWtHERZjWg4Pj6cS8lThfg3a8ERiCBC2yRvj4eEx/wg8kSiFJxJDADZFeEnCw2NhwC6wQ6wRWzD0NSREtIqIrieih7P/K0vaXUREDxLRFiK6Uhz/dyK6I/t7hIjuyI6fREQz4reP9dPPYYf3bvLwWFh4SaIc/RqurwRwo1LqqmzzvxLA78sGRBQC+AiACwFsB3ArEV2rlLpPKfVLot0HABwSp25VSp3TZ/8WBRZjWg4Pj6cUvMq3FP0KV5cAuCb7fA2A1zjanAdgi1Jqm1KqAeCz2XkalLoWvAHAZ/rsz6JE7t3kZV0Pj4WAlyTK0e+udIxSaicAZP/XOdocD+Bx8X17dkzixQB2KaUeFsdOJqIfE9F3iejFZR0gosuJaDMRbd6zZ8/cnmIIEAbk9aEeHguEwCf4K0VHdRMR3QDgWMdP7+ryHq5hV9b3N8KUInYCOEEptY+IzgXwZSJ6plLqcOFCSl0N4GoA2LRpk33dRYNlIxGWjVQWuhseHkcldNEhr24qoCORUEpdUPYbEe0iovVKqZ1EtB7Abkez7QA2iu8bAOwQ14gA/AKAc8U96wDq2efbiGgrgKcD2Nypv4sVX3z7C7F2qc+W6uGxEAiGPEvtQqJfBce1AC7LPl8G4CuONrcCOI2ITiaiKoBLs/MYFwB4QCm1nQ8Q0drM4A0iOgXAaQC29dnXocbJa8axpOYD4D08FhI+wV8R/RKJqwBcSEQPI/VeugoAiOg4IroOAJRSLQBXAPgmgPsBfE4pda+4xqUoGqxfAuAuIroTwOcBvE0ptb/Pvnp4eHg4oYsiebtgAX2xrkqpfQDOdxzfAeBi8f06ANeVXONXHce+AOAL/fTNw8PDo1t476ZyeLrp4eFx1MOnCi+HJxIeHh5HPfIa3J5I2PBEwsPDwyODJxJFeCLh4eFx1COXJBa4I0MITyQ8PDyOenibRDk8kfDw8DjqEfgEf6XwRMLDw+OoR+4Cu7D9GEZ4IuHh4eGhE/x5KmHDEwkPD4+jHixBkCcSBXgi4eHhcdQjL1/qiYQNTyQ8PDyOegTeu6kUnkh4eHgc9SAy/3vk8ETCw8PjqAd5w3UpPJHw8PA46qHLl3p1UwGeSHh4eBz1YNLgvZuK8ETCw8PjqAcXGwo9jSigLyJBRKuI6Hoiejj7v7Kk3SeJaDcR3dPt+UT0TiLaQkQPEtGr+umnh4eHRzuwC6xPy1FEv5LElQBuVEqdBuDG7LsLnwJwUbfnE9GZSMuaPjM77++55rWHh4fHoEE6LYcnEjb6JRKXALgm+3wNgNe4GimlvgfAVaO67PxLAHxWKVVXSv0EwBYA5/XZVw8PDw8nvOG6HP0SiWOUUjsBIPu/bkDnHw/gcdFue3asACK6nIg2E9HmPXv29Hh7Dw8PDylJLGw/hhFRpwZEdAOAYx0/vWvw3clv6zimXA2VUlcDuBoANm3a5Gzj4eHh0Q6+fGk5OhIJpdQFZb8R0S4iWq+U2klE6wHs7vH+ZedvB7BRtNsAYEeP1/bw8PDoCkwaPJEool9107UALss+XwbgKwM6/1oAlxJRjYhOBnAagFv67KuHh4eHE+RtEqXol0hcBeBCInoYwIXZdxDRcUR0HTcios8A+CGA04loOxG9td35Sql7AXwOwH0AvgHgHUqpuM++enh4eDihbRKeSBTQUd3UDkqpfQDOdxzfAeBi8f2NvZyf/fbnAP68n/55eHh4dIPcJrHAHRlC+IhrDw+Pox46Vbi3SRTgiYSHh8dRjzxVuCcSNjyR8PDwOOrhDdfl8ETCw8PjqAeThtDviAX4IfHw8DjqwYZrr24qwhMJDw+Pox7kDdel8ETCw8PjqIdPy1EOTyQ8PDyOeuTBdAvbj2GEHxIPD4+jHs/ZuBL/8yWn4LknOOumHdXoK+Law8PD46mA0WqId158xkJ3YyjhJQkPDw8Pj1J4IuHh4eHhUQpPJDw8PDw8SuGJhIeHh4dHKTyR8PDw8PAohScSHh4eHh6l8ETCw8PDw6MUnkh4eHh4eJSClFIL3YeBgYj2AHi0j0usAbB3QN050vB9Xzgs5v4v5r4Di7v/w9T3E5VSa10/PKWIRL8gos1KqU0L3Y+5wPd94bCY+7+Y+w4s7v4vlr57dZOHh4eHRyk8kfDw8PDwKIUnEiauXugO9AHf94XDYu7/Yu47sLj7vyj67m0SHh4eHh6l8JKEh4eHh0cpPJHw8PDw8CiFJxIAiOgiInqQiLYQ0ZUL3Z9OIKJHiOhuIrqDiDZnx1YR0fVE9HD2f2hKbBHRJ4loNxHdI46V9peI3pm9iweJ6FUL02vdF1ff/4SInsjG/w4iulj8Nkx930hE/0VE9xPRvUT0v7Pji2Xsy/o/9ONPRCNEdAsR3Zn1/T3Z8UUx9gaUUkf1H4AQwFYApwCoArgTwJkL3a8OfX4EwBrr2PsBXJl9vhLAXy50P0XfXgLguQDu6dRfAGdm76AG4OTs3YRD1vc/AfB7jrbD1vf1AJ6bfV4K4KGsj4tl7Mv6P/TjD4AALMk+VwD8CMDzF8vYyz8vSQDnAdiilNqmlGoA+CyASxa4T3PBJQCuyT5fA+A1C9cVE0qp7wHYbx0u6+8lAD6rlKorpX4CYAvSd7QgKOl7GYat7zuVUrdnnycA3A/geCyesS/rfxmGpv8qxWT2tZL9KSySsZfwRCKddI+L79vRfiIOAxSAbxHRbUR0eXbsGKXUTiBdXADWLVjvukNZfxfL+7iCiO7K1FGsMhjavhPRSQCeg5SjXXRjb/UfWATjT0QhEd0BYDeA65VSi3LsPZFIxUIbw+4X/EKl1HMBvBrAO4joJQvdoQFiMbyPjwI4FcA5AHYC+EB2fCj7TkRLAHwBwG8ppQ63a+o4Noz9XxTjr5SKlVLnANgA4Dwielab5kPVdwlPJFKKvVF83wBgxwL1pSsopXZk/3cD+BJSsXQXEa0HgOz/7oXrYVco6+/Qvw+l1K5sA0gAfBy5WmDo+k5EFaQb7L8ppb6YHV40Y+/q/2IafwBQSh0E8B0AF2ERjT3DEwngVgCnEdHJRFQFcCmAaxe4T6UgonEiWsqfAbwSwD1I+3xZ1uwyAF9ZmB52jbL+XgvgUiKqEdHJAE4DcMsC9K8UvMgzvBbp+AND1nciIgD/COB+pdQHxU+LYuzL+r8Yxp+I1hLRiuzzKIALADyARTL2Bhbacj4MfwAuRuo5sRXAuxa6Px36egpSL4g7AdzL/QWwGsCNAB7O/q9a6L6KPn8GqVqgiZRjemu7/gJ4V/YuHgTw6iHs+78AuBvAXUgX9/oh7fuLkKos7gJwR/Z38SIa+7L+D/34AzgLwI+zPt4D4I+y44ti7OWfT8vh4eHh4VEKr27y8PDw8CiFJxIeHh4eHqXwRMLDw8PDoxSeSHh4eHh4lMITCQ8PDw+PUngi4eHh4eFRCk8kPDw8PDxK8f8DyjbiL+UbSgEAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "middle_position_first_feature = window_size // 2\n",
    "middle_position_second_feature = -1 + ( (window_size + 6) + (window_size + 6 + window_size + 1) )//2\n",
    "\n",
    "df_two_assays = df_merged.iloc[:, \n",
    "                               np.r_[middle_position_first_feature-(operative_half_window_size):\n",
    "                                     middle_position_first_feature+(operative_half_window_size)+1, \n",
    "                                     middle_position_second_feature-(operative_half_window_size):\n",
    "                                     middle_position_second_feature+(operative_half_window_size)+1,\n",
    "                                     window_size,\n",
    "                                     window_size + 1,\n",
    "                                     window_size + 2,\n",
    "                                     window_size + 3]]\n",
    "\n",
    "df_subsampled = df_two_assays\n",
    "\n",
    "# Train on all but CXCR4 and TGFBR1 chromosomes\n",
    "training_chromosomes = [\"chr\" + str(c) for c in range(1, 23, 1)]\n",
    "training_chromosomes.remove(\"chr2\") # exclude chr2 where CXCR4 lies\n",
    "training_chromosomes.remove(\"chr9\") # exclude chr2 where TGFBR1 lies\n",
    "\n",
    "df_training = df_subsampled.loc[df_subsampled['chrom'].isin(training_chromosomes)]\n",
    "xTrain = df_training.iloc[:, pairwise_multiplier*(2 * operative_half_window_size + 1):2 * (2 * operative_half_window_size + 1)] \n",
    "yTrain = df_training.iloc[:, 2 * (2 * operative_half_window_size + 1)]\n",
    "\n",
    "testing_chromosomes = [\"chr\" + str(c) for c in range(2, 23, 2)]\n",
    "df_testing = df_subsampled.loc[df_subsampled['chrom'].isin(testing_chromosomes)]\n",
    "xTest = df_testing.iloc[:, pairwise_multiplier*(2 * operative_half_window_size + 1):2 * (2 * operative_half_window_size + 1)] \n",
    "yTest = df_testing.iloc[:, 2 * (2 * operative_half_window_size + 1)]\n",
    "\n",
    "if(len(xTrain) < 10):\n",
    "    print(\"len(xTrain) = \"+str(len(xTrain)))\n",
    "\n",
    "quadratic_transform = PolynomialFeatures(degree=1, interaction_only=True).fit(xTrain)\n",
    "# xTrain = quadratic_transform.transform(xTrain)\n",
    "# xTest = quadratic_transform.transform(xTest)\n",
    "\n",
    "# The quadratic transform step adds an additional feature to the input features\n",
    "print(xTrain.shape, xTest.shape, yTrain.shape, yTest.shape)\n",
    "\n",
    "# Fit a ridge regression model\n",
    "l1_alpha = 0.005\n",
    "model = linear_model.ElasticNet(alpha=l1_alpha, l1_ratio=0.0, max_iter=1000)\n",
    "\n",
    "xTrain = np.asarray(xTrain)\n",
    "yTrain = np.asarray(yTrain)\n",
    "xTest = np.asarray(xTest)\n",
    "yTest = np.asarray(yTest)\n",
    "\n",
    "model.fit(xTrain, yTrain)\n",
    "\n",
    "# Compute statistics\n",
    "yTrain_Pred = model.predict(xTrain)\n",
    "yTest_Pred = model.predict(xTest)\n",
    "mse_Train = mean_squared_error(yTrain, yTrain_Pred)\n",
    "pc_Train, _ = pearsonr(yTrain, yTrain_Pred)\n",
    "sc_Train, _ = spearmanr(yTrain, yTrain_Pred)\n",
    "mse_Test = mean_squared_error(yTest, yTest_Pred)\n",
    "pc_Test, _ = pearsonr(yTest, yTest_Pred)\n",
    "sc_Test, _ = spearmanr(yTest, yTest_Pred)\n",
    "\n",
    "print(\"Train: MSE = \", round(mse_Train, 3), \"Pearson =\", round(pc_Train, 3), \"Spearman =\", round(sc_Train, 3))\n",
    "print(\"Test: MSE = \", round(mse_Test, 3), \"Pearson =\", round(pc_Test, 3), \"Spearman =\", round(sc_Test, 3))\n",
    "\n",
    "plt.plot(model.coef_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "ad8d4934",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "../../Data/Gene_Expression_Data/T01.tsv.TPM.headered doesn't exist!\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '../../Data/Gene_Expression_Data/T01.tsv.TPM.headered'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/scratch/4164177.1.0_gpu_t4_/ipykernel_91444/3823448495.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0mcell_type_choice\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0mpath_to_save\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"../../Data/\"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m\"linear_model\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m \u001b[0mxInference\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0myInference\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgene_list\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mCHROM\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mTSS\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSTRAND\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgenerate_data_vectors\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcell_type_choice\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m2\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0moperative_half_window_size\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpath_to_save\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     13\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0mpeak_width_choices\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;36m6\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m8\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.chrmt/Code/deepENCODE/chrmt_inference.py\u001b[0m in \u001b[0;36mgenerate_data_vectors\u001b[0;34m(cell_type_choice, window_size, path_to_save)\u001b[0m\n\u001b[1;32m     37\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mgene\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgene_list\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     38\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 39\u001b[0;31m         prediction_generator = TranscriptomePredictor(window_size,\n\u001b[0m\u001b[1;32m     40\u001b[0m                                \u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     41\u001b[0m                                \u001b[0mshuffle\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.chrmt/Code/deepENCODE/chrmt_generator.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, window_size, batch_size, shuffle, mode, masking_probability, chrom, start, strand, cell_type)\u001b[0m\n\u001b[1;32m    594\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcell_type\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcell_type\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    595\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 596\u001b[0;31m         EpigenomeGenerator.__init__(self, self.window_size, self.batch_size,\n\u001b[0m\u001b[1;32m    597\u001b[0m                                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshuffle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmode\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    598\u001b[0m                                     self.masking_probability)\n",
      "\u001b[0;32m~/.chrmt/Code/deepENCODE/chrmt_generator.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, window_size, batch_size, shuffle, mode, masking_probability)\u001b[0m\n\u001b[1;32m    142\u001b[0m             \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mTSS_DATA\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"doesn't exist!\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfile\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstderr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    143\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 144\u001b[0;31m         \u001b[0mf_TSS\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mTSS_DATA\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'r'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    145\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Loading TSS data\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfile\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstderr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    146\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '../../Data/Gene_Expression_Data/T01.tsv.TPM.headered'"
     ]
    }
   ],
   "source": [
    "sys.path.insert(1, '/home/sbatra/.chrmt/Code/deepENCODE/')\n",
    "from chrmt_inference import generate_data_vectors\n",
    "from chrmt_inference import ise_pairwise_input\n",
    "from chrmt_inference import perform_ise\n",
    "\n",
    "# use above ridge regression model\n",
    "trained_model = model\n",
    "\n",
    "# Generate data vectors for CXCR4 and TGFBR1\n",
    "cell_type_choice = -1\n",
    "path_to_save = \"../../Data/\" + \"linear_model\"\n",
    "xInference, yInference, gene_list, CHROM, TSS, STRAND = generate_data_vectors(cell_type_choice, 2 * operative_half_window_size + 1, path_to_save)\n",
    "\n",
    "peak_width_choices = [6, 8]\n",
    "inserted_lnp1_minuslog10_p_value_choices = [1.5]  # corresponds to 0.0003\n",
    "fig, axs = plt.subplots(len(peak_width_choices) * len(inserted_lnp1_minuslog10_p_value_choices), 2)\n",
    "\n",
    "fig = perform_ise(fig, axs,\n",
    "            \"../../../Data/p300_epigenome_editing_dataset.tsv\", ise_pairwise_input,\n",
    "            peak_width_choices,\n",
    "            inserted_lnp1_minuslog10_p_value_choices,\n",
    "            xInference, yInference,\n",
    "            gene_list, CHROM, TSS, STRAND,                \n",
    "            trained_model, run_name)\n",
    "\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "333ad3aa",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# The goal is to perform in-silico perturbation using a model trained to predict bulk gene expression from \n",
    "# epigenetic features, on CRISPRa data\n",
    "\n",
    "cell_type_choice = 13 # corresponds to HEK293T\n",
    "assay_index = assay_index_choice\n",
    "\n",
    "inserted_pvalue_choice = 1.5 # corresponds to 0.0003\n",
    "for peak_width_choice in range(6, 7, 2):\n",
    "    \n",
    "    for MNase_offset in range(-15, 15):\n",
    "#         print(\"peak_width = \", peak_width_choice)\n",
    "        TPM = {}\n",
    "\n",
    "        # Perform inference for epigenome editing data\n",
    "        def ise(cell_type_choice_input, \n",
    "                gene,\n",
    "                gene_strand,\n",
    "                gRNA_strand_dict,\n",
    "                trained_model,\n",
    "                assay_index_value,\n",
    "                inserted_lnp1_minuslog10p_value = 3,\n",
    "                peak_width = 2,\n",
    "                pairwise_features=False):\n",
    "\n",
    "            # Load epigenetic features for all assays\n",
    "            X = np.load(\"../Data/\" + gene + \".T\" + '%02d' % cell_type_choice_input + \".npy\")\n",
    "\n",
    "            # obtain the middle portion of this\n",
    "            X = X[:, (window_size//2)-operative_half_window_size:(window_size//2)+operative_half_window_size+1, :]\n",
    "\n",
    "            # Perform inference by introducing p-value changes with a peak width\n",
    "            yPred = []\n",
    "            center = operative_half_window_size\n",
    "            positions = range(center - center, center + center + 1)\n",
    "            for pos in positions:\n",
    "                X_modified = np.copy(X)\n",
    "\n",
    "                # some positions won't be used in the plot and hence we don't perturb them\n",
    "                if(pos not in gRNA_strand_dict):\n",
    "                    gRNA_strand = \"plus\" # Change 1\n",
    "                else:\n",
    "                    gRNA_strand = gRNA_strand_dict[pos]\n",
    "\n",
    "                ise_start_position = pos - (peak_width // 2)\n",
    "                ise_end_position = pos + (peak_width // 2 + 1) \n",
    "\n",
    "                for p in range(ise_start_position, ise_end_position):\n",
    "                    if( (p>=0) and (p < max(positions)) ):\n",
    "                        if(X_modified[:, p, 2] > 10): # Remember this is ln( -log10(p-value) + 1)\n",
    "                            # If H3K27me3 peak exists, then p300 doesn't work\n",
    "                            print(\"H3K27me3 exists!\")\n",
    "                            pass\n",
    "                        else:\n",
    "                            if (pairwise_features):\n",
    "                                # Modify the H3K27ac peak with the shape being derived from MNase (-1) \n",
    "                                if(gRNA_strand == \"plus\"):\n",
    "                                    X_modified[:, p, 3] = X_modified[:, p, 3] + (X_modified[:, min(p+MNase_offset, 2*operative_half_window_size), -1] * inserted_lnp1_minuslog10p_value)\n",
    "#                                     print(\"Delta = \", gene, peak_width_choice, gRNA_strand, pos, p, (X_modified[:, min(p+MNase_offset, 2*operative_half_window_size), -1]), file=sys.stderr)\n",
    "                                else: #(gRNA_strand == \"minus\"):\n",
    "                                    X_modified[:, p, 3] = X_modified[:, p, 3] + (X_modified[:, min(p-MNase_offset, 2*operative_half_window_size), -1] * inserted_lnp1_minuslog10p_value)\n",
    "#                                     print(\"Delta = \", gene, peak_width_choice, gRNA_strand, pos, p, (X_modified[:, min(p-MNase_offset, 2*operative_half_window_size), -1]), file=sys.stderr)\n",
    "#                                 else:\n",
    "#                                     pass # Change 2\n",
    "                            else:\n",
    "                                # Modify the assay itself\n",
    "                                X_modified[:, p, assay_index_value] += inserted_lnp1_minuslog10p_value\n",
    "\n",
    "                # Prepare input\n",
    "                if (pairwise_features):\n",
    "                    X_modified = np.concatenate([X_modified[:, :, 3], X_modified[:, :, assay_index_value]], axis=1)\n",
    "                else:                 \n",
    "                    X_modified = X_modified[:, :, assay_index_value]\n",
    "\n",
    "                yPred_value = trained_model.predict(quadratic_transform.transform(X_modified))\n",
    "                yy = yPred_value[0]\n",
    "                yPred.append(yy)\n",
    "\n",
    "            # Prepare input for predicting native expression\n",
    "            X_modified_2 = np.copy(X)\n",
    "            if(pairwise_features):\n",
    "                X_modified_2 = np.concatenate([X[:, :, 3], X[:, :, assay_index_value]], axis=1) \n",
    "            else:\n",
    "                X_modified_2 = X[:, :, assay_index_value]\n",
    "\n",
    "            # Instead of scaling, divide by yPred\n",
    "            yPred_value = trained_model.predict(quadratic_transform.transform(X_modified_2))[0] + 0.00000001 # to avoid divby0\n",
    "\n",
    "        #             print(\"Predicted TPM for \", gene, \" = \", yPred_TPM)\n",
    "        #             print((np.power(10, yPred) -1), (np.power(10, yPred_value) -1))\n",
    "\n",
    "            yPred = (np.power(10, yPred) -1) / (np.power(10, yPred_value) -1)\n",
    "            yPred_TPM = (np.power(10, yPred_value) -1)\n",
    "            return yPred, yPred_TPM\n",
    "\n",
    "\n",
    "        def p_value_mapping(inserted_lnp1_minuslog10p_value):\n",
    "            minuslog10p_value = np.expm1(inserted_lnp1_minuslog10p_value)\n",
    "            p_value = np.power(10, -1 * minuslog10p_value)\n",
    "            return round(minuslog10p_value, 4)\n",
    "\n",
    "\n",
    "        def convert_to_2D(idx, nrows, ncols):\n",
    "            return idx//ncols, idx%ncols\n",
    "\n",
    "\n",
    "        # Now that we have the helper functions declared, we perform the in silico perturbation\n",
    "\n",
    "        # Load p300 epigenome editing data\n",
    "        df_p300_epigenome_editing = pd.read_csv(\"../Data/p300_epigenome_editing_dataset.tsv\", sep=\"\\t\")\n",
    "\n",
    "        # we want to compute \n",
    "        TSS = {}\n",
    "        STRANDS = {}\n",
    "        CHROMS = {}\n",
    "        GENES = {}\n",
    "        gRNA_STRANDS = {}\n",
    "        for index in range(len(df_p300_epigenome_editing)):\n",
    "            tss = df_p300_epigenome_editing.iloc[index, 13]\n",
    "            gene_strand = df_p300_epigenome_editing.iloc[index, 3]\n",
    "            chrom = df_p300_epigenome_editing.iloc[index, 2]\n",
    "            gene = df_p300_epigenome_editing.iloc[index, 0]\n",
    "\n",
    "            TSS[gene] = int(tss)\n",
    "\n",
    "            if(gene_strand == \"plus\"):\n",
    "                STRANDS[gene] = \"+\"\n",
    "            elif(gene_strand == \"minus\"):\n",
    "                STRANDS[gene] = \"-\"\n",
    "            else:\n",
    "                print(\"something wrong with strand!\")\n",
    "\n",
    "            CHROMS[gene] = chrom\n",
    "            GENES[gene] = 1                                                    \n",
    "\n",
    "        GENES_LIST = set(list(GENES.keys()))    \n",
    "\n",
    "        df_GENES_values = {}\n",
    "        df_GENES_means = {}\n",
    "        gene_strand_dict = {}\n",
    "        for gene in GENES_LIST:\n",
    "            df_GENES_values[gene] = df_p300_epigenome_editing[df_p300_epigenome_editing[\"p300 target gene\"] == gene]\n",
    "\n",
    "            df_GENES_values[gene][\"Position_wrt_TSS\"] = ( (pd.to_numeric(df_GENES_values[gene][\"gRNA position  wrt TSS (hg38)\"])) / RESOLUTION ) # Change 3\n",
    "\n",
    "            df_GENES_values[gene][\"gRNA_strand\"] = df_GENES_values[gene][\"gRNA target strand\"].apply(lambda s: \"red\" if(s==\"minus\") else \"blue\")\n",
    "            df_GENES_values[gene][\"gRNA_strand_bool\"] = df_GENES_values[gene][\"gRNA target strand\"].apply(lambda s: -1 if(s==\"minus\") else +1)\n",
    "\n",
    "            gene_strand_dict[gene] = STRANDS[gene]\n",
    "        #     print(gene_strand_dict)\n",
    "\n",
    "            df_GENES_means[gene] = df_GENES_values[gene].groupby('gRNA ID').mean()\n",
    "            df_GENES_means[gene].index.name = 'gRNA ID'\n",
    "            df_GENES_means[gene].reset_index(inplace=True)\n",
    "#             print(gene, df_GENES_means[gene])\n",
    "\n",
    "        # Perform in-silico epigenesis\n",
    "        assay_color = ['black', 'red', 'green', 'blue', 'cyan', 'pink', 'brown']\n",
    "        xticklabels = range(-operative_half_window_size, operative_half_window_size + 1)\n",
    "\n",
    "        GENES_LIST = [\"CXCR4\", \"TGFBR1\"]\n",
    "\n",
    "        fig, axes = plt.subplots(nrows=len(GENES_LIST), ncols=2, figsize=(40, 30), sharey=False)\n",
    "        fig.tight_layout(pad=1, w_pad=20, h_pad=25)\n",
    "\n",
    "        for idx, gene in enumerate(sorted(GENES_LIST)):\n",
    "            TPM[gene] = {}\n",
    "\n",
    "            idx_x, idx_y = convert_to_2D(idx, nrows=len(GENES_LIST), ncols=1)\n",
    "            ax_1 = axes[idx_x, 0]\n",
    "            ax_2 = axes[idx_x, 1]\n",
    "\n",
    "            gene_features = np.squeeze(np.load(\"../Data/\" + gene + \".T\" + '%02d' % cell_type_choice + \".npy\"), axis=0)\n",
    "            gene_features = gene_features[(window_size//2)-operative_half_window_size:(window_size//2)+operative_half_window_size+1, :]\n",
    "\n",
    "            df_values = df_GENES_values[gene]\n",
    "            df_means = df_GENES_means[gene]\n",
    "            # print(gene, df_values) \n",
    "            # print(gene, df_means)\n",
    "\n",
    "            inserted_lnp1_minuslog10p_value = inserted_pvalue_choice\n",
    "            peak_width = peak_width_choice\n",
    "\n",
    "            # Compute gRNA strands for all positions\n",
    "            gRNA_strand_dict = {}\n",
    "            for p_idx in range(len(df_means)):\n",
    "                position_m = int( (df_means.iloc[p_idx, 9] + operative_half_window_size) )\n",
    "                gRNA_strand = df_means.iloc[p_idx, 10]\n",
    "                if(gRNA_strand == +1):\n",
    "                    gRNA_strand_dict[position_m] = \"plus\"\n",
    "                elif(gRNA_strand == -1):\n",
    "                    gRNA_strand_dict[position_m] = \"minus\"\n",
    "                else:\n",
    "                    print(\"Something wrong with parsing gRNA strand\", file=sys.stderr)\n",
    "\n",
    "            # Pass assay_index as -1 to make sure we select MNase\n",
    "            gene_ise, yPred_TPM = ise(cell_type_choice, gene, gene_strand_dict[gene], gRNA_strand_dict, model, 1*assay_index - 1, inserted_lnp1_minuslog10p_value, peak_width, pairwise_flag)\n",
    "\n",
    "            # Create a scatter plot of the means with the predictions of those positions\n",
    "            gene_ise_at_means = []\n",
    "            alan_means = []\n",
    "            gRNA_strands = []\n",
    "\n",
    "            for p_idx in range(len(df_means)):\n",
    "                position_m = df_means.iloc[p_idx, 9]\n",
    "                alan_mean = df_means.iloc[p_idx, 1]\n",
    "                gRNA_strand = df_means.iloc[p_idx, 10]\n",
    "                if(position_m + operative_half_window_size < 0):\n",
    "                    continue\n",
    "                elif( (int(position_m) + operative_half_window_size) >= len(gene_ise) ):\n",
    "                    continue\n",
    "                else:\n",
    "        #             print(len(gene_ise), position_m, operative_half_window_size)\n",
    "                    gene_ise_at_means.append(gene_ise[int(position_m) + operative_half_window_size])\n",
    "                    alan_means.append(alan_mean)\n",
    "                    gRNA_strands.append(gRNA_strand)\n",
    "\n",
    "            pc, pp = pearsonr(list(alan_means), gene_ise_at_means)\n",
    "            sc, sp = spearmanr(list(alan_means), gene_ise_at_means)\n",
    "\n",
    "            print(gene, sc, sp, MNase_offset, file=sys.stderr)\n",
    "\n",
    "            gRNA_strands_colors = []\n",
    "            for iii in gRNA_strands:\n",
    "                if(iii == -1):\n",
    "                    gRNA_strands_colors.append(\"red\")\n",
    "                else:\n",
    "                    gRNA_strands_colors.append(\"blue\")\n",
    "\n",
    "            ax_1.scatter(list(alan_means), gene_ise_at_means, color=\"#FF1493\", s=1000) # =\"#FF1493\")\n",
    "            ax_1.set_xlim(0, 1.1 * max(alan_means))\n",
    "            ax_1.set_ylim(0, 1.1 * max(gene_ise_at_means))\n",
    "            ax_1.tick_params(axis='both', which='major', labelsize=40)\n",
    "            ax_1.tick_params(axis='both', which='minor', labelsize=40)\n",
    "            ax_1.set_xlabel(\"Mean experimental fold change\", size=60)\n",
    "            ax_1.set_ylabel(\"Model prediction's fold change\", size=45)\n",
    "        #     if(pp > 0.05):\n",
    "        #         p_asterisk = \"NS\"\n",
    "        #     if(sp > 0.05):\n",
    "        #         s_asterisk = \"NS\"\n",
    "            ax_1.set_title(\"Correlation between experimental and model predictions fold change\\nPearson = \"+\n",
    "                           str(round(pc, 2))+\n",
    "                           \" (\"+str(round(pp, 3))+\n",
    "                           \") Spearman = \"+\n",
    "                           str(round(sc, 2))+\n",
    "                           \" (\"+str(round(sp, 3))+\n",
    "                           \")\", size=40)\n",
    "\n",
    "            # Determine whether we are doing H3K27ac ISE in the background of another assay's features\n",
    "            # or we have marginal features and are doing that track's ISE\n",
    "            if(pairwise_flag):\n",
    "                epigenetic_features = gene_features[:, -1] # MNase\n",
    "                epigenetic_features_2 = gene_features[:, 3] # H3K27ac\n",
    "                color_for_assay = assay_color[3]\n",
    "                label_for_assay = assays[3]\n",
    "            else:\n",
    "                epigenetic_features = gene_features[:, assay_index-1]\n",
    "                color_for_assay = assay_color[assay_index-1]\n",
    "                label_for_assay = assays[assay_index-1]\n",
    "\n",
    "            # Scale the model predictions     \n",
    "            scaling_ratio = np.median(df_means['Measured fold change'])/np.median(gene_ise - 0.0)\n",
    "            scaled_model_predictions = 0.5 * (scaling_ratio * (gene_ise - 0.0)) + 1.0\n",
    "\n",
    "            # Scale the epigenetic features\n",
    "            epigenetic_features_scaling_ratio = max(df_means['Measured fold change'])/max(epigenetic_features - 0.0)\n",
    "            scaled_epigenetic_features = (epigenetic_features_scaling_ratio * (epigenetic_features - 0.0)) + 0.5\n",
    "\n",
    "            epigenetic_features_scaling_ratio_2 = max(df_means['Measured fold change'])/max(epigenetic_features_2 - 0.0)\n",
    "            scaled_epigenetic_features_2 = (epigenetic_features_scaling_ratio_2 * (epigenetic_features_2 - 0.0)) + 0.0 \n",
    "\n",
    "            ax_2.plot(xticklabels, scaled_model_predictions, 'o-', color=\"#4daf4a\", linewidth=5, markersize=2, label=\"(Scaled) Model Predictions \" + label_for_assay)\n",
    "            ax_2.plot(xticklabels, scaled_epigenetic_features, 'o-', color=\"#8470FF\", linewidth=5, markersize=1, label=\"(Scaled) Epigenetic Features MNase\") # + label_for_assay)\n",
    "            ax_2.plot(xticklabels, scaled_epigenetic_features_2, 'o-', color=\"darkblue\", linewidth=5, markersize=1, label=\"(Scaled) Epigenetic Features H3K27ac\") # + label_for_assay)\n",
    "\n",
    "            ax_2.bar(df_means['Position_wrt_TSS'], 0.0 + (df_means['Measured fold change']), color=\"#f781bf\", bottom=0, width=2, label=\"Experimental mean from qPCR\")\n",
    "\n",
    "            gRNA_strand_groups = df_values.groupby(\"gRNA target strand\")\n",
    "\n",
    "            color_dict = {}\n",
    "            color_dict[\"+\"] = {\"plus\":\"blue\", \"minus\":\"red\"}\n",
    "            color_dict[\"-\"] = {\"plus\":\"red\", \"minus\":\"blue\"}\n",
    "            direction = {\"blue\":\"rightward\", \"red\":\"leftward\"}\n",
    "            color_index = 0\n",
    "            for name, group in gRNA_strand_groups:\n",
    "                ax_2.plot(group['Position_wrt_TSS'], 0.0 + (group['Measured fold change']), 'o', color=color_dict[STRANDS[gene]][name], label=\"qPCR facing \"+direction[color_dict[STRANDS[gene]][name]], markersize=15)\n",
    "                color_index += 1\n",
    "\n",
    "            ax_2.set_xlim(-operative_half_window_size-10, operative_half_window_size+10)\n",
    "            ax_2.set_ylim(-1, 1.0 + max(df_means['Measured fold change'])*1.5)\n",
    "            x_vals = ax_2.get_xticks()\n",
    "            ax_2.set_xticklabels(['{:3.0f}'.format(x * RESOLUTION) for x in x_vals])\n",
    "            ax_2.yaxis.set_major_locator(MaxNLocator(integer=True))\n",
    "            ax_2.tick_params(axis='both', which='major', labelsize=35)\n",
    "            ax_2.tick_params(axis='both', which='minor', labelsize=35)\n",
    "            ax_2.set_xlabel(\"Peak Position (in bp) w.r.t TSS\", size=50)\n",
    "            ax_2.set_ylabel(\"Gene expression fold change\", size=50)\n",
    "            ax_2.set_title(gene+\" with H3K27ac + \"+assays[assay_index-1]+\"\\nincreasing \" + str(peak_width * RESOLUTION) + \"bp peaks by -log10(p_value)=\"+str(p_value_mapping(inserted_lnp1_minuslog10p_value)), size=40) #, y=1.1)\n",
    "\n",
    "            ax_2.legend(loc='upper center', prop={'size': 30}, ncol=2)\n",
    "\n",
    "            # Now also keep storing the actual TPM vs predicted TPM for each gene, for each cell type\n",
    "            actual_TPM = np.load(\"../Data/\" + gene + \".T\" + '%02d' % cell_type_choice + \".TPM.npy\")\n",
    "        #             print(gene, cell_type_choice, (np.power(10, actual_TPM[0][0]) -1), yPred_TPM)\n",
    "            TPM[gene][cell_type_choice] = ((np.power(10, actual_TPM[0][0]) -1), yPred_TPM)\n",
    "\n",
    "#         plt.show()\n",
    "        plt.close()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d68a73fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot spearman as a function of MNase offset\n",
    "plt.rcParams[\"figure.figsize\"] = (20, 10)\n",
    "colnames = [\"gene\", \"sc\", \"sp\", \"offset\"] \n",
    "df_offset = pd.read_csv('../Data/offset_Spearman.plus_plus', sep=\" \", names=colnames, low_memory=False)\n",
    "df_offset_sorted = df_offset.sort_values(['gene', 'offset'], ascending=[1, 1])\n",
    "df_offset_sorted\n",
    "plt.plot(df_offset_sorted.iloc[1:29, 3], df_offset_sorted.iloc[1:29, 1],  'o-', color=\"#4daf4a\", linewidth=5, markersize=10, label=\"CXCR4\")\n",
    "plt.plot(df_offset_sorted.iloc[30:, 3], df_offset_sorted.iloc[30:, 1],  'o-', color=\"#8470FF\", linewidth=5, markersize=10, label=\"TGFBR1\")\n",
    "plt.xlabel(\"Offset / 25bp \", fontsize=25)\n",
    "plt.ylabel(\"Spearman of linear model's predictions with Alan's data\", fontsize=20)\n",
    "plt.xlim(-16, 16)\n",
    "plt.ylim(-0.1, 1)\n",
    "plt.xticks(fontsize=15, rotation=0)\n",
    "plt.yticks(fontsize=15, rotation=0)\n",
    "plt.title(\"(strand-specific) Offset to extract MNase-seq for perturbation\", fontsize=35)\n",
    "plt.legend(loc='upper right', prop={'size': 30}, ncol=2)\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05772279",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We analyze the HER2 gene dataset here:\n",
    "colnames = [\"chrom\", \"gRNA_start\", \"gRNA_end\", \"gRNA_ID\", \"average_log2(fold_change)\", \"d_TSS\"] \n",
    "df_HER2_DHS = pd.read_csv('../Data/HER2.DHS.hg38.tsv', sep=\"\\t\", names=colnames, low_memory=False)\n",
    "df_HER2_sat = pd.read_csv('../Data/HER2.sat.hg38.tsv', sep=\"\\t\", names=colnames, low_memory=False)\n",
    "plt.rcParams[\"figure.figsize\"] = (20, 10)\n",
    "plt.plot(df_HER2_sat['d_TSS'], df_HER2_sat['average_log2(fold_change)'], 'o', color=\"skyblue\", label=\"sat (\"+str(len(df_HER2_sat))+\" gRNAs)\")\n",
    "plt.plot(df_HER2_DHS['d_TSS'], df_HER2_DHS['average_log2(fold_change)'], 'o', color=\"red\", label=\"DHS (\"+str(len(df_HER2_DHS))+\" gRNAs)\")\n",
    "plt.xlabel(\"distance of gRNA to TSS in bp\", fontsize=25)\n",
    "plt.ylabel(\"average(log2(fold-change)) between high and low expression\", fontsize=20)\n",
    "plt.xlim(-2100, 2100)\n",
    "plt.ylim(-4, 4)\n",
    "plt.xticks(fontsize=15, rotation=0)\n",
    "plt.yticks(fontsize=15, rotation=0)\n",
    "plt.title(\"HER2 NBT17\", fontsize=35)\n",
    "plt.legend(loc='upper center', prop={'size': 30}, ncol=2, markerscale=2)\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "329acdb2",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Perform perturbation analysis in HER2\n",
    "df_HER2_p300_epigenome_editing = pd.read_csv(\"../Data/HER2.p300_epigenome_editing_dataset.tsv\", sep=\"\\t\")\n",
    "\n",
    "def her2_ise(position, cell_type_choice, gene_features, gRNA_strand, trained_model, assay_index_value, inserted_lnp1_minuslog10p_value, peak_width):\n",
    "\n",
    "    # Load epigenetic features for all assays\n",
    "    X = gene_features\n",
    "\n",
    "    # obtain the middle portion of this\n",
    "    X = X[:, (window_size//2)-operative_half_window_size:(window_size//2)+operative_half_window_size+1, :]\n",
    "\n",
    "    # Perform inference by introducing p-value changes with a peak width\n",
    "    yPred = []\n",
    "\n",
    "    positions = [position + operative_half_window_size]\n",
    "    for pos in positions:\n",
    "        X_modified = np.copy(X)\n",
    "\n",
    "        ise_start_position = max(0, pos - (peak_width // 2))\n",
    "        ise_end_position = min(pos + (peak_width // 2 + 1), 2*operative_half_window_size)\n",
    "\n",
    "        for p in range(ise_start_position, ise_end_position):\n",
    "            if( (p>=0) and (p < max(positions)) ):\n",
    "                if(X_modified[:, p, 2] > 10): # Remember this is ln( -log10(p-value) + 1)\n",
    "                    # If H3K27me3 peak exists, then p300 doesn't work\n",
    "                    print(\"H3K27me3 exists!\")\n",
    "                    pass\n",
    "                else:\n",
    "                    # Modify the H3K27ac peak with the shape being derived from MNase (-1) \n",
    "                    if(gRNA_strand == \"plus\"):\n",
    "                        X_modified[:, p, 3] = X_modified[:, p, 3] + (X_modified[:, min(p+MNase_offset, 2*operative_half_window_size), -1] * inserted_lnp1_minuslog10p_value)\n",
    "#                         print(\"Delta = \", gene, peak_width_choice, gRNA_strand, pos, p, (X_modified[:, min(p+MNase_offset, 2*operative_half_window_size), -1]), file=sys.stderr)\n",
    "                    elif(gRNA_strand == \"minus\"):\n",
    "                        X_modified[:, p, 3] = X_modified[:, p, 3] + (X_modified[:, min(p-MNase_offset, 2*operative_half_window_size), -1] * inserted_lnp1_minuslog10p_value)\n",
    "#                         print(\"Delta = \", gene, peak_width_choice, gRNA_strand, pos, p, (X_modified[:, min(p-MNase_offset, 2*operative_half_window_size), -1]), file=sys.stderr)\n",
    "                    else:\n",
    "                        pass # Change 2\n",
    "\n",
    "        # Prepare input\n",
    "        X_modified = np.concatenate([X_modified[:, :, 3], X_modified[:, :, assay_index_value]], axis=1)\n",
    "\n",
    "        yPred_value = trained_model.predict(quadratic_transform.transform(X_modified))\n",
    "        yy = yPred_value[0]\n",
    "        yPred.append(yy)\n",
    "\n",
    "    # Prepare input for predicting native expression\n",
    "    X_modified_2 = np.copy(X)\n",
    "    X_modified_2 = np.concatenate([X[:, :, 3], X[:, :, assay_index_value]], axis=1) \n",
    "    \n",
    "    # Instead of scaling, divide by yPred\n",
    "    yPred_value = trained_model.predict(quadratic_transform.transform(X_modified_2))[0] + 0.00000001 # to avoid divby0\n",
    "\n",
    "#             print(\"Predicted TPM for \", gene, \" = \", yPred_TPM)\n",
    "#             print((np.power(10, yPred) -1), (np.power(10, yPred_value) -1))\n",
    "\n",
    "    yPred = (np.power(10, yPred) -1) / (np.power(10, yPred_value) -1)\n",
    "    yPred_TPM = (np.power(10, yPred_value) -1)\n",
    "    \n",
    "    assert(len(yPred) == 1)\n",
    "    \n",
    "    return yPred[0]/yPred_TPM\n",
    "\n",
    "\n",
    "cell_type_choice = 13 # corresponds to HEK293T\n",
    "assay_index = assay_index_choice\n",
    "\n",
    "inserted_pvalue_choice = 1.5 # corresponds to 0.0003\n",
    "peak_width_choice = 6\n",
    "\n",
    "for MNase_offset in range(-15,15):\n",
    "\n",
    "    # Perform in-silico epigenesis\n",
    "    assay_color = ['black', 'red', 'green', 'blue', 'cyan', 'pink', 'brown']\n",
    "    xticklabels = range(-operative_half_window_size, operative_half_window_size + 1)\n",
    "\n",
    "    GENES_LIST = [\"HER2\", \"HER2\"] # [\"CXCR4\", \"TGFBR1\"]\n",
    "\n",
    "    fig, axes = plt.subplots(nrows=len(GENES_LIST), ncols=2, figsize=(40, 30), sharey=False)\n",
    "    fig.tight_layout(pad=1, w_pad=20, h_pad=25)\n",
    "\n",
    "    for idx, gene in enumerate(sorted(GENES_LIST)):\n",
    "        TPM[gene] = {}\n",
    "\n",
    "        idx_x, idx_y = convert_to_2D(idx, nrows=len(GENES_LIST), ncols=1)\n",
    "        ax_1 = axes[idx_x, 0]\n",
    "        ax_2 = axes[idx_x, 1]\n",
    "        \n",
    "    # we want to compute \n",
    "    TSS = {}\n",
    "    STRANDS = {}\n",
    "    CHROMS = {}\n",
    "    GENES = {}\n",
    "    gRNA_STRANDS = {}\n",
    "    gRNA_strands_colors = []\n",
    "        \n",
    "    measured_fold_change_list = []\n",
    "    predicted_fold_change_list = []\n",
    "    \n",
    "    for index in range(len(df_HER2_p300_epigenome_editing)):\n",
    "        tss = df_HER2_p300_epigenome_editing.iloc[index, 13]\n",
    "        gene_strand = df_HER2_p300_epigenome_editing.iloc[index, 3]\n",
    "        chrom = df_HER2_p300_epigenome_editing.iloc[index, 2]\n",
    "        gene = df_HER2_p300_epigenome_editing.iloc[index, 0]\n",
    "\n",
    "        if(gene != \"HER2\"):\n",
    "            continue\n",
    "        \n",
    "        TSS[gene] = int(tss)\n",
    "\n",
    "        if(gene_strand == \"plus\"):\n",
    "            STRANDS[gene] = \"+\"\n",
    "        elif(gene_strand == \"minus\"):\n",
    "            STRANDS[gene] = \"-\"\n",
    "        else:\n",
    "            print(\"something wrong with strand!\")\n",
    "\n",
    "        CHROMS[gene] = chrom\n",
    "        GENES[gene] = 1                                                    \n",
    "\n",
    "        \n",
    "        \n",
    "        inserted_lnp1_minuslog10p_value = inserted_pvalue_choice\n",
    "        peak_width = peak_width_choice\n",
    "\n",
    "        gRNA_strand = df_HER2_p300_epigenome_editing.iloc[index, 8]\n",
    "        position = pd.to_numeric(df_HER2_p300_epigenome_editing.iloc[index, 15]) // RESOLUTION\n",
    "                \n",
    "        gene_features = np.load(\"../Data/\" + gene + \".T\" + '%02d' % cell_type_choice + \".npy\")\n",
    "\n",
    "        predicted_fold_change = her2_ise(position, cell_type_choice, gene_features, gRNA_strand, model, assay_index - 1, inserted_lnp1_minuslog10p_value, peak_width)\n",
    "        predicted_fold_change_list.append(predicted_fold_change)\n",
    "        \n",
    "        measured_fold_change = df_HER2_p300_epigenome_editing.iloc[index, 7]\n",
    "        measured_fold_change_list.append(measured_fold_change)\n",
    "        \n",
    "    measured_fold_change_list = np.asarray(measured_fold_change_list)    \n",
    "    predicted_fold_change_list = np.asarray(predicted_fold_change_list)    \n",
    "        \n",
    "    # Create a scatter plot of the means with the predictions of those positions\n",
    "    pc, pp = pearsonr(measured_fold_change_list, predicted_fold_change_list)\n",
    "    sc, sp = spearmanr(measured_fold_change_list, predicted_fold_change_list)\n",
    "\n",
    "    print(gene, sc, sp, MNase_offset, file=sys.stderr)\n",
    "\n",
    "    gRNA_strands_colors.append(\"blue\")\n",
    "\n",
    "    ax_1.scatter(measured_fold_change_list, predicted_fold_change_list, color=\"#FF1493\", s=100) # =\"#FF1493\")\n",
    "    ax_1.set_xlim(-4, 1.1 * max(measured_fold_change_list))\n",
    "    ax_1.set_ylim(-0.1, 1.1 * max(predicted_fold_change_list))\n",
    "    ax_1.tick_params(axis='both', which='major', labelsize=40)\n",
    "    ax_1.tick_params(axis='both', which='minor', labelsize=40)\n",
    "    ax_1.set_xlabel(\"Measured experimental log2 fold change\", size=60)\n",
    "    ax_1.set_ylabel(\"Model prediction's fold change\", size=45)\n",
    "#     if(pp > 0.05):\n",
    "#         p_asterisk = \"NS\"\n",
    "#     if(sp > 0.05):\n",
    "#         s_asterisk = \"NS\"\n",
    "    ax_1.set_title(\"Correlation between experimental and model predictions fold change\\nPearson = \"+\n",
    "                   str(round(pc, 2))+\n",
    "                   \" (\"+str(round(pp, 3))+\n",
    "                   \") Spearman = \"+\n",
    "                   str(round(sc, 2))+\n",
    "                   \" (\"+str(round(sp, 3))+\n",
    "                   \")\", size=40)\n",
    "\n",
    "    # Determine whether we are doing H3K27ac ISE in the background of another assay's features\n",
    "    # or we have marginal features and are doing that track's ISE\n",
    "    epigenetic_features = gene_features[:, -1] # MNase\n",
    "    epigenetic_features_2 = gene_features[:, 3] # H3K27ac\n",
    "    color_for_assay = assay_color[3]\n",
    "    label_for_assay = assays[3]\n",
    "\n",
    "    # Scale the model predictions     \n",
    "    scaling_ratio = np.median(measured_fold_change_list)/np.median(predicted_fold_change_list)\n",
    "    scaled_model_predictions = 0.5 * (scaling_ratio * (predicted_fold_change_list))\n",
    "\n",
    "    # Scale the epigenetic features\n",
    "    epigenetic_features_scaling_ratio = max(measured_fold_change_list)/max(epigenetic_features)\n",
    "    scaled_epigenetic_features = (epigenetic_features_scaling_ratio * (epigenetic_features))\n",
    "\n",
    "    epigenetic_features_scaling_ratio_2 = max(measured_fold_change_list)/max(epigenetic_features_2)\n",
    "    scaled_epigenetic_features_2 = (epigenetic_features_scaling_ratio_2 * (epigenetic_features_2))\n",
    "\n",
    "#     ax_2.plot(xticklabels, scaled_model_predictions, 'o-', color=\"#4daf4a\", linewidth=5, markersize=2, label=\"(Scaled) Model Predictions \" + label_for_assay)\n",
    "#     ax_2.plot(xticklabels, scaled_epigenetic_features, 'o-', color=\"#8470FF\", linewidth=5, markersize=1, label=\"(Scaled) Epigenetic Features MNase\") # + label_for_assay)\n",
    "#     ax_2.plot(xticklabels, scaled_epigenetic_features_2, 'o-', color=\"darkblue\", linewidth=5, markersize=1, label=\"(Scaled) Epigenetic Features H3K27ac\") # + label_for_assay)\n",
    "\n",
    "#     ax_2.bar(positions_TSS, (measured_fold_change_list), color=\"#f781bf\", bottom=0, width=2, label=\"Experimental mean from qPCR\")\n",
    "\n",
    "#     color_dict = {}\n",
    "#     color_dict[\"+\"] = {\"plus\":\"blue\", \"minus\":\"red\"}\n",
    "#     color_dict[\"-\"] = {\"plus\":\"red\", \"minus\":\"blue\"}\n",
    "#     direction = {\"blue\":\"rightward\", \"red\":\"leftward\"}\n",
    "#     color_index = 0\n",
    "#     for name, group in gRNA_strand_groups:\n",
    "#         ax_2.plot(group['Position_wrt_TSS'], 0.0 + (group['Measured fold change']), 'o', color=color_dict[STRANDS[gene]][name], label=\"qPCR facing \"+direction[color_dict[STRANDS[gene]][name]], markersize=15)\n",
    "#         color_index += 1\n",
    "\n",
    "#     ax_2.set_xlim(-operative_half_window_size-10, operative_half_window_size+10)\n",
    "#     ax_2.set_ylim(-1, 1.0 + max(measured_fold_change_list)*1.5)\n",
    "#     x_vals = ax_2.get_xticks()\n",
    "#     ax_2.set_xticklabels(['{:3.0f}'.format(x * RESOLUTION) for x in x_vals])\n",
    "#     ax_2.yaxis.set_major_locator(MaxNLocator(integer=True))\n",
    "#     ax_2.tick_params(axis='both', which='major', labelsize=35)\n",
    "#     ax_2.tick_params(axis='both', which='minor', labelsize=35)\n",
    "#     ax_2.set_xlabel(\"Peak Position (in bp) w.r.t TSS\", size=50)\n",
    "#     ax_2.set_ylabel(\"Gene expression fold change\", size=50)\n",
    "#     ax_2.set_title(gene+\" with H3K27ac + \"+assays[assay_index-1]+\"\\nincreasing \" + str(peak_width * RESOLUTION) + \"bp peaks by -log10(p_value)=\"+str(p_value_mapping(inserted_lnp1_minuslog10p_value)), size=40) #, y=1.1)\n",
    "\n",
    "#     ax_2.legend(loc='upper center', prop={'size': 30}, ncol=2)\n",
    "\n",
    "#     # Now also keep storing the actual TPM vs predicted TPM for each gene, for each cell type\n",
    "#     actual_TPM = np.load(\"../Data/\" + gene + \".T\" + '%02d' % cell_type_choice + \".TPM.npy\")\n",
    "# #             print(gene, cell_type_choice, (np.power(10, actual_TPM[0][0]) -1), yPred_TPM)\n",
    "#     TPM[gene][cell_type_choice] = ((np.power(10, actual_TPM[0][0]) -1), yPred_TPM)\n",
    "\n",
    "    plt.show()\n",
    "    plt.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "854eb32b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Analyze predicted TPMs across cell types for a given gene\n",
    "plt.rcParams[\"figure.figsize\"] = (10, 10)\n",
    "for gene in TPM.keys():\n",
    "    gene_TPM = [(v[0], v[1]) for v in TPM[gene].values()]\n",
    "#     print(gene, gene_TPM)\n",
    "\n",
    "    gene_TPM = np.asarray(gene_TPM)\n",
    "    pc, pp = pearsonr(gene_TPM[:, 0], gene_TPM[:, 1])\n",
    "    sc, sp = spearmanr(gene_TPM[:, 0], gene_TPM[:, 1])\n",
    "\n",
    "    plt.plot(gene_TPM[:, 0], gene_TPM[:, 1], 'o', markersize=10)\n",
    "    plt.xlabel(\"True TPM\", fontsize=20)\n",
    "    plt.ylabel(\"Predicted TPM\", fontsize=20)\n",
    "    plt.xlim(-1, 50)\n",
    "    plt.ylim(-1, 5)\n",
    "    plt.title(gene+ \": Pearson = \"+str(round(pc, 2))+\" Spearman = \"+str(round(sc, 2)), fontsize=25)\n",
    "    plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa0b7796",
   "metadata": {},
   "outputs": [],
   "source": [
    "list( zip( [x[1] for x in TPM[\"CXCR4\"].values()], [x[1] for x in TPM[\"TGFBR1\"].values()] ) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ce5a1da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Analysis of MNase-seq and H3K27ac in K562\n",
    "import pyBigWig\n",
    "K562_H3K27ac = pyBigWig.open(\"../Data/MNase_H3K27ac/K562.H3K27ac.ENCFF469JMR.bigWig\")\n",
    "K562_MNase = pyBigWig.open(\"../Data/MNase_H3K27ac/K562.MNase.hg38.sorted.chromosomes.non_overlapping.ENCFF000VNN.bigWig\")\n",
    "chrom = \"chr9\"\n",
    "K562_H3K27ac_values = np.nan_to_num( K562_H3K27ac.values(chrom, 0, K562_H3K27ac.chroms(chrom), numpy=True) )\n",
    "K562_MNase_values = np.nan_to_num( K562_MNase.values(chrom, 0, K562_MNase.chroms(chrom), numpy=True) )\n",
    "\n",
    "print(K562_H3K27ac_values.shape, K562_MNase_values.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c940f3a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (10, 10)\n",
    "plt.plot(K562_MNase_values, K562_H3K27ac_values, 'o', markersize=0.1)\n",
    "plt.xlabel(\"MNase-seq __\", fontsize=20)\n",
    "plt.ylabel(\"H3K27ac -log10(p-values)\", fontsize=20)\n",
    "plt.xlim(-1, 20)\n",
    "plt.ylim(-1, 200)\n",
    "plt.title(gene+ \": Pearson = \"+str(round(pc, 2))+\" Spearman = \"+str(round(sc, 2)), fontsize=25)\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88dcc85f",
   "metadata": {},
   "outputs": [],
   "source": [
    "for mnase_level in [0, 1, 2, 3, 4, 5, 10, 20]:\n",
    "    for H3K27ac_threshold in [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 20]:\n",
    "        print(mnase_level, H3K27ac_threshold, np.where(K562_H3K27ac_values[np.where(K562_MNase_values >= mnase_level)] >= H3K27ac_threshold)[0].shape )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b97547c6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d2109d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Update with CRISPOR.tsv for both DHS and sat (why does sat CRISPOR have so many missing gRNAs?)\n",
    "df_HER2_DHS = pd.read_csv(\"../Data/HER2_MAGeCK/HER2.DHS.hg19.MAGeCK.fold_change\", sep=\"\\t\")\n",
    "df_HER2_sat = pd.read_csv(\"../Data/HER2_MAGeCK/HER2.sat.hg19.MAGeCK.fold_change\", sep=\"\\t\")\n",
    "\n",
    "fig, axes = plt.subplots(nrows=2, ncols=3, figsize=(30, 15), sharey=False)\n",
    "# fig.tight_layout(pad=1, w_pad=20, h_pad=25)\n",
    "\n",
    "axes[0, 0].plot(df_HER2_DHS['H_L_LFC'], df_HER2_DHS['U_L_LFC'], 'o')\n",
    "axes[0, 0].set_xlim(-5, 5)\n",
    "axes[0, 0].set_ylim(-5, 5)\n",
    "axes[0, 0].set_xlabel(\"High vs Low Log Fold Change\", size=20)\n",
    "axes[0, 0].set_ylabel(\"Unsorted vs Low Log Fold Change\", size=20)\n",
    "axes[0, 0].set_title(\"DHS: Spearman = \" + str(round( spearmanr(df_HER2_DHS['H_L_LFC'], df_HER2_DHS['U_L_LFC'])[0], 3) ), size=30)\n",
    "\n",
    "axes[0, 1].plot(df_HER2_DHS['H_L_LFC'], df_HER2_DHS['H_U_LFC'], 'o')\n",
    "axes[0, 1].set_xlim(-5, 5)\n",
    "axes[0, 1].set_ylim(-5, 5)\n",
    "axes[0, 1].set_xlabel(\"High vs Low Log Fold Change\", size=20)\n",
    "axes[0, 1].set_ylabel(\"High vs Unsorted Log Fold Change\", size=20)\n",
    "axes[0, 1].set_title(\"DHS: Spearman = \" + str(round( spearmanr(df_HER2_DHS['H_L_LFC'], df_HER2_DHS['H_U_LFC'])[0], 3) ), size=30)\n",
    "\n",
    "axes[0, 2].plot(df_HER2_DHS['H_U_LFC'], df_HER2_DHS['U_L_LFC'], 'o')\n",
    "axes[0, 2].set_xlim(-5, 5)\n",
    "axes[0, 2].set_ylim(-5, 5)\n",
    "axes[0, 2].set_xlabel(\"High vs Unsorted Log Fold Change\", size=20)\n",
    "axes[0, 2].set_ylabel(\"Unsorted vs Low Log Fold Change\", size=20)\n",
    "axes[0, 2].set_title(\"DHS: Spearman = \" + str(round( spearmanr(df_HER2_DHS['H_U_LFC'], df_HER2_DHS['U_L_LFC'])[0], 3) ), size=30)\n",
    "\n",
    "axes[1, 0].plot(df_HER2_sat['H_L_LFC'], df_HER2_sat['U_L_LFC'], 'o')\n",
    "axes[1, 0].set_xlim(-5, 5)\n",
    "axes[1, 0].set_ylim(-5, 5)\n",
    "axes[1, 0].set_xlabel(\"High vs Low Log Fold Change\", size=20)\n",
    "axes[1, 0].set_ylabel(\"Unsorted vs Low Log Fold Change\", size=20)\n",
    "axes[1, 0].set_title(\"sat: Spearman = \" + str(round( spearmanr(df_HER2_sat['H_L_LFC'], df_HER2_sat['U_L_LFC'])[0], 3) ), size=30)\n",
    "\n",
    "axes[1, 1].plot(df_HER2_sat['H_L_LFC'], df_HER2_sat['H_U_LFC'], 'o')\n",
    "axes[1, 1].set_xlim(-5, 5)\n",
    "axes[1, 1].set_ylim(-5, 5)\n",
    "axes[1, 1].set_xlabel(\"High vs Low Log Fold Change\", size=20)\n",
    "axes[1, 1].set_ylabel(\"High vs Unsorted Log Fold Change\", size=20)\n",
    "axes[1, 1].set_title(\"sat: Spearman = \" + str(round( spearmanr(df_HER2_sat['H_L_LFC'], df_HER2_sat['H_U_LFC'])[0], 3) ), size=30)\n",
    "\n",
    "axes[1, 2].plot(df_HER2_sat['H_U_LFC'], df_HER2_sat['U_L_LFC'], 'o')\n",
    "axes[1, 2].set_xlim(-5, 5)\n",
    "axes[1, 2].set_ylim(-5, 5)\n",
    "axes[1, 2].set_xlabel(\"High vs Unsorted Log Fold Change\", size=20)\n",
    "axes[1, 2].set_ylabel(\"Unsorted vs Low Log Fold Change\", size=20)\n",
    "axes[1, 2].set_title(\"sat: Spearman = \" + str(round( spearmanr(df_HER2_sat['H_U_LFC'], df_HER2_sat['U_L_LFC'])[0], 3) ), size=30)\n",
    "\n",
    "\n",
    "# ax = plt.axes(projection =\"3d\")\n",
    "# ax.plot3D(df_HER2['H_L_LFC'], df_HER2['H_U_LFC'], df_HER2['U_L_LFC'], 'o', color = \"green\", markersize=1)\n",
    "# ax.set_xlabel('H_L_LFC')\n",
    "# ax.set_ylabel('H_U_LFC')\n",
    "# ax.set_zlabel('U_L_LFC')\n",
    "# ax.view_init(10, 120)\n",
    "\n",
    "plt.show()\n",
    "# plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "552c77b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Analyze gRNA scores for HER2 in juxtaposition with MAGeCK LFC and FDR\n",
    "\n",
    "def parse_df(file_name):\n",
    "    df_DHS = pd.read_csv(file_name, sep=\"\\t\", index_col=False,\n",
    "                                 names=[\"gRNA\", \"Protospacer_PAM\", \"CRISPOR_search_window\", \"CRISPOR_pos_strand\",\n",
    "                                        \"CRISPOR_Protospacer_PAM\", \"mitSpecScore\", \"cfdSpecScore\", \"offtargetCount\",\n",
    "                                        \"targetGenomeGeneLocus\", \"Doench16Score\", \"Moreno-Mateos-Score\", \"Out-Of-Frame-Score\", \"Lindel-Score\", \"GrafEtAlStatus\",\n",
    "                                        \"chromosome\", \"start\", \"end\", \"gRNA_2\", \"Protospacer\", \"PAM\",\n",
    "                                        \"H_L_LFC\", \"H_U_LFC\", \"U_L_LFC\", \n",
    "                                        \"H_L_FDR\", \"H_U_FDR\", \"U_L_FDR\"])\n",
    "    df_DHS[\"strand\"] = [(lambda x: \"+\" if x == \"w\" else \"-\")(x[-1]) for x in df_DHS[\"CRISPOR_pos_strand\"]]\n",
    "    df_DHS[\"d_TSS\"] = [(x - 37844347) for x in df_DHS[\"start\"]] # TODO: Refine taking into account strand information\n",
    "    df_DHS[\"log10p1_d_TSS\"] = [np.log10(abs(x - 37844347) + 1) for x in df_DHS[\"start\"]]\n",
    "    \n",
    "    return df_DHS\n",
    "\n",
    "df_training = parse_df(\"../Data/HER2_MAGeCK/HER2.sat.hg19.MAGeCK.CRISPOR.tsv\")\n",
    "df_testing = parse_df(\"../Data/HER2_MAGeCK/HER2.DHS.hg19.MAGeCK.CRISPOR.tsv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05788c6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_testing[df_testing['H_L_FDR']<0.05]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebb04ea0",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (10, 10)\n",
    "plt.plot(df_training['start'], df_training['H_L_FDR'], 'o')\n",
    "plt.axvline(x=37844347, color='grey')\n",
    "plt.ylabel(\"FDR q-value\")\n",
    "plt.xlim(37844347 - 50000, 37844347 + 50000)\n",
    "plt.ylim(-1.1, 1.1)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "255c3241",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_common = pd.read_csv(\"../Data/HER2_MAGeCK/T1\", sep=\"\\t\", index_col=False,\n",
    "                       names=[\"gRNA_1\", \"Protospacer_PAM_1\", \"CRISPOR_search_window_1\", \"CRISPOR_pos_strand_1\",\n",
    "                                        \"CRISPOR_Protospacer_PAM_1\", \"mitSpecScore_1\", \"cfdSpecScore_1\", \"offtargetCount_1\",\n",
    "                                        \"targetGenomeGeneLocus_1\", \"Doench16Score_1\", \"Moreno-Mateos-Score_1\", \"Out-Of-Frame-Score_1\", \"Lindel-Score_1\", \"GrafEtAlStatus_1\",\n",
    "                                        \"chromosome_1\", \"start_1\", \"end_1\", \"gRNA_2_1\", \"Protospacer_1\", \"PAM_1\",\n",
    "                                        \"H_L_LFC_1\", \"H_U_LFC_1\", \"U_L_LFC_1\", \n",
    "                                        \"H_L_FDR_1\", \"H_U_FDR_1\", \"U_L_FDR_1\",\n",
    "                              \"gRNA_2\", \"Protospacer_PAM_2\", \"CRISPOR_search_window_2\", \"CRISPOR_pos_strand_2\",\n",
    "                                        \"CRISPOR_Protospacer_PAM_2\", \"mitSpecScore_2\", \"cfdSpecScore_2\", \"offtargetCount_2\",\n",
    "                                        \"targetGenomeGeneLocus_2\", \"Doench16Score_2\", \"Moreno-Mateos-Score_2\", \"Out-Of-Frame-Score_2\", \"Lindel-Score_2\", \"GrafEtAlStatus_2\",\n",
    "                                        \"chromosome_2\", \"start_2\", \"end_2\", \"gRNA_2_2\", \"Protospacer_2\", \"PAM_2\",\n",
    "                                        \"H_L_LFC_2\", \"H_U_LFC_2\", \"U_L_LFC_2\", \n",
    "                                        \"H_L_FDR_2\", \"H_U_FDR_2\", \"U_L_FDR_2\"])\n",
    "df_common\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = (10, 10)\n",
    "plt.plot(df_common['H_L_FDR_1'], df_common['H_L_FDR_2'], 'o', markersize=5, color=\"#FF1493\")\n",
    "plt.title(\"Concordance between 1,076 gRNAs common to DHS and Saturation screen\", fontsize=20)\n",
    "plt.xlabel(\"DHS gRNA FDR q-value\", fontsize=20)\n",
    "plt.ylabel(\"Satration gRNA FDR q-value\", fontsize=20)\n",
    "plt.xlim(-0.1, 1.1)\n",
    "plt.ylim(-0.1, 1.1)\n",
    "plt.show()\n",
    "plt.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d14e44f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_common[(df_common['H_L_FDR_1'] > 0.8) & (df_common['H_L_FDR_2'] > 0.8)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "340c9841",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now compute pairs of gRNAs that are within D (for now 1) of each other on the \n",
    "D = 100\n",
    "FDR_threshold = 0.05\n",
    "\n",
    "# close_gRNAs = [x for x in [(lambda v: v[0] if ( (v[1][2] - v[1][0] < D) and (v[1][1] == v[1][3]) ) else -1)(x) for x in enumerate(list(zip(df_DHS[\"start\"][0:-1], df_DHS[\"strand\"][0:-1], df_DHS[\"start\"][1:], df_DHS[\"strand\"][1:])))] if x >= 0]\n",
    "\n",
    "def find_close_gRNAs(df_DHS):\n",
    "\n",
    "    starts = df_DHS[\"start\"]\n",
    "    strands = df_DHS[\"strand\"]\n",
    "    FDRs = df_DHS[\"H_L_FDR\"]\n",
    "\n",
    "    close_gRNAs = []\n",
    "    random_gRNAs = []\n",
    "    for i in tqdm(range(len(df_DHS))):\n",
    "        for j in range(max(0, i-100), min(len(df_DHS), i+100)):\n",
    "            if(i == j):\n",
    "                continue\n",
    "            else:\n",
    "                if( (strands[i] == strands[i]) and # This controls whether we require the strands to be the same \n",
    "                    (np.abs(starts[i] - starts[j]) < D) and\n",
    "                    (min(FDRs[i], FDRs[j]) < FDR_threshold) and\n",
    "                    (max(FDRs[i], FDRs[j]) >= FDR_threshold) ):\n",
    "\n",
    "                    u, v = random.sample(range(len(df_DHS)), 2)\n",
    "                    random_gRNAs.append((u, v))\n",
    "\n",
    "                    if(FDRs[i] > FDRs[j]):\n",
    "                        close_gRNAs.append((i, j))\n",
    "                    else:\n",
    "                        close_gRNAs.append((j, i))\n",
    "\n",
    "    close_gRNAs = np.unique(close_gRNAs, axis=0)                    \n",
    "    print(D, len(close_gRNAs), len(df_DHS))\n",
    "    \n",
    "    return close_gRNAs, random_gRNAs\n",
    "\n",
    "close_gRNAs_training, random_gRNAs_training = find_close_gRNAs(df_training)\n",
    "close_gRNAs_testing, random_gRNAs_testing = find_close_gRNAs(df_testing)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bee9ccbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Train logistic regression for binary classification\n",
    "from sklearn.metrics import average_precision_score\n",
    "from sklearn.preprocessing import Normalizer\n",
    "import random\n",
    "\n",
    "def create_augmented_data(N, features_class_1):\n",
    "    \n",
    "    features_class_0 = []\n",
    "    n = features_class_1.shape[0]\n",
    "    D = features_class_1.shape[1]\n",
    "    for i in range(N):\n",
    "        feature_class_1 = features_class_1[random.randint(0, n-1), :]\n",
    "        \n",
    "        reduced_feature_class_0 = np.copy(feature_class_1)\n",
    "        sampled_feature = random.randint(3, 3)\n",
    "        reduced_feature_class_0[sampled_feature] = max(reduced_feature_class_0[sampled_feature] - 30, 0)\n",
    "\n",
    "        features_class_0.append(reduced_feature_class_0)\n",
    "        \n",
    "    return np.asarray(features_class_0)\n",
    "\n",
    "\n",
    "def create_features_for_classification(df_DHS, close_gRNAs, N_augmented):\n",
    "    \n",
    "    features_class_0 = np.unique(np.asarray([list(df_DHS.iloc[x[0], np.r_[5, 6, 7, 9, 10]]) for x in close_gRNAs]), axis=0)\n",
    "    features_class_1 = np.unique(np.asarray([list(df_DHS.iloc[x[1], np.r_[5, 6, 7, 9, 10]]) for x in close_gRNAs]), axis=0)\n",
    "\n",
    "    xTrain_augmented = create_augmented_data(N_augmented, features_class_1)\n",
    "    fp = features_class_1.shape[0] / (features_class_0.shape[0] + features_class_1.shape[0] + N_augmented)\n",
    "\n",
    "#     print(features_class_0.shape, features_class_1.shape, xTrain_augmented.shape)\n",
    "    \n",
    "    xTrain = np.concatenate([features_class_0, features_class_1, xTrain_augmented], axis=0) \n",
    "    yTrain = np.concatenate([[0]*len(features_class_0), [1]*len(features_class_1), [0] * len(xTrain_augmented)], \n",
    "                            axis=0)\n",
    "\n",
    "    return xTrain, yTrain, fp\n",
    "\n",
    "xTrain, yTrain, fpTrain = create_features_for_classification(df_training, close_gRNAs_training, 10)\n",
    "xTest, yTest, fpTest = create_features_for_classification(df_testing, close_gRNAs_testing, 1)\n",
    "\n",
    "normalizer = Normalizer().fit(xTrain)                                             \n",
    "xTrain_normalized = normalizer.transform(xTrain)\n",
    "xTest_normalized = normalizer.transform(xTest)\n",
    "\n",
    "subsampled_xTrain_negatives = [i for i in range(len(xTrain_normalized)) if ((yTrain[i] == 0) and \n",
    "                                                                            (random.uniform(0, 1) < 0.2))]\n",
    "xTrain_subsampled_negatives = xTrain[subsampled_xTrain_negatives, :]\n",
    "yTrain_subsampled_negatives = yTrain[subsampled_xTrain_negatives]\n",
    "\n",
    "subsampled_xTrain_positives = [i for i in range(len(xTrain_normalized)) if ((yTrain[i] == 1) and \n",
    "                                                                            (random.uniform(0, 1) < 2))]\n",
    "xTrain_subsampled_positives = xTrain[subsampled_xTrain_positives, :]\n",
    "yTrain_subsampled_positives = yTrain[subsampled_xTrain_positives]\n",
    "\n",
    "xTrain_normalized = np.concatenate([xTrain_subsampled_positives, xTrain_subsampled_negatives], axis=0)\n",
    "yTrain_normalized = np.concatenate([yTrain_subsampled_positives, yTrain_subsampled_negatives], axis=0)\n",
    "\n",
    "model = logreg(random_state=0).fit(xTrain_normalized, yTrain_normalized)\n",
    "print(list(zip(gRNA_scores.keys(), [round(x, 3) for x in model.coef_[0]])))\n",
    "yTrain_pred = model.predict(xTrain_normalized)\n",
    "yTest_pred = model.predict(xTest_normalized)\n",
    "\n",
    "print(\"Fraction of positives training:\", \n",
    "      round(len(xTrain_subsampled_positives) / (len(xTrain_subsampled_positives) + \n",
    "                                                len(xTrain_subsampled_negatives)), 3), \n",
    "      \"\\tAUPRC training:\", round(average_precision_score(yTrain_normalized, yTrain_pred), 3))\n",
    "\n",
    "print(\"Fraction of positives testing:\", round(fpTest, 3), \"\\tAUPRC testing:\", \n",
    "      round(average_precision_score(yTest, yTest_pred), 3))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6c0167a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Train logistic regression for binary classification but with differences of scores\n",
    "from sklearn.metrics import average_precision_score\n",
    "from sklearn import svm\n",
    "from sklearn.ensemble import RandomForestClassifier as rfc\n",
    "import random\n",
    "\n",
    "def create_augmented_data(N, features_class_1):\n",
    "    \n",
    "    xTrain_augmented = []\n",
    "    yTrain_augmented = []\n",
    "    n = features_class_1.shape[0]\n",
    "    D = features_class_1.shape[1]\n",
    "    for i in range(N):\n",
    "        feature_class_1 = features_class_1[random.randint(0, n-1), :]\n",
    "        \n",
    "        reduced_feature_class_0 = np.copy(feature_class_1)\n",
    "        sampled_feature = random.randint(0, 4)\n",
    "        if(sampled_feature == 0):\n",
    "            reduced_feature_class_0[sampled_feature] = max(reduced_feature_class_0[sampled_feature] - 2, 0)\n",
    "        elif(sampled_feature == 1):\n",
    "            reduced_feature_class_0[sampled_feature] = max(reduced_feature_class_0[sampled_feature] - 2, 0)\n",
    "        elif(sampled_feature == 2):\n",
    "            reduced_feature_class_0[sampled_feature] = min(reduced_feature_class_0[sampled_feature] + 2, 1000)\n",
    "        elif(sampled_feature == 3):\n",
    "            reduced_feature_class_0[sampled_feature] = max(reduced_feature_class_0[sampled_feature] - 2, 0)\n",
    "        elif(sampled_feature == 4):\n",
    "            reduced_feature_class_0[sampled_feature] = max(reduced_feature_class_0[sampled_feature] - 20, 0)\n",
    "    \n",
    "        positive_data_point = feature_class_1 - reduced_feature_class_0\n",
    "        negative_data_point = - feature_class_1 + reduced_feature_class_0\n",
    "    \n",
    "        xTrain_augmented.append(list(feature_class_1) + list(reduced_feature_class_0))\n",
    "        xTrain_augmented.append(list(reduced_feature_class_0) + list(feature_class_1))\n",
    "        \n",
    "        yTrain_augmented.append(1)\n",
    "        yTrain_augmented.append(0)\n",
    "        \n",
    "    return np.asarray(xTrain_augmented), np.asarray(yTrain_augmented)\n",
    "\n",
    "\n",
    "def create_features_for_classification(df_DHS, close_gRNAs, N_augmented):\n",
    "    \n",
    "    features_class_0 = np.asarray([list(df_DHS.iloc[x[0], np.r_[5, 6, 7, 9, 10]]) for x in close_gRNAs])\n",
    "    features_class_1 = np.asarray([list(df_DHS.iloc[x[1], np.r_[5, 6, 7, 9, 10]]) for x in close_gRNAs])\n",
    "\n",
    "    positive_data_points = [list(features_class_1[i, :]) + list(features_class_0[i, :]) for i in range(features_class_1.shape[0])]\n",
    "    negative_data_points = [list(features_class_0[i, :]) + list(features_class_1[i, :]) for i in range(features_class_1.shape[0])]    \n",
    "    \n",
    "    xTrain_augmented, yTrain_augmented = create_augmented_data(N_augmented, features_class_1)\n",
    "    fp = features_class_1.shape[0] / (features_class_0.shape[0] + features_class_1.shape[0] + len(xTrain_augmented))\n",
    "\n",
    "    print(features_class_0.shape, features_class_1.shape, xTrain_augmented.shape)\n",
    "    \n",
    "    xTrain = np.concatenate([negative_data_points, positive_data_points, xTrain_augmented], axis=0) \n",
    "    yTrain = np.concatenate([[0]*len(negative_data_points), [1]*len(positive_data_points), yTrain_augmented], axis=0)\n",
    "\n",
    "    print(\"x shape: \", xTrain.shape)\n",
    "    \n",
    "    return xTrain, yTrain, fp\n",
    "\n",
    "xTrain, yTrain, fpTrain = create_features_for_classification(df_training, close_gRNAs_training, 10000)\n",
    "xTest, yTest, fpTest = create_features_for_classification(df_testing, close_gRNAs_testing, 1)\n",
    "\n",
    "# model = logreg(random_state=0).fit(xTrain, yTrain)\n",
    "# model = svm.SVC(kernel='rbf', C=1).fit(xTrain, yTrain)\n",
    "model = rfc(max_depth=2, random_state=0).fit(xTrain, yTrain)\n",
    "# print(list(zip(gRNA_scores.keys(), [round(x, 3) for x in model.coef_[0]])))\n",
    "yTrain_pred = model.predict(xTrain)\n",
    "yTest_pred = model.predict(xTest)\n",
    "\n",
    "print(\"Fraction of positives training:\", round(fpTrain, 3), \"\\tAUPRC training:\", round(average_precision_score(yTrain, yTrain_pred), 3))\n",
    "print(\"Fraction of positives testing:\", round(fpTest, 3), \"\\tAUPRC testing:\", round(average_precision_score(yTest, yTest_pred), 3))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c7a7410",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3da99ed3",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "df_DHS = df_training\n",
    "close_gRNAs = close_gRNAs_training\n",
    "random_gRNAs = random_gRNAs_training\n",
    "\n",
    "gRNA_score_names = {1:\"mitSpecScore\", 2:\"cfdSpecScore\", 3:\"offtargetCount\", 4:\"Doench16Score\", 5:\"Moreno-Mateos-Score\", 6:\"Out-Of-Frame-Score\", 7:\"Lindel-Score\", 8:\"log10p1_d_TSS\"}\n",
    "gRNA_scores = {\"mitSpecScore\":5, \"cfdSpecScore\":6, \"offtargetCount\":7, \"Doench16Score\":9, \"Moreno-Mateos-Score\":10, \"Out-Of-Frame-Score\":11, \"Lindel-Score\":12, \"log10p1_d_TSS\":28}\n",
    "\n",
    "axis_limits = {1:(-25, 125), 2:(-25, 125), 3:(-100, 1000), 4:(-25, 125), 5:(-25, 125), 6:(-25, 125), 7:(-25, 125), 8:(-1, 10)}\n",
    "\n",
    "for gRNA_score_index in range(1, 9):\n",
    "    gRNA_score_name = gRNA_score_names[gRNA_score_index]\n",
    "\n",
    "    x_1 = [float(df_DHS.iloc[x[0], gRNA_scores[gRNA_score_name]]) for x in close_gRNAs]\n",
    "    x_2 = [float(df_DHS.iloc[x[1], gRNA_scores[gRNA_score_name]]) for x in close_gRNAs]\n",
    "\n",
    "    random_x_1 = [float(df_DHS.iloc[x[0], gRNA_scores[gRNA_score_name]]) for x in random_gRNAs]\n",
    "    random_x_2 = [float(df_DHS.iloc[x[1], gRNA_scores[gRNA_score_name]]) for x in random_gRNAs]\n",
    "\n",
    "    plt.rcParams[\"figure.figsize\"] = (10, 10)\n",
    "    plt.scatter(random_x_1, random_x_2, 10, color=\"darkgrey\")\n",
    "    plt.scatter(x_1, x_2, 2, color=\"purple\")\n",
    "    plt.plot([axis_limits[gRNA_score_index][0], axis_limits[gRNA_score_index][1]], [axis_limits[gRNA_score_index][0], axis_limits[gRNA_score_index][1]], 'o-', color=\"grey\")\n",
    "    plt.title(\"Score: \" + gRNA_score_name, fontsize=30)\n",
    "    plt.xlabel(\"gRNA with non-significant FC\", fontsize=20)\n",
    "    plt.ylabel(\"gRNA with significant FC\", fontsize=20)\n",
    "    plt.xlim(axis_limits[gRNA_score_index][0], axis_limits[gRNA_score_index][1])\n",
    "    plt.ylim(axis_limits[gRNA_score_index][0], axis_limits[gRNA_score_index][1])\n",
    "    plt.legend(loc=\"upper left\", labels=[\"y=x\", \"close gRNAs\", \"random gRNAs\"], fontsize=15)\n",
    "    plt.show()\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2968d174",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
